diff --git a/radicale/storage/multifilesystem.py b/radicale/storage/multifilesystem.py
index ca2d70c..647d7e7 100644
--- a/radicale/storage/multifilesystem.py
+++ b/radicale/storage/multifilesystem.py
@@ -48,30 +48,31 @@ class Collection(filesystem.Collection):
         headers = headers or self.headers
         items = items if items is not None else self.items
         timezones = list(set(i for i in items if isinstance(i, ical.Timezone)))
         components = [i for i in items if isinstance(i, ical.Component)]
         for component in components:
             text = ical.serialize(self.tag, headers, [component] + timezones)
             name = (
                 component.name if sys.version_info[0] >= 3 else
                 component.name.encode(filesystem.FILESYSTEM_ENCODING))
             path = os.path.join(self._path, name)
             with filesystem.open(path, "w") as fd:
                 fd.write(text)
 
     def delete(self):
         shutil.rmtree(self._path)
+        os.remove(self._props_path)
 
     def remove(self, name):
         if os.path.exists(os.path.join(self._path, name)):
             os.remove(os.path.join(self._path, name))
 
     @property
     def text(self):
         components = (
             ical.Timezone, ical.Event, ical.Todo, ical.Journal, ical.Card)
         items = set()
         try:
             for filename in os.listdir(self._path):
                 with filesystem.open(os.path.join(self._path, filename)) as fd:
                     items.update(self._parse(fd.read(), components))
         except IOError:
