diff --git a/.gitignore b/.gitignore
index beed18f..008153b 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,2 +1,3 @@
 .idea/
 bower_components/
+node_modules/
diff --git a/bower.json b/bower.json
index ccfdd5f..3f8493b 100644
--- a/bower.json
+++ b/bower.json
@@ -1,20 +1,22 @@
 {
   "name": "i18n-node-angular",
+  "version": "0.2.0-beta1",
   "description": "i18n-node for Angular",
-  "version": "0.1.2",
+  "main": "i18n-node-angular.js",
   "homepage": "https://github.com/oliversalzburg/i18n-node-angular",
   "authors": [
     "Oliver Salzburg <oliver.salzburg@gmail.com>"
   ],
-  "main": "i18n-node-angular.js",
   "license": "MIT",
   "private": false,
   "ignore": [
     "**/.*",
-    "bower_components"
+    "bower_components",
+    "node_modules",
+    "i18n-node-routes.js"
   ],
   "dependencies": {
     "angular": "~1.2.9",
     "sprintf-js": "git://github.com/alexei/sprintf.js.git"
   }
 }
diff --git a/i18n-node-routes.js b/i18n-node-routes.js
new file mode 100644
index 0000000..bd640d4
--- /dev/null
+++ b/i18n-node-routes.js
@@ -0,0 +1,77 @@
+/**
+ * Copyright (C) 2014, Oliver Salzburg
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Created: 2014-01-21 19:40
+ *
+ * @author Oliver Salzburg
+ * @copyright Copyright (C) 2014, Oliver Salzburg
+ * @license http://opensource.org/licenses/mit-license.php MIT License
+ */
+
+var i18n = require( "i18n" );
+
+var configuration = {
+  directory : "locales/",
+  extension : ".json"
+};
+
+/**
+ * Configure the express routes through which translations are served.
+ * @param app
+ * @param {Object} [configObject]
+ */
+var configure = function( app, configObject ) {
+  if( typeof configObject !== "undefined" ) {
+    configuration.directory = ( typeof configObject.directory === "string" ) ? configObject.directory : configuration.directory;
+    configuration.extension = ( typeof configObject.extension === "string" ) ? configObject.extension : configuration.extension;
+  }
+
+  // Register routes
+  app.get( "/i18n/:locale", i18nRoutes.i18n );
+  app.get( "/i18n/:locale/:phrase", i18nRoutes.translate );
+};
+
+var i18nRoutes =
+{
+  /**
+   * Sends a translation file to the client.
+   * @param request
+   * @param response
+   */
+  i18n : function( request, response ) {
+    var locale = request.params.locale;
+    response.sendfile( configuration.directory + locale + configuration.extension );
+  },
+
+  /**
+   * Translate a given string and provide the result.
+   * @param request
+   * @param response
+   */
+  translate : function( request, response ) {
+    var locale = request.params.locale;
+    var phrase = request.params.phrase;
+    var result = i18n.__( {phrase : phrase, locale : locale} );
+    response.send( result );
+  }
+};
+
+module.exports.configure = configure;
\ No newline at end of file
diff --git a/package.json b/package.json
new file mode 100644
index 0000000..85880fe
--- /dev/null
+++ b/package.json
@@ -0,0 +1,24 @@
+{
+  "name": "i18n-node-angular",
+  "version": "0.2.0-beta1",
+  "description": "express routes for i18n-node-angular",
+  "main": "i18n-node-routes.js",
+  "homepage": "https://github.com/oliversalzburg/i18n-node-angular",
+  "author": "Oliver Salzburg",
+  "license": "MIT",
+  "files":[
+    "i18n-node-routes.js"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/oliversalzburg/i18n-node-angular.git"
+  },
+  "keywords": [
+    "i18n",
+    "i18n-node",
+    "angular"
+  ],
+  "dependencies": {
+    "i18n": "~0.4.1"
+  }
+}
