diff --git a/index.js b/index.js
index 798c4be9..b781692a 100755
--- a/index.js
+++ b/index.js
@@ -6085,31 +6085,31 @@ F.compile_virtual = function(res) {
 			delete F.temporary.processing[req.$key];
 			res.$file();
 		}
 	});
 
 	return;
 };
 
 function compile_check(res) {
 
 	var req = res.req;
 	var uri = req.uri;
 
 	if (F.routes.merge[uri.pathname]) {
 		compile_merge(res);
-		return F;
+		return;
 	}
 
 	fsFileExists(res.options.filename, function(e, size, sfile, stats) {
 
 		if (e) {
 
 			if (!res.noCompress && COMPRESSIONSPECIAL[req.extension] && CONF.allow_compile && !REG_NOCOMPRESS.test(res.options.filename))
 				return compile_file(res);
 
 			var tmp = [res.options.filename, size, stats.mtime.toUTCString()];
 			if (CONF.allow_gzip && COMPRESSION[U.getContentType(req.extension)]) {
 				compile_gzip(tmp, function(tmp) {
 					F.temporary.path[req.$key] = tmp;
 					res.$file();
 					delete F.temporary.processing[req.$key];
@@ -7314,45 +7314,31 @@ F.service = function(count) {
  */
 F.listener = function(req, res) {
 
 	req.options = res.options = {};
 	res.req = req;
 	req.res = res;
 
 	if (F._length_wait)
 		return F.response503(req, res);
 	else if (!req.host) // HTTP 1.0 without host
 		return res.throw400();
 
 	var headers = req.headers;
 	req.$protocol = ((req.connection && req.connection.encrypted) || ((headers['x-forwarded-proto'] || ['x-forwarded-protocol']) === 'https')) ? 'https' : 'http';
 
-	var beg = 0;
-
-	// Removes directory browsing
-	for (var i = 0; i < req.url.length; i++) {
-		if (req.url[i] === '.' && req.url[i + 1] === '/')
-			beg = i + 1;
-		else if (req.url[i] === '?')
-			break;
-	}
-
-	if (beg)
-		req.url = req.url.substring(beg);
-
 	req.uri = framework_internal.parseURI(req);
-
 	F.stats.request.request++;
 	F.$events.request && EMIT('request', req, res);
 
 	if (F._request_check_redirect) {
 		var redirect = F.routes.redirects[req.$protocol + '://' + req.host];
 		if (redirect) {
 			F.stats.response.forward++;
 			res.options.url = redirect.url + (redirect.path ? req.url : '');
 			res.options.permanent = redirect.permanent;
 			res.$redirect();
 			return;
 		}
 	}
 
 	req.path = framework_internal.routeSplit(req.uri.pathname);
@@ -7386,30 +7372,38 @@ function requestcontinue_middleware(req, res)  {
  * Continue to process
  * @private
  * @param {Request} req
  * @param {Response} res
  * @param {Object} headers
  * @param {String} protocol [description]
  * @return {Framework}
  */
 F.$requestcontinue = function(req, res, headers) {
 
 	if (!req || !res || res.headersSent || res.success)
 		return;
 
 	// Validates if this request is the file (static file)
 	if (req.isStaticFile) {
+
+		// Stops path travelsation outside of "public" directory
+		// A potential security issue
+		if (req.uri.pathname.indexOf('./') !== -1) {
+			req.$total_status(404);
+			return;
+		}
+
 		F.stats.request.file++;
 		if (F._length_files)
 			req.$total_file();
 		else
 			res.continue();
 		return;
 	}
 
 	if (!PERF[req.method]) {
 		req.$total_status(404);
 		return;
 	}
 
 	F.stats.request.web++;
 
