diff --git a/src/lxml/html/clean.py b/src/lxml/html/clean.py
index d619b77b..91b43947 100644
--- a/src/lxml/html/clean.py
+++ b/src/lxml/html/clean.py
@@ -32,52 +32,43 @@ __all__ = ['clean_html', 'clean', 'Cleaner', 'autolink', 'autolink_html',
 
 
 # This is an IE-specific construct you can have in a stylesheet to
 # run some Javascript:
 _css_javascript_re = re.compile(
     r'expression\s*\(.*?\)', re.S|re.I)
 
 # Do I have to worry about @\nimport?
 _css_import_re = re.compile(
     r'@\s*import', re.I)
 
 # All kinds of schemes besides just javascript: that can cause
 # execution:
 _javascript_scheme_re = re.compile(
     r'\s*(?:javascript|jscript|livescript|vbscript|about|mocha):', re.I)
-_whitespace_re = re.compile(r'\s+')
+_substitute_whitespace = re.compile(r'\s+').sub
 # FIXME: should data: be blocked?
 
 # FIXME: check against: http://msdn2.microsoft.com/en-us/library/ms537512.aspx
 _conditional_comment_re = re.compile(
     r'\[if[\s\n\r]+.*?][\s\n\r]*>', re.I|re.S)
 
 _find_styled_elements = etree.XPath(
     "descendant-or-self::*[@style]")
 
 _find_external_links = etree.XPath(
     "descendant-or-self::a[normalize-space(@href) and substring(normalize-space(@href),1,1) != '#']")
 
-def clean_html(html, **kw):
-    """
-    Like clean(), but takes a text input document, and returns a text
-    document.
-    """
-    doc = fromstring(html)
-    clean(doc, **kw)
-    return tostring(doc)
-
 class Cleaner(object):
     """
     Instances cleans the document of each of the possible offending
     elements.  The cleaning is controlled by attributes; you can
     override attributes in a subclass, or set them in the constructor.
 
     ``scripts``:
         Removes any ``<script>`` tags.
 
     ``javascript``:
         Removes any Javascript, like an ``onclick`` attribute.
 
     ``comments``:
         Removes any comments.
 
@@ -193,138 +184,138 @@ class Cleaner(object):
         # FIXME: not looking at the action currently, because it is more complex
         # than than -- if you keep the form, you should keep the form controls.
         ##form='action',
         a='href',
         )
 
     def __call__(self, doc):
         """
         Cleans the document.
         """
         if hasattr(doc, 'getroot'):
             # ElementTree instance, instead of an element
             doc = doc.getroot()
         # Normalize a case that IE treats <image> like <img>, and that
         # can confuse either this step or later steps.
-        for el in doc.getiterator('image'):
+        for el in doc.iter('image'):
             el.tag = 'img'
         if not self.comments:
             # Of course, if we were going to kill comments anyway, we don't
             # need to worry about this
             self.kill_conditional_comments(doc)
         kill_tags = set()
         remove_tags = set(self.remove_tags or ())
         if self.allow_tags:
             allow_tags = set(self.allow_tags)
         else:
             allow_tags = set()
         if self.scripts:
             kill_tags.add('script')
         if self.safe_attrs_only:
             safe_attrs = set(defs.safe_attrs)
-            for el in doc.getiterator():
+            for el in doc.iter():
                 attrib = el.attrib
                 for aname in attrib.keys():
                     if aname not in safe_attrs:
                         del attrib[aname]
         if self.javascript:
             if not self.safe_attrs_only:
                 # safe_attrs handles events attributes itself
-                for el in doc.getiterator():
+                for el in doc.iter():
                     attrib = el.attrib
                     for aname in attrib.keys():
                         if aname.startswith('on'):
                             del attrib[aname]
             doc.rewrite_links(self._remove_javascript_link,
                               resolve_base_href=False)
             if not self.style:
                 # If we're deleting style then we don't have to remove JS links
                 # from styles, otherwise...
                 for el in _find_styled_elements(doc):
                     old = el.get('style')
                     new = _css_javascript_re.sub('', old)
                     new = _css_import_re.sub('', old)
                     if self._has_sneaky_javascript(new):
                         # Something tricky is going on...
                         del el.attrib['style']
                     elif new != old:
                         el.set('style', new)
-                for el in list(doc.getiterator('style')):
+                for el in list(doc.iter('style')):
                     if el.get('type', '').lower().strip() == 'text/javascript':
                         el.drop_tree()
                         continue
                     old = el.text or ''
                     new = _css_javascript_re.sub('', old)
                     # The imported CSS can do anything; we just can't allow:
                     new = _css_import_re.sub('', old)
                     if self._has_sneaky_javascript(new):
                         # Something tricky is going on...
                         el.text = '/* deleted */'
                     elif new != old:
                         el.text = new
         if self.comments or self.processing_instructions:
             # FIXME: why either?  I feel like there's some obscure reason
             # because you can put PIs in comments...?  But I've already
             # forgotten it
             kill_tags.add(etree.Comment)
         if self.processing_instructions:
             kill_tags.add(etree.ProcessingInstruction)
         if self.style:
             kill_tags.add('style')
             for el in _find_styled_elements(doc):
                 del el.attrib['style']
         if self.links:
             kill_tags.add('link')
         elif self.style or self.javascript:
             # We must get rid of included stylesheets if Javascript is not
             # allowed, as you can put Javascript in them
-            for el in list(doc.getiterator('link')):
+            for el in list(doc.iter('link')):
                 if 'stylesheet' in el.get('rel', '').lower():
                     # Note this kills alternate stylesheets as well
                     el.drop_tree()
         if self.meta:
             kill_tags.add('meta')
         if self.page_structure:
             remove_tags.update(('head', 'html', 'title'))
         if self.embedded:
             # FIXME: is <layer> really embedded?
             # We should get rid of any <param> tags not inside <applet>;
             # These are not really valid anyway.
-            for el in list(doc.getiterator('param')):
+            for el in list(doc.iter('param')):
                 found_parent = False
                 parent = el.getparent()
                 while parent is not None and parent.tag not in ('applet', 'object'):
                     parent = parent.getparent()
                 if parent is None:
                     el.drop_tree()
             kill_tags.update(('applet',))
             # The alternate contents that are in an iframe are a good fallback:
             remove_tags.update(('iframe', 'embed', 'layer', 'object', 'param'))
         if self.frames:
             # FIXME: ideally we should look at the frame links, but
             # generally frames don't mix properly with an HTML
             # fragment anyway.
             kill_tags.update(defs.frame_tags)
         if self.forms:
             remove_tags.add('form')
             kill_tags.update(('button', 'input', 'select', 'textarea'))
         if self.annoying_tags:
             remove_tags.update(('blink', 'marque'))
 
         _remove = []
         _kill = []
-        for el in doc.getiterator():
+        for el in doc.iter():
             if el.tag in kill_tags:
                 if self.allow_element(el):
                     continue
                 _kill.append(el)
             elif el.tag in remove_tags:
                 if self.allow_element(el):
                     continue
                 _remove.append(el)
 
         if _remove and _remove[0] == doc:
             # We have to drop the parent-most tag, which we can't
             # do.  Instead we'll rewrite it:
             el = _remove.pop(0)
             el.tag = 'div'
             el.attrib.clear()
@@ -337,31 +328,31 @@ class Cleaner(object):
             el.clear()
 
         for el in _kill:
             el.drop_tree()
         for el in _remove:
             el.drop_tag()
 
         allow_tags = self.allow_tags
         if self.remove_unknown_tags:
             if allow_tags:
                 raise ValueError(
                     "It does not make sense to pass in both allow_tags and remove_unknown_tags")
             allow_tags = set(defs.tags)
         if allow_tags:
             bad = []
-            for el in doc.getiterator():
+            for el in doc.iter():
                 if el.tag not in allow_tags:
                     bad.append(el)
             for el in bad:
                 el.drop_tag()
         if self.add_nofollow:
             for el in _find_external_links(doc):
                 if not self.allow_follow(el):
                     el.set('rel', 'nofollow')
 
     def allow_follow(self, anchor):
         """
         Override to suppress rel="nofollow" on some anchors.
         """
         return False
 
@@ -396,60 +387,60 @@ class Cleaner(object):
         return False
 
     def kill_conditional_comments(self, doc):
         """
         IE conditional comments basically embed HTML that the parser
         doesn't normally see.  We can't allow anything like that, so
         we'll kill any comments that could be conditional.
         """
         bad = []
         self._kill_elements(
             doc, lambda el: _conditional_comment_re.search(el.text),
             etree.Comment)                
 
     def _kill_elements(self, doc, condition, iterate=None):
         bad = []
-        for el in doc.getiterator(iterate):
+        for el in doc.iter(iterate):
             if condition(el):
                 bad.append(el)
         for el in bad:
             el.drop_tree()
 
     def _remove_javascript_link(self, link):
         # links like "j a v a s c r i p t:" might be interpreted in IE
-        new = _whitespace_re.sub('', link)
+        new = _substitute_whitespace('', link)
         if _javascript_scheme_re.search(new):
             # FIXME: should this be None to delete?
             return ''
         return link
 
-    _decomment_re = re.compile(r'/\*.*?\*/', re.S)
+    _substitute_comments = re.compile(r'/\*.*?\*/', re.S).sub
 
     def _has_sneaky_javascript(self, style):
         """
         Depending on the browser, stuff like ``e x p r e s s i o n(...)``
         can get interpreted, or ``expre/* stuff */ssion(...)``.  This
         checks for attempt to do stuff like this.
 
         Typically the response will be to kill the entire style; if you
         have just a bit of Javascript in the style another rule will catch
         that and remove only the Javascript from the style; this catches
         more sneaky attempts.
         """
-        style = self._decomment_re.sub('', style)
+        style = self._substitute_comments('', style)
         style = style.replace('\\', '')
-        style = _whitespace_re.sub('', style)
+        style = _substitute_whitespace('', style)
         style = style.lower()
         if 'javascript:' in style:
             return True
         if 'expression(' in style:
             return True
         return False
 
     def clean_html(self, html):
         if isinstance(html, basestring):
             return_string = True
             doc = fromstring(html)
         else:
             return_string = False
             doc = copy.deepcopy(html)
         self(doc)
