diff --git a/djblets/util/templatetags/djblets_js.py b/djblets/util/templatetags/djblets_js.py
index 152ee269..d281661c 100644
--- a/djblets/util/templatetags/djblets_js.py
+++ b/djblets/util/templatetags/djblets_js.py
@@ -13,30 +13,34 @@
 # the following conditions:
 #
 # The above copyright notice and this permission notice shall be included
 # in all copies or substantial portions of the Software.
 #
 # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 # EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 # MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 # IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 # CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 # TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 # SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 
 from django import template
+from django.core.serializers import serialize
+from django.db.models.query import QuerySet
+from django.utils import simplejson
+from django.utils.safestring import mark_safe
 
 
 register = template.Library()
 
 
 @register.simple_tag
 def form_dialog_fields(form):
     """
     Translates a Django Form object into a JavaScript list of fields.
     The resulting list of fields can be used to represent the form
     dynamically.
     """
     s = ''
 
     for field in form:
@@ -45,15 +49,26 @@ def form_dialog_fields(form):
         if field.is_hidden:
             s += "hidden: true, "
         else:
             s += "label: '%s', " % field.label_tag(field.label + ":")
 
             if field.field.required:
                 s += "required: true, "
 
             if field.field.help_text:
                 s += "help_text: '%s', " % field.field.help_text
 
         s += "widget: '%s' }," % unicode(field)
 
     # Chop off the last ','
     return "[ %s ]" % s[:-1]
+
+
+
+@register.filter
+def json_dumps(value, indent=0):
+    if isinstance(value, QuerySet):
+        result = serialize('json', value, indent=indent)
+    else:
+        result = simplejson.dumps(value, indent=indent)
+
+    return mark_safe(result)
