diff --git a/src/messagetypes/__init__.py b/src/messagetypes/__init__.py
index 0b63519d..16d97ae7 100644
--- a/src/messagetypes/__init__.py
+++ b/src/messagetypes/__init__.py
@@ -2,31 +2,31 @@ from importlib import import_module
 from os import path, listdir
 from string import lower
 
 from debug import logger
 
 
 class MsgBase(object):
     def encode(self):
         self.data = {"": lower(type(self).__name__)}
 
 
 def constructObject(data):
     try:
         classBase = eval(data[""] + "." + data[""].title())
     except NameError:
-        print "Don't know how to handle message type: \"%s\"" % (data[""])
+        logger.error("Don't know how to handle message type: \"%s\"", data[""])
         return None
     try:
         returnObj = classBase()
         returnObj.decode(data)
     except KeyError as e:
         logger.error("Missing mandatory key %s", e)
         return None
     except:
         logger.error("classBase fail", exc_info=True)
         return None
     else:
         return returnObj
 
 for mod in listdir(path.dirname(__file__)):
     if mod == "__init__.py":
