diff --git a/application/controllers/AdminController.php b/application/controllers/AdminController.php
index 7fe7b5aa9be..e24d552a405 100644
--- a/application/controllers/AdminController.php
+++ b/application/controllers/AdminController.php
@@ -34,7 +34,7 @@ protected function _init()
 		if (Yii::app()->getConfig('buildnumber') != "" && Yii::app()->getConfig('updatecheckperiod') > 0 && $updatelastcheck < date_shift(date("Y-m-d H:i:s"), "Y-m-d H:i:s", "-". Yii::app()->getConfig('updatecheckperiod')." days"))
 			updatecheck();
 
-		unset(Yii::app()->session['FileManagerContext']);
+		//unset(Yii::app()->session['FileManagerContext']);
 
 		$this->user_id = Yii::app()->user->getId();
 
@@ -184,6 +184,7 @@ public function getActionClasses()
 			'globalsettings'   => 'globalsettings',
 			'htmleditor_pop'   => 'htmleditor_pop',
 			'index'            => 'index',
+            'kcfinder'         => 'kcfinder',
 			'labels'           => 'labels',
 			'participants'     => 'participantsaction',
 			'printablesurvey'  => 'printablesurvey',
diff --git a/application/controllers/admin/assessments.php b/application/controllers/admin/assessments.php
index ec790301940..47d68d9ffd1 100644
--- a/application/controllers/admin/assessments.php
+++ b/application/controllers/admin/assessments.php
@@ -39,6 +39,8 @@ public function index($iSurveyId)
         $languages = Survey::model()->findByPk($iSurveyId)->additionalLanguages;
         $surveyLanguage = Survey::model()->findByPk($iSurveyId)->language;
 
+        $_SESSION['FileManagerContext'] = "edit:assessments:{$iSurveyId}";
+
         array_unshift($languages, $surveyLanguage); // makes an array with ALL the languages supported by the survey -> $assessmentlangs
 
         Yii::app()->setConfig("baselang", $surveyLanguage);
diff --git a/application/controllers/admin/emailtemplates.php b/application/controllers/admin/emailtemplates.php
index 360d924065b..dc6d6a05863 100644
--- a/application/controllers/admin/emailtemplates.php
+++ b/application/controllers/admin/emailtemplates.php
@@ -37,6 +37,8 @@ function index($iSurveyId)
         Yii::app()->loadHelper('admin.htmleditor');
         Yii::app()->loadHelper('surveytranslator');
 
+        $_SESSION['FileManagerContext'] = "edit:assessments:{$iSurveyId}";
+
         if(isset($iSurveyId) && getEmailFormat($iSurveyId) == 'html')
         {
             $ishtml = true;
diff --git a/application/controllers/admin/export.php b/application/controllers/admin/export.php
index 2f1036dd67b..bc4709fe7ae 100644
--- a/application/controllers/admin/export.php
+++ b/application/controllers/admin/export.php
@@ -1095,52 +1095,50 @@ public function showquexmlsurvey()
 
     public function resources()
     {
-        $id = sanitize_int(Yii::app()->request->getParam('id'));
-        $action = Yii::app()->request->getParam('action');
-
-        $this->load->library("admin/Phpzip");
-        $z = $this->phpzip;
-
-        if ( $action == "exportsurvresources" )
+        switch ( Yii::app()->request->getParam('export') )
 		{
-            $surveyid = $id;
-            $resourcesdir = $this->config->item("uploaddir") . "/surveys/$surveyid/";
-            $zipfile = $this->config->item("tempdir") . "/resources-survey-$surveyid.zip";
-            $z -> Zip($resourcesdir, $zipfile);
-
-            if ( is_file($zipfile) )
-			{
-                //Send the file for download!
-				$fn = "resources-survey-{$surveyid}.zip";
-				$this->_addHeaders($fn, "application/force-download", 0);
-
-                @readfile($zipfile);
-
-                //Delete the temporary file
-                unlink($zipfile);
-                return;
-            }
-        }
-
-        if ( $action == "exportlabelresources" )
-		{
-            $lid = $id;
-            $resourcesdir = $this->config->item("uploaddir") . "/labels/$lid/";
-            $zipfile = $this->config->item("tempdir") . "/resources-labelset-$lid.zip";
-            $z -> Zip($resourcesdir, $zipfile);
-
-            if ( is_file($zipfile) )
-			{
-                //Send the file for download!
-				$fn = "resources-label-{$lid}.zip";
-				$this->_addHeaders($fn, "application/force-download", 0);
-
-                @readfile($zipfile);
-
-                //Delete the temporary file
-                unlink($zipfile);
-                return;
-            }
+            case 'survey' :
+                $surveyid = sanitize_int(CHttpRequest::getParam('surveyid'));
+                $resourcesdir = Yii::app()->getConfig('uploaddir') . "/surveys/{$surveyid}/";
+                $tmpdir = Yii::app()->getConfig('tempdir') . '/';
+                $zipfilename = "resources-survey-$surveyid.zip";
+                $zipfilepath = $tmpdir . $zipfilename;
+
+                Yii::app()->loadLibrary('admin.pclzip.pclzip');
+                $zip = new PclZip($zipfilepath);
+                if ($zip->create($resourcesdir, PCLZIP_OPT_REMOVE_PATH, $resourcesdir) === 0)
+                {
+                    die("Error : ".$archive->errorInfo(true));
+                }
+                elseif (file_exists($zipfilepath))
+                {
+                    $this->_addHeaders($zipfilename, 'application/force-download', 0);
+                    readfile($zipfilepath);
+                    unlink($zipfilepath);
+                    exit;
+                }
+                break;
+            case 'label' :
+                $lid = sanitize_int(CHttpRequest::getParam('lid'));
+                $resourcesdir = Yii::app()->getConfig('uploaddir') . "/labels/{$lid}/";
+                $tmpdir = Yii::app()->getConfig('tempdir') . '/';
+                $zipfilename = "resources-labelset-$lid.zip";
+                $zipfilepath = $tmpdir . $zipfilename;
+
+                Yii::app()->loadLibrary('admin.pclzip.pclzip');
+                $zip = new PclZip($zipfilepath);
+                if ($zip->create($resourcesdir, PCLZIP_OPT_REMOVE_PATH, $resourcesdir) === 0)
+                {
+                    die("Error : ".$archive->errorInfo(true));
+                }
+                elseif (file_exists($zipfilepath))
+                {
+                    $this->_addHeaders($zipfilename, 'application/force-download', 0);
+                    readfile($zipfilepath);
+                    unlink($zipfilepath);
+                    exit;
+                }
+                break;
         }
     }
 
diff --git a/application/controllers/admin/htmleditor_pop.php b/application/controllers/admin/htmleditor_pop.php
index 53cfb17a19b..7c3e243e916 100644
--- a/application/controllers/admin/htmleditor_pop.php
+++ b/application/controllers/admin/htmleditor_pop.php
@@ -19,14 +19,14 @@ class htmleditor_pop extends Survey_Common_Action
 
     function index()
     {
-        $sFieldName = isset($_GET['name'])?$_GET['name']:0;
-        $sFieldText = isset($_GET['text'])?$_GET['text']:0;
-        $sFieldType = isset($_GET['type'])?$_GET['type']:0;
-        $sAction = isset($_GET['action'])?$_GET['action']:0;
-        $iSurveyId = isset($_GET['sid'])?$_GET['sid']:0;
-        $iGroupId = isset($_GET['gid'])?$_GET['gid']:0;
-        $iQuestionId = isset($_GET['qid'])?$_GET['qid']:0;
-        $sLanguage = isset($_GET['lang'])?$_GET['lang']:0;
+        $sFieldName = isset($_GET['name']) ? $_GET['name'] : 0;
+        $sFieldText = isset($_GET['text']) ? $_GET['text'] : 0;
+        $sFieldType = isset($_GET['type']) ? $_GET['type'] : 0;
+        $sAction = isset($_GET['action']) ? $_GET['action'] : 0;
+        $iSurveyId = isset($_GET['sid']) ? $_GET['sid'] : 0;
+        $iGroupId = isset($_GET['gid']) ? $_GET['gid'] : 0;
+        $iQuestionId = isset($_GET['qid']) ? $_GET['qid'] : 0;
+        $sLanguage = isset($_GET['lang']) ? $_GET['lang'] : 0;
         $aData['clang'] = $this->getController()->lang;
         $aData['sFieldName'] = $sFieldName;
         if (get_magic_quotes_gpc())
diff --git a/application/controllers/admin/kcfinder.php b/application/controllers/admin/kcfinder.php
new file mode 100644
index 00000000000..c9af5fa8b58
--- /dev/null
+++ b/application/controllers/admin/kcfinder.php
@@ -0,0 +1,361 @@
+<?php if (!defined('BASEPATH')) exit('No direct script access allowed');
+/*
+ * LimeSurvey
+ * Copyright (C) 2007-2011 The LimeSurvey Project Team / Carsten Schmitz
+ * All rights reserved.
+ * License: GNU/GPL License v2 or later, see LICENSE.php
+ * LimeSurvey is free software. This version may have been modified pursuant
+ * to the GNU General Public License, and as distributed it includes or
+ * is derivative of works licensed under the GNU General Public License or
+ * other free or open source software licenses.
+ * See COPYRIGHT.php for copyright notices and details.
+ *
+ * 	$Id: Admin_Controller.php 11256 2011-10-25 13:52:18Z c_schmitz $
+ */
+
+class kcfinder extends Survey_Common_Action
+{
+
+    function index($load = false)
+    {
+        $usquery = "SELECT stg_value FROM {{settings_global}} where stg_name='SessionName'";
+        $usresult = Yii::app()->db->createCommand($usquery)->queryRow();
+        if ($usresult)
+        {
+            @session_name($usrow['stg_value']);
+        }
+        else
+        {
+            session_name("LimeSurveyAdmin");
+        }
+
+        session_set_cookie_params(0, Yii::app()->createUrl('admin/kcfinder/index/load/' . (string) $load . '/'));
+
+        if (session_id() == "")
+            @session_start();
+
+        $_SESSION['KCFINDER'] = array();
+
+        $sAllowedExtensions = implode(' ', array_map('trim', explode(',', Yii::app()->getConfig('allowedresourcesuploads'))));
+        $_SESSION['KCFINDER']['types'] = array(
+            'files' => $sAllowedExtensions,
+            'flash' => $sAllowedExtensions,
+            'images' => $sAllowedExtensions
+        );
+
+        if (Yii::app()->getConfig('demoMode') === false &&
+                isset($_SESSION['loginID']) &&
+                isset($_SESSION['FileManagerContext']))
+        {
+            // disable upload at survey creation time
+            // because we don't know the sid yet
+            if (preg_match('/^(create|edit):(question|group|answer)/', $_SESSION['FileManagerContext']) != 0 ||
+                    preg_match('/^edit:survey/', $_SESSION['FileManagerContext']) != 0 ||
+                    preg_match('/^edit:assessments/', $_SESSION['FileManagerContext']) != 0 ||
+                    preg_match('/^edit:emailsettings/', $_SESSION['FileManagerContext']) != 0)
+            {
+                $contextarray = explode(':', $_SESSION['FileManagerContext'], 3);
+                $surveyid = $contextarray[2];
+
+                if (bHasSurveyPermission($surveyid, 'surveycontent', 'update'))
+                {
+                    $_SESSION['KCFINDER']['disabled'] = false;
+                    $_SESSION['KCFINDER']['uploadURL'] = $this->getController()->createUrl("upload/surveys/{$surveyid}/");
+                    $_SESSION['KCFINDER']['uploadDir'] = ROOT . "/upload/surveys/{$surveyid}/";
+                }
+            }
+            elseif (preg_match('/^edit:label/', $_SESSION['FileManagerContext']) != 0)
+            {
+                $contextarray = explode(':', $_SESSION['FileManagerContext'], 3);
+                $labelid = $contextarray[2];
+                // check if the user has label management right and labelid defined
+                if ($_SESSION['USER_RIGHT_MANAGE_LABEL'] == 1 && isset($labelid) && $labelid != '')
+                {
+                    $_SESSION['KCFINDER']['disabled'] = false;
+                    $_SESSION['KCFINDER']['uploadURL'] = $this->getController()->createUrl("upload/labels/{$labelid}/");
+                    $_SESSION['KCFINDER']['uploadDir'] = ROOT . "/upload/labels/{$labelid}/";
+                }
+            }
+        }
+
+        Yii::registerAutoloader(array($this, 'kcfinder_autoload'), true);
+        if (!empty($load) && file_exists(ROOT . "/scripts/admin/kcfinder/" . $load . EXT))
+        {
+            require_once(ROOT . "/scripts/admin/kcfinder/" . $load . EXT);
+        }
+    }
+
+    function kcfinder_autoload($class)
+    {
+        if ($class == "uploader")
+            require ROOT . "/scripts/admin/kcfinder/core/uploader.php";
+        elseif ($class == "browser")
+            require ROOT . "/scripts/admin/kcfinder/core/browser.php";
+        elseif (file_exists(ROOT . "/scripts/admin/kcfinder/core/types/$class.php"))
+            require ROOT . "/scripts/admin/kcfinder/core/types/$class.php";
+        elseif (file_exists(ROOT . "/scripts/admin/kcfinder/lib/class_$class.php"))
+            require ROOT . "/scripts/admin/kcfinder/lib/class_$class.php";
+        elseif (file_exists(ROOT . "/scripts/admin/kcfinder/lib/helper_$class.php"))
+            require ROOT . "/scripts/admin/kcfinder/lib/helper_$class.php";
+    }
+
+    function css()
+    {
+        $this->index();
+
+        $mtime = @filemtime(__FILE__);
+        if ($mtime)
+            httpCache::checkMTime($mtime);
+        $browser = new browser();
+        $config = $browser->config;
+        ob_start();
+        ?>
+        html, body {
+        overflow: hidden;
+        }
+
+        body, form, th, td {
+        margin: 0;
+        padding: 0;
+        }
+
+        a {
+        cursor:pointer;
+        }
+
+        * {
+        font-family: Tahoma, Verdana, Arial, sans-serif;
+        font-size: 11px;
+        }
+
+        table {
+        border-collapse: collapse;
+        }
+
+        #all {
+        visibility: hidden;
+        }
+
+        #left {
+        float: left;
+        display: block;
+        width: 25%;
+        }
+
+        #right {
+        float: left;
+        display: block;
+        width: 75%;
+        }
+
+        #settings {
+        display: none;
+        padding: 0;
+        float: left;
+        width: 100%;
+        }
+
+        #settings > div {
+        float: left;
+        }
+
+        #folders {
+        padding: 5px;
+        overflow: auto;
+        }
+
+        #toolbar {
+        padding: 5px;
+        }
+
+        #files {
+        padding: 5px;
+        overflow: auto;
+        }
+
+        #status {
+        padding: 5px;
+        float: left;
+        overflow: hidden;
+        }
+
+        #fileinfo {
+        float: left;
+        }
+
+        #clipboard div {
+        width: 16px;
+        height: 16px;
+        }
+
+        .folders {
+        margin-left: 16px;
+        }
+
+        div.file {
+        overflow-x: hidden;
+        width: <?php echo $config['thumbWidth'] ?>px;
+        float: left;
+        text-align: center;
+        cursor: default;
+        white-space: nowrap;
+        }
+
+        div.file .thumb {
+        width: <?php echo $config['thumbWidth'] ?>px;
+        height: <?php echo $config['thumbHeight'] ?>px;
+        background: no-repeat center center;
+        }
+
+        #files table {
+        width: 100%;
+        }
+
+        tr.file {
+        cursor: default;
+        }
+
+        tr.file > td {
+        white-space: nowrap;
+        }
+
+        tr.file > td.name {
+        background-repeat: no-repeat;
+        background-position: left center;
+        padding-left: 20px;
+        width: 100%;
+        }
+
+        tr.file > td.time,
+        tr.file > td.size {
+        text-align: right;
+        }
+
+        #toolbar {
+        cursor: default;
+        white-space: nowrap;
+        }
+
+        #toolbar a {
+        padding-left: 20px;
+        text-decoration: none;
+        background: no-repeat left center;
+        }
+
+        #toolbar a:hover, a[href="#upload"].uploadHover {
+        color: #000;
+        }
+
+        #upload {
+        position: absolute;
+        overflow: hidden;
+        opacity: 0;
+        filter: alpha(opacity:0);
+        }
+
+        #upload input {
+        cursor: pointer;
+        }
+
+        #uploadResponse {
+        display: none;
+        }
+
+        span.brace {
+        padding-left: 11px;
+        cursor: default;
+        }
+
+        span.brace.opened, span.brace.closed {
+        cursor: pointer;
+        }
+
+        #shadow {
+        position: absolute;
+        top: 0;
+        left: 0;
+        display: none;
+        background: #fff;
+        z-index: 100;
+        opacity: 0.5;
+        filter: alpha(opacity:50);
+        }
+
+        #dialog, #clipboard {
+        position: absolute;
+        display: none;
+        z-index: 101;
+        cursor: default;
+        }
+
+        #clipboard {
+        z-index: 99;
+        }
+
+        #loading {
+        display: none;
+        float: right;
+        }
+
+        .menu {
+        background: #888;
+        white-space: nowrap;
+        }
+
+        .menu a {
+        display: block;
+        }
+
+        .menu .list {
+        max-height: 0;
+        overflow-y: auto;
+        overflow-x: hidden;
+        white-space: nowrap;
+        }
+
+        .file .access, .file .hasThumb {
+        display: none;
+        }
+
+        #dialog img {
+        cursor: pointer;
+        }
+
+        #resizer {
+        position: absolute;
+        z-index: 98;
+        top: 0;
+        background: #000;
+        opacity: 0;
+        filter: alpha(opacity:0);
+        }
+        <?php
+        header("Content-Type: text/css");
+        echo text::compressCSS(ob_get_clean());
+    }
+
+    function js_localise()
+    {
+        $this->index();
+
+        if (function_exists('set_magic_quotes_runtime'))
+            @set_magic_quotes_runtime(false);
+
+        $input = new input();
+        if (!isset($input->get['lng']) || ($input->get['lng'] == 'en'))
+            die;
+        $file = ROOT . "/scripts/admin/kcfinder/lang/" . $input->get['lng'] . ".php";
+        $files = glob(ROOT . "/scripts/admin/kcfinder/lang/*.php");
+        if (!in_array($file, $files))
+            die;
+        $mtime = @filemtime($file);
+        if ($mtime)
+            httpCache::checkMTime($mtime);
+        require $file;
+        header("Content-Type: text/javascript; charset={$lang['_charset']}");
+        foreach ($lang as $english => $native)
+            if (substr($english, 0, 1) != "_")
+                echo "browser.labels['" . text::jsValue($english) . "']=\"" . text::jsValue($native) . "\";";
+    }
+
+}
\ No newline at end of file
diff --git a/application/controllers/admin/labels.php b/application/controllers/admin/labels.php
index 5abe7e8e7a5..0de7819e7a9 100644
--- a/application/controllers/admin/labels.php
+++ b/application/controllers/admin/labels.php
@@ -49,28 +49,26 @@ public function run($sa)
     public function importlabelresources()
     {
         $clang = $this->getController()->lang;
-        $action = returnglobal('action');
         $lid = returnglobal('lid');
 
-        if ($action == "importlabelresources" && $lid)
+        if (!empty($lid))
         {
             if (Yii::app()->getConfig('demoMode'))
-                $this->getController()->error($clang->gT("Demo mode only: Uploading files is disabled in this system."));
-
-            Yii::import('application.libraries.admin.Phpzip');
-
-            $zipfile = $_FILES['the_file']['tmp_name'];
-            $z = new Phpzip();
+                $this->getController()->error($clang->gT("Demo mode only: Uploading files is disabled in this system."), $this->getController()->createUrl("admin/labels/view/lid/{$lid}"));
 
             // Create temporary directory
             // If dangerous content is unzipped
             // then no one will know the path
             $extractdir = self::_tempdir(Yii::app()->getConfig('tempdir'));
-            $basedestdir = Yii::app()->getConfig('publicdir') . "/upload/labels";
+            $zipfilename = $_FILES['the_file']['tmp_name'];
+            $basedestdir = Yii::app()->getConfig('uploaddir') . "/labels";
             $destdir = $basedestdir . "/$lid/";
 
+            Yii::app()->loadLibrary('admin.pclzip.pclzip');
+            $zip = new PclZip($zipfilename);
+
             if (!is_writeable($basedestdir))
-                $this->getController()->error(sprintf($clang->gT("Incorrect permissions in your %s folder."), $basedestdir));
+                $this->getController()->error(sprintf($clang->gT("Incorrect permissions in your %s folder."), $basedestdir), $this->getController()->createUrl("admin/labels/view/lid/{$lid}"));
 
             if (!is_dir($destdir))
                 mkdir($destdir);
@@ -78,68 +76,22 @@ public function importlabelresources()
             $aImportedFilesInfo = null;
             $aErrorFilesInfo = null;
 
-            if (is_file($zipfile))
+            if (is_file($zipfilename))
             {
-
-                if ($z->extract($extractdir, $zipfile) != 'OK')
-                    $this->getController()->error($clang->gT("This file is not a valid ZIP file archive. Import failed."));
+                if ($zip->extract($extractdir) <= 0)
+                    $this->getController()->error($clang->gT("This file is not a valid ZIP file archive. Import failed. " . $zip->errorInfo(true)), $this->getController()->createUrl("admin/labels/view/lid/{$lid}"));
 
                 // now read tempdir and copy authorized files only
-                $dh = opendir($extractdir);
-                $aErrorFilesInfo = array();
-                $aImportedFilesInfo = array();
-                while ($direntry = readdir($dh))
-                {
-                    if (($direntry != ".") && ($direntry != ".."))
-                    {
-                        if (is_file($extractdir . "/" . $direntry))
-                        {
-                            // is  a file
-                            $extfile = substr(strrchr($direntry, '.'), 1);
-                            if (!(stripos(',' . Yii::app()->getConfig('allowedresourcesuploads') . ',', ',' . $extfile . ',') === false))
-                            {
-                                // Extension allowed
-                                if (!copy($extractdir . "/" . $direntry, $destdir . $direntry))
-                                {
-                                    $aErrorFilesInfo[] = Array(
-                                        "filename" => $direntry,
-                                        "status" => $clang->gT("Copy failed")
-                                    );
-                                    unlink($extractdir . "/" . $direntry);
-                                }
-                                else
-                                {
-                                    $aImportedFilesInfo[] = Array(
-                                        "filename" => $direntry,
-                                        "status" => $clang->gT("OK")
-                                    );
-                                    unlink($extractdir . "/" . $direntry);
-                                }
-                            }
-                            else
-                            {
-                                // Extension forbidden
-                                $aErrorFilesInfo[] = Array(
-                                    "filename" => $direntry,
-                                    "status" => $clang->gT("Error") . " (" . $clang->gT("Forbidden Extension") . ")"
-                                );
-                                unlink($extractdir . "/" . $direntry);
-                            }
-                        }
-                    }
-                }
+                list($aImportedFilesInfo, $aErrorFilesInfo) = $this->_filterImportedResources($extractdir, $destdir);
 
                 // Delete the temporary file
-                unlink($zipfile);
-
-                // Delete temporary folder
-                rmdir($extractdir);
+                unlink($zipfilename);
 
                 if (is_null($aErrorFilesInfo) && is_null($aImportedFilesInfo))
-                    $this->getController()->error($clang->gT("This ZIP archive contains no valid Resources files. Import failed.") . '<br /><br />' . $clang->gT("Remember that we do not support subdirectories in ZIP archives."));
+                    $this->getController()->error($clang->gT("This ZIP archive contains no valid Resources files. Import failed."), $this->getController()->createUrl("admin/labels/view/lid/{$lid}"));
             }
             else
-                $this->getController()->error(sprintf($clang->gT("An error occurred uploading your file. This may be caused by incorrect permissions in your %s folder."), $basedestdir));
+                $this->getController()->error(sprintf($clang->gT("An error occurred uploading your file. This may be caused by incorrect permissions in your %s folder."), $basedestdir), $this->getController()->createUrl("admin/labels/view/lid/{$lid}"));
 
             $aData = array(
                 'aErrorFilesInfo' => $aErrorFilesInfo,
@@ -213,7 +165,7 @@ public function import()
             $aViewUrls['import_view'][] = array('aImportResults' => $aImportResults);
         }
 
-        $this->_renderWrappedTemplate($aViewUrls, $aData);
+        $this->_renderWrappedTemplate($aViewUrls);
     }
 
     /**
@@ -291,6 +243,8 @@ public function view($lid = 0)
         if ($lid != false)
             $lid = sanitize_int($lid);
 
+        $_SESSION['FileManagerContext'] = "edit:label:{$lid}";
+
         // Gets the current language
         $clang = $this->getController()->lang;
         $action = 'labels';
diff --git a/application/controllers/admin/question.php b/application/controllers/admin/question.php
index 5d364f4396b..6adc1925638 100644
--- a/application/controllers/admin/question.php
+++ b/application/controllers/admin/question.php
@@ -288,7 +288,7 @@ public function _editansweroptions($surveyid, $gid, $qid)
         $qtype = $qrow['type'];
 
         $qtypes = getqtypelist('', 'array');
-        
+
         $scalecount = $qtypes[$qtype]['answerscales'];
 
         $clang = $this->getController()->lang;
@@ -480,7 +480,7 @@ public function _editsubquestion($surveyid, $gid, $qid)
 
             // Check that there are subquestions for every language supported by the survey
             foreach ($anslangs as $language)
-            {                
+            {
                 foreach ($subquestiondata as $row)
                 {
                     $qrow = Questions::model()->count('
@@ -621,9 +621,13 @@ public function index($sa, $surveyid, $gid, $qid=null)
         $aData['qid'] = $qid;
         $aData['display']['menu_bars']['surveysummary'] = 'viewgroup';
         $aData['display']['menu_bars']['gid_action'] = 'addquestion';
+        $_SESSION['FileManagerContext'] = "create:question:{$surveyid}";
 
         if ($action != "addquestion")
+        {
+            $_SESSION['FileManagerContext'] = "edit:question:{$surveyid}";
             $aData['display']['menu_bars']['qid_action'] = 'editquestion';
+        }
 
         if (bHasSurveyPermission($surveyid, 'surveycontent', 'read'))
         {
diff --git a/application/controllers/admin/questiongroup.php b/application/controllers/admin/questiongroup.php
index 22050326df7..966e5d5fdb6 100644
--- a/application/controllers/admin/questiongroup.php
+++ b/application/controllers/admin/questiongroup.php
@@ -110,6 +110,8 @@ function add($surveyid)
             $action = "addgroup";
             $clang = $this->getController()->lang;
 
+            $_SESSION['FileManagerContext'] = "create:group:{$surveyid}";
+
             if ($action == "addgroup")
             {
                 Yii::app()->loadHelper('admin/htmleditor');
@@ -267,9 +269,11 @@ public function edit($surveyid, $gid)
 
         if (bHasSurveyPermission($surveyid, 'surveycontent', 'read'))
         {
-            $action = "editgroup"; //$this->input->post('action');
+            $action = "editgroup";
             $clang = $this->getController()->lang;
 
+            $_SESSION['FileManagerContext'] = "edit:group:{$surveyid}";
+
             if ($action == "editgroup")
             {
                 Yii::app()->loadHelper('admin/htmleditor');
diff --git a/application/controllers/admin/surveyaction.php b/application/controllers/admin/surveyaction.php
index 7ff06a143d6..b7107188b73 100644
--- a/application/controllers/admin/surveyaction.php
+++ b/application/controllers/admin/surveyaction.php
@@ -133,9 +133,10 @@ function editsurveysettings($surveyid)
         Yii::app()->loadHelper('surveytranslator');
         $clang = $this->getController()->lang;
 
+        $_SESSION['FileManagerContext'] = "edit:survey:{$surveyid}";
+
         $esrow = array();
         $esrow = self::_fetchSurveyInfo('editsurvey', $surveyid);
-        //$esrow = $esrow[0]->tableSchema->columns;
         $aData['esrow'] = $esrow;
 
         $aData = array_merge($aData, $this->_generalTabEditSurvey($surveyid, $esrow));
@@ -146,8 +147,6 @@ function editsurveysettings($surveyid)
         $aData = array_merge($aData, $this->_tabPanelIntegration($esrow));
         $aData = array_merge($aData, $this->_tabResourceManagement($surveyid));
 
-        //echo $editsurvey;
-        //$this->load->model('questions_model');
         $oResult = Questions::model()->getQuestionsWithSubQuestions($surveyid, $esrow['language'], "({{questions}}.type = 'T'  OR  {{questions}}.type = 'Q'  OR  {{questions}}.type = 'T' OR {{questions}}.type = 'S')");
 
         $aData['questions'] = $oResult;
@@ -158,43 +157,36 @@ function editsurveysettings($surveyid)
     }
 
     /**
-     * survey::importsurveyresources()
-     * Function responsible to import survey resources.
+     * Function responsible to import survey resources from a '.zip' file.
+     *
      * @access public
      * @return void
      */
     function importsurveyresources()
     {
         $clang = $this->getController()->lang;
-        $action = Yii::app()->request->getPost('action');
-        $surveyid = Yii::app()->request->getPost('sid');
-        $aViewUrls = array();
-
-        $aData = array(
-            'surveyid' => $surveyid,
-            'clang' => $clang
-        );
-
-        $aData['display']['menu_bars']['surveysummary'] = $action;
+        $surveyid = Yii::app()->request->getPost('surveyid');
 
-        if ($action == "importsurveyresources" && $surveyid)
+        if (!empty($surveyid))
         {
-            if (Yii::app()->getConfig('demoMode'))
-                $this->getController()->error($clang->gT("Demo Mode Only: Uploading file is disabled in this system."));
-
-            Yii::app()->loadLibrary('admin/Phpzip');
+            $aData['display']['menu_bars']['surveysummary'] = 'importsurveyresources';
 
-            $zipfile = $_FILES['the_file']['tmp_name'];
-            $z = new PHPZip();
+            if (Yii::app()->getConfig('demoMode'))
+                $this->getController()->error($clang->gT("Demo mode only: Uploading files is disabled in this system."), $this->getController()->createUrl("admin/survey/view/surveyid/{$surveyid}"));
 
             // Create temporary directory
-            // If dangerous content is unzipped then no one will know the path
-            $extractdir = $this->_tempdir(Yii::app()->getConfig('tempdir'));
-            $basedestdir = Yii::app()->getConfig('publicdir') . "/upload/surveys";
+            // If dangerous content is unzipped
+            // then no one will know the path
+            $extractdir = self::_tempdir(Yii::app()->getConfig('tempdir'));
+            $zipfilename = $_FILES['the_file']['tmp_name'];
+            $basedestdir = Yii::app()->getConfig('uploaddir') . "/surveys";
             $destdir = $basedestdir . "/$surveyid/";
 
+            Yii::app()->loadLibrary('admin.pclzip.pclzip');
+            $zip = new PclZip($zipfilename);
+
             if (!is_writeable($basedestdir))
-                $this->getController()->error(sprintf($clang->gT("Incorrect permissions in your %s folder."), $basedestdir));
+                $this->getController()->error(sprintf($clang->gT("Incorrect permissions in your %s folder."), $basedestdir), $this->getController()->createUrl("admin/survey/view/surveyid/{$surveyid}"));
 
             if (!is_dir($destdir))
                 mkdir($destdir);
@@ -202,75 +194,31 @@ function importsurveyresources()
             $aImportedFilesInfo = null;
             $aErrorFilesInfo = null;
 
-            if (is_file($zipfile))
+            if (is_file($zipfilename))
             {
-                if ($z->extract($extractdir, $zipfile) != 'OK')
-                    $this->getController()->error($clang->gT("This file is not a valid ZIP file $CI. Import failed."));
+                if ($zip->extract($extractdir) <= 0)
+                    $this->getController()->error($clang->gT("This file is not a valid ZIP file archive. Import failed. " . $zip->errorInfo(true)), $this->getController()->createUrl("admin/survey/view/surveyid/{$surveyid}"));
 
                 // now read tempdir and copy authorized files only
-                $dh = opendir($extractdir);
-                while ($direntry = readdir($dh))
-                {
-                    if (($direntry != ".") && ($direntry != ".."))
-                    {
-                        if (is_file($extractdir . "/" . $direntry))
-                        {
-                            // is  a file
-                            $extfile = substr(strrchr($direntry, '.'), 1);
-                            if (!(stripos(',' . $allowedresourcesuploads . ',', ',' . $extfile . ',') === false))
-                            {
-                                // Extension allowed
-                                if (!copy($extractdir . "/" . $direntry, $destdir . $direntry))
-                                {
-                                    $aErrorFilesInfo[] = Array(
-                                        "filename" => $direntry,
-                                        "status" => $clang->gT("Copy failed")
-                                    );
-                                    unlink($extractdir . "/" . $direntry);
-                                }
-                                else
-                                {
-                                    $aImportedFilesInfo[] = Array(
-                                        "filename" => $direntry,
-                                        "status" => $clang->gT("OK")
-                                    );
-                                    unlink($extractdir . "/" . $direntry);
-                                }
-                            }
-                            else
-                            { // Extension forbidden
-                                $aErrorFilesInfo[] = Array(
-                                    "filename" => $direntry,
-                                    "status" => $clang->gT("Error") . " (" . $clang->gT("Forbidden Extension") . ")"
-                                );
-                                unlink($extractdir . "/" . $direntry);
-                            }
-                        }
-                    }
-                }
+                list($aImportedFilesInfo, $aErrorFilesInfo) = $this->_filterImportedResources($extractdir, $destdir);
 
                 // Delete the temporary file
-                unlink($zipfile);
-                // Delete temporary folder
-                rmdir($extractdir);
+                unlink($zipfilename);
 
-                // display summary
                 if (is_null($aErrorFilesInfo) && is_null($aImportedFilesInfo))
-                    $this->getController()->error(
-                         $clang->gT("This ZIP $CI contains no valid Resources files. Import failed.") . '<br />' .
-                         $clang->gT("Remember that we do not support subdirectories in ZIP $CIs.")
-                    );
-
-                $aData['aErrorFilesInfo'] = $aErrorFilesInfo;
-                $aData['aImportedFilesInfo'] = $aImportedFilesInfo;
+                    $this->getController()->error($clang->gT("This ZIP archive contains no valid Resources files. Import failed."), $this->getController()->createUrl("admin/survey/view/surveyid/{$surveyid}"));
             }
             else
-                $this->getController()->error(sprintf($clang->gT("An error occurred uploading your file. This may be caused by incorrect permissions in your %s folder."), $basedestdir));
+                $this->getController()->error(sprintf($clang->gT("An error occurred uploading your file. This may be caused by incorrect permissions in your %s folder."), $basedestdir), $this->getController()->createUrl("admin/survey/view/surveyid/{$surveyid}"));
 
-            $aViewUrls[] = 'importSurveyResources_view';
-        }
+            $aData = array(
+                'aErrorFilesInfo' => $aErrorFilesInfo,
+                'aImportedFilesInfo' => $aImportedFilesInfo,
+                'surveyid' => $surveyid
+            );
 
-        $this->_renderWrappedTemplate($aViewUrls, $aData);
+            $this->_renderWrappedTemplate('importSurveyResources_view', $aData);
+        }
     }
 
     /**
diff --git a/application/core/Survey_Common_Action.php b/application/core/Survey_Common_Action.php
index 32f101f2957..1b9df8111c7 100644
--- a/application/core/Survey_Common_Action.php
+++ b/application/core/Survey_Common_Action.php
@@ -917,4 +917,64 @@ function _userGroupBar($ugid = 0)
         $this->getController()->render('/admin/usergroup/usergroupbar_view', $data);
     }
 
+    protected function _filterImportedResources($extractdir, $destdir)
+    {
+        $clang = $this->getController()->lang;
+        $dh = opendir($extractdir);
+        $aErrorFilesInfo = array();
+        $aImportedFilesInfo = array();
+
+        if (!is_dir($destdir))
+            mkdir($destdir);
+
+        while ($direntry = readdir($dh))
+        {
+            if ($direntry != "." && $direntry != "..")
+            {
+                if (is_file($extractdir . "/" . $direntry))
+                {
+                    // is  a file
+                    $extfile = substr(strrchr($direntry, '.'), 1);
+                    if (!(stripos(',' . Yii::app()->getConfig('allowedresourcesuploads') . ',', ',' . $extfile . ',') === false))
+                    {
+                        // Extension allowed
+                        if (!copy($extractdir . "/" . $direntry, $destdir . "/" . $direntry))
+                        {
+                            $aErrorFilesInfo[] = Array(
+                                "filename" => $direntry,
+                                "status" => $clang->gT("Copy failed")
+                            );
+                        }
+                        else
+                        {
+                            $aImportedFilesInfo[] = Array(
+                                "filename" => $direntry,
+                                "status" => $clang->gT("OK")
+                            );
+                        }
+                    }
+                    else
+                    {
+                        // Extension forbidden
+                        $aErrorFilesInfo[] = Array(
+                            "filename" => $direntry,
+                            "status" => $clang->gT("Forbidden Extension")
+                        );
+                    }
+                    unlink($extractdir . "/" . $direntry);
+                }
+                elseif (is_dir($extractdir . "/" . $direntry))
+                {
+                    list($_aImportedFilesInfo, $_aErrorFilesInfo) = $this->_filterImportedFiles($extractdir . "/" . $direntry, $destdir . "/" . $direntry);
+                    $aImportedFilesInfo = array_merge($aImportedFilesInfo, $_aImportedFilesInfo);
+                    $aErrorFilesInfo = array_merge($aErrorFilesInfo, $_aErrorFilesInfo);
+                }
+            }
+        }
+
+        rmdir($extractdir);
+
+        return array($aImportedFilesInfo, $aErrorFilesInfo);
+    }
+
 }
diff --git a/application/helpers/admin/label_helper.php b/application/helpers/admin/label_helper.php
index 8fe47e9c045..4d086afdacb 100644
--- a/application/helpers/admin/label_helper.php
+++ b/application/helpers/admin/label_helper.php
@@ -137,7 +137,7 @@ function modlabelsetanswers($lid)
         $_POST['method'] = $clang->gT("Save");
     }
 
-    $data = json_decode(stripslashes($_POST['dataToSend']));
+    $data = json_decode(CHttpRequest::getPost('dataToSend'));
 
     if ($ajax)
         $lid = insertlabelset();
@@ -145,7 +145,7 @@ function modlabelsetanswers($lid)
     if (count(array_unique($data->{'codelist'})) == count($data->{'codelist'}))
     {
 
-        $query = "DELETE FROM {{labels}}  WHERE `lid` = '$lid'";
+        $query = "DELETE FROM {{labels}} WHERE `lid` = '$lid'";
 
         $result = Yii::app()->db->createCommand($query)->execute();
 
diff --git a/application/views/admin/labels/import_view.php b/application/views/admin/labels/import_view.php
index f4ccd619dfa..85278066aad 100644
--- a/application/views/admin/labels/import_view.php
+++ b/application/views/admin/labels/import_view.php
@@ -30,5 +30,5 @@
         <li><?php echo $clang->gT("Labels") . ": {$aImportResults['labels']}" ?></li>
     </ul>
     <strong><?php $clang->eT("Import of label set(s) is completed.") ?></strong><br /><br />
-    <input type='submit' value='<?php $clang->gT("Return to label set administration") ?>' onclick="window.open('<?php echo $this->createUrl('admin/labels/view') ?>', '_top')" />
+    <input type='submit' value='<?php $clang->eT("Return to label set administration"); ?>' onclick="window.open('<?php echo $this->createUrl('admin/labels/view') ?>', '_top')" />
 </div><br />
\ No newline at end of file
diff --git a/application/views/admin/labels/importlabelresources_view.php b/application/views/admin/labels/importlabelresources_view.php
index d6066d9b324..9ba0bd6542c 100644
--- a/application/views/admin/labels/importlabelresources_view.php
+++ b/application/views/admin/labels/importlabelresources_view.php
@@ -1,4 +1,4 @@
-<div class='header ui-widget-header'><?php $clang->eT("Import Label Set") ?></div>
+<div class='header ui-widget-header'><?php $clang->eT("Import Label Set Resources") ?></div>
 <div class='messagebox ui-corner-all'>
     <div class="successheader"><?php $clang->eT("Success") ?></div><br />
     <?php $clang->eT("File upload succeeded.") ?><br /><br />
diff --git a/application/views/admin/labels/labelview_view.php b/application/views/admin/labels/labelview_view.php
index 9c0c0f90ffa..2609c42978b 100644
--- a/application/views/admin/labels/labelview_view.php
+++ b/application/views/admin/labels/labelview_view.php
@@ -124,13 +124,13 @@
                         <li>
                             <label>&nbsp;</label>
                             <input type='button'<?php echo hasResources($lid, 'label') === false ? ' disabled="disabled"' : '' ?>
-                                   onclick='window.open("<?php echo Yii::app()->getConfig('sCKEditorURL') ?>/editor/filemanager/browser/default/browser.html?Connector=../../connectors/php/connector.php?", "_blank")'
+                                   onclick='window.open("<?php echo $this->createUrl("admin/kcfinder/index/load/browse"); ?>", "_blank")'
                                    value="<?php $clang->eT("Browse Uploaded Resources") ?>" />
                         </li>
                         <li>
                             <label>&nbsp;</label>
                             <input type='button'<?php echo hasResources($lid, 'label') === false ? ' disabled="disabled"' : '' ?>
-                                   onclick='window.open("<?php echo $this->createUrl("/admin/labels/exportlabelresources/lid/$lid"); ?>", "_blank")'
+                                   onclick='window.open("<?php echo $this->createUrl("/admin/export/resources/export/label/lid/$lid"); ?>", "_blank")'
                                    value="<?php $clang->eT("Export Resources As ZIP Archive") ?>"  />
                         </li>
                         <li>
diff --git a/application/views/admin/survey/deleteSurvey_view.php b/application/views/admin/survey/deleteSurvey_view.php
index 689beaa717f..c197f4a0976 100644
--- a/application/views/admin/survey/deleteSurvey_view.php
+++ b/application/views/admin/survey/deleteSurvey_view.php
@@ -8,7 +8,7 @@
         <font color='red'><strong><?php $clang->eT("Error"); ?></strong></font>
         <?php $clang->eT("You have not selected a survey to delete"); ?>
         <br /><br />
-        <input type='submit' value='<?php $clang->eT("Main Admin Screen"); ?>' onclick="window.open('<?php echo Yii::app()->getController()->createUrl("admin"); ?>', '_top')"/>
+        <input type='submit' value='<?php $clang->eT("Main Admin Screen"); ?>' onclick="window.open('<?php echo $this->createUrl('admin/index'); ?>', '_top')"/>
         </td></tr></table>
         </body></html>
         <?php return;
@@ -21,18 +21,18 @@
         <?php $clang->eT("It will also delete any resources/files that have been uploaded for this survey."); ?><br /><br />
         <?php $clang->eT("We recommend that before you delete this survey you export the entire survey from the main administration screen.");
 
-        if (Yii::app()->db->schema->getTable('{{survey_$surveyid}}'))
+        if (tableExists("{{survey_{$surveyid}}}"))
         { ?>
             <br /><br /><?php $clang->eT("This survey is active and a responses table exists. If you delete this survey, these responses (and files) will be deleted. We recommend that you export the responses before deleting this survey."); ?><br /><br />
         <?php }
 
-        if (Yii::app()->db->schema->getTable('{{tokens_$surveyid}}'))
+        if (tableExists("{{tokens_{$surveyid}}}"))
         { ?>
             <?php $clang->eT("This survey has an associated tokens table. If you delete this survey this tokens table will be deleted. We recommend that you export or backup these tokens before deleting this survey."); ?><br /><br />
         <?php } ?>
 
         <p>
-        <input type='submit'  value='<?php $clang->eT("Delete survey"); ?>' onclick="window.open('<?php echo Yii::app()->getController()->createUrl("admin/survey/delete/surveyid/$surveyid"); ?>', '_top')" />
-        <input type='submit'  value='<?php $clang->eT("Cancel"); ?>' onclick="window.open('<?php echo Yii::app()->getController()->createUrl("admin/survey/view/$surveyid"); ?>', '_top')" />
+        <input type='submit' value='<?php $clang->eT("Delete survey"); ?>' onclick="window.open('<?php echo $this->createUrl("admin/survey/delete/surveyid/$surveyid/delete/yes"); ?>', '_top')" />
+        <input type='submit' value='<?php $clang->eT("Cancel"); ?>' onclick="window.open('<?php echo $this->createUrl("admin/survey/view/$surveyid"); ?>', '_top')" />
     <?php } ?>
 </div><br />&nbsp;
\ No newline at end of file
diff --git a/application/views/admin/survey/importSurveyResources_view.php b/application/views/admin/survey/importSurveyResources_view.php
index 5814518ef71..35f229e4702 100644
--- a/application/views/admin/survey/importSurveyResources_view.php
+++ b/application/views/admin/survey/importSurveyResources_view.php
@@ -1,24 +1,52 @@
-<div class='header ui-widget-header'><?php echo $clang->gT("Import survey resources"); ?></div>
+<div class='header ui-widget-header'><?php $clang->eT("Import survey resources"); ?></div>
 <div class='messagebox ui-corner-all'>
-    <strong><?php echo $clang->gT("Imported Resources for"); ?>" SID:</strong> <?php echo $surveyid; ?><br /><br />
+    <div class="successheader"><?php $clang->eT("Success") ?></div><br />
+    <?php $clang->eT("File upload succeeded.") ?><br /><br />
+    <?php $clang->eT("Reading file..") ?><br /><br />
+<?php
+    $ImportListHeader = '';
+    if (!count($aErrorFilesInfo) &&count($aImportedFilesInfo))
+    {
+        $status = $clang->gT("Success");
+        $statusClass = 'successheader';
+        $okfiles = count($aImportedFilesInfo);
+        $errfiles=0;
+        $ImportListHeader .= "<br /><strong><u>" . $clang->gT("Imported Files List") . ":</u></strong><br />\n";
+    }
+    elseif (count($aErrorFilesInfo) &&count($aImportedFilesInfo))
+    {
+        $status = $clang->gT("Partial");
+        $statusClass = 'partialheader';
+        $okfiles = count($aImportedFilesInfo);
+        $errfiles = count($aErrorFilesInfo);
+        $ErrorListHeader = "<br /><strong><u>" . $clang->gT("Error Files List") . ":</u></strong><br />\n";
+        $ImportListHeader .= "<br /><strong><u>" . $clang->gT("Imported Files List") . ":</u></strong><br />\n";
+    }
+    else
+    {
+        $okfiles = 0;
+        $status = $clang->gT("Error");
+        $statusClass = 'warningheader';
+        $errfiles = count($aErrorFilesInfo);
+        $ErrorListHeader = "<br /><strong><u>" . $clang->gT("Error Files List") . ":</u></strong><br />\n";
+    }
+?>
+    <strong><?php $clang->eT("Imported Resources for"); ?>" SID:</strong> <?php echo $surveyid; ?><br /><br />
     <div class="<?php echo $statusClass; ?>">
         <?php echo $status; ?>
     </div>
     <br />
     <strong>
-        <u><?php echo $clang->gT("Resources Import Summary"); ?></u>
+        <u><?php $clang->eT("Resources Import Summary"); ?></u>
     </strong>
     <br />
-    <?php echo $clang->gT("Total Imported files"); ?>: <?php echo count($aImportedFilesInfo); ?><br />
-    <?php echo $clang->gT("Total Errors"); ?>: <?php echo $count($aErrorFilesInfo); ?><br />
-    <div class="successheader"><?php echo !empty($aImportedFilesInfo) && empty($aErrorFilesInfo)? $clang->gT("Success") : $clang->gT("Partial"); ?></div><br /><br />
-    <?php echo $clang->gT("File upload succeeded.") ?><br /><br />
-    <?php echo $clang->gT("Reading file...") ?><br /><br />
-<?php
+    <?php $clang->eT("Total Imported files"); ?>: <?php echo count($okfiles); ?><br />
+    <?php $clang->eT("Total Errors"); ?>: <?php echo count($errfiles); ?><br />
+    <?php
 if (!empty($aImportedFilesInfo))
 {
 ?>
-    <strong><?php echo $clang->gT("Imported Files List") ?>:</strong><br />
+    <strong><?php $clang->eT("Imported Files List") ?>:</strong><br />
     <ul>
 <?php
     foreach ($aImportedFilesInfo as $entry)
@@ -33,18 +61,15 @@
 <?php
 if (!empty($aErrorFilesInfo))
 {
-?>
-    <strong><?php echo $clang->gT("Error Files List") ?>:</strong><br />
-    <ul>
-<?php
+    echo $ErrorListHeader;
     foreach ($aErrorFilesInfo as $entry)
     {
-        echo CHtml::tag('li', array(), $clang->gT("File") . ': ' . $entry["filename"]);
+        echo CHtml::tag('li', array(), $clang->gT("File") . ': ' . $entry['filename'] . " (" . $entry['status'] . ")");
     }
 ?>
     </ul>
 <?php
 }
 ?>
-    <input type='submit' value='<?php echo $clang->gT("Back"); ?>' onclick="window.open('<?php echo $this->createUrl('admin/survey/editsurveysettings/surveyid/' . $surveyid); ?>', '_top')" />\n";
+    <input type='submit' value='<?php $clang->eT("Back"); ?>' onclick="window.open('<?php echo $this->createUrl('admin/survey/editsurveysettings/surveyid/' . $surveyid); ?>', '_top')" />
 </div>
diff --git a/application/views/admin/survey/subview/tabResourceManagement_view.php b/application/views/admin/survey/subview/tabResourceManagement_view.php
index 57b74f56936..d19e821ef7f 100644
--- a/application/views/admin/survey/subview/tabResourceManagement_view.php
+++ b/application/views/admin/survey/subview/tabResourceManagement_view.php
@@ -1,21 +1,16 @@
-<?php
-	$yii = Yii::app();
-	$controller = $yii->getController();
-?>
 <div id='resources'>
-            <form enctype='multipart/form-data'  class='form30' id='importsurveyresources' name='importsurveyresources' action='<?php echo $yii->homeUrl.('/admin/survey/importsurveyresources/'); ?>' method='post' onsubmit='return validatefilename(this,"<?php $clang->eT('Please select a file to import!','js'); ?>");'>
-            <input type='hidden' name='sid' value='<?php echo $surveyid; ?>' />
-            <input type='hidden' name='action' value='importsurveyresources' />
-            <ul>
+    <form enctype='multipart/form-data'  class='form30' id='importsurveyresources' name='importsurveyresources' action='<?php echo $this->createUrl('admin/survey/importsurveyresources/'); ?>' method='post' onsubmit='return validatefilename(this,"<?php $clang->eT('Please select a file to import!', 'js'); ?>");'>
+        <input type='hidden' name='surveyid' value='<?php echo $surveyid; ?>' />
+        <input type='hidden' name='action' value='importsurveyresources' />
+        <ul>
             <li><label>&nbsp;</label>
-            <input type='button' onclick='window.open("<?php echo $yii->getConfig('sCKEditorURL'); ?>/editor/filemanager/browser/default/browser.html?Connector=../../connectors/php/connector.php", "_blank")' value="<?php $clang->eT("Browse Uploaded Resources"); ?>" <?php echo $disabledIfNoResources; ?> /></li>
+                <input type='button' onclick='window.open("<?php echo $this->createUrl("admin/kcfinder/index/load/browse"); ?>", "_blank")' value="<?php $clang->eT("Browse Uploaded Resources"); ?>" <?php echo $disabledIfNoResources; ?> /></li>
             <li><label>&nbsp;</label>
-            <input type='button' onclick='window.open("<?php echo $yii->homeUrl.("/admin/export/resources/exportsurvresources/surveyid/$surveyid/"); ?>", "_blank")' value="<?php $clang->eT("Export Resources As ZIP Archive"); ?>" <?php echo $disabledIfNoResources; ?> /></li>
+                <input type='button' onclick='window.open("<?php echo $this->createUrl("admin/export/resources/export/survey/surveyid/$surveyid/"); ?>", "_blank")' value="<?php $clang->eT("Export Resources As ZIP Archive"); ?>" <?php echo $disabledIfNoResources; ?> /></li>
             <li><label for='the_file'><?php $clang->eT("Select ZIP File:"); ?></label>
-            <input id='the_file' name='the_file' type='file' size='50' /></li>
+                <input id='the_file' name='the_file' type='file' size='50' /></li>
             <li><label>&nbsp;</label>
-            <input type='button' value='<?php $clang->eT("Import Resources ZIP Archive"); ?>' <?php echo $ZIPimportAction; ?> /></li>
-            </ul></form>
-
-        
+                <input type='button' value='<?php $clang->eT("Import Resources ZIP Archive"); ?>' <?php echo $ZIPimportAction; ?> /></li>
+        </ul>
+    </form>
 </div>
\ No newline at end of file
diff --git a/scripts/admin/admin_core.js b/scripts/admin/admin_core.js
index 14458fad9b8..1d2f05f00bc 100644
--- a/scripts/admin/admin_core.js
+++ b/scripts/admin/admin_core.js
@@ -713,7 +713,7 @@ function js2php(object){
     for (property in object){
         var value = object[property];
         if (typeof(value)=="string"){
-            json += '"'+property+'":"'+value+'",'
+            json += '"'+property+'":"'+value.replace(/[\\"']/g, '\\$&')+'",'
         }
         else{
             if (!value[0]){
diff --git a/scripts/admin/ckeditor.36/limesurvey-config.js b/scripts/admin/ckeditor.36/limesurvey-config.js
index 5ff8c512dbd..b785753f8bc 100644
--- a/scripts/admin/ckeditor.36/limesurvey-config.js
+++ b/scripts/admin/ckeditor.36/limesurvey-config.js
@@ -1,13 +1,13 @@
 CKEDITOR.editorConfig = function( config )
 {
 
-    config.filebrowserBrowseUrl = CKEDITOR.basePath+'../kcfinder/browse.php?type=files';
-    config.filebrowserImageBrowseUrl = CKEDITOR.basePath+'../kcfinder/browse.php?type=images';
-    config.filebrowserFlashBrowseUrl = CKEDITOR.basePath+'../kcfinder/browse.php?type=flash';
+    config.filebrowserBrowseUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/browse?type=files';
+    config.filebrowserImageBrowseUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/browse?type=images';
+    config.filebrowserFlashBrowseUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/browse?type=flash';
 
-    config.filebrowserUploadUrl = CKEDITOR.basePath+'../kcfinder/upload.php?type=files';
-    config.filebrowserImageUploadUrl = CKEDITOR.basePath+'../kcfinder/upload.php?type=images';
-    config.filebrowserFlashUploadUrl = CKEDITOR.basePath+'../kcfinder/upload.php?type=flash';
+    config.filebrowserUploadUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/upload?type=files';
+    config.filebrowserImageUploadUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/upload?type=images';
+    config.filebrowserFlashUploadUrl = CKEDITOR.basePath+'../../../../admin/kcfinder/index/load/upload?type=flash';
 
 	config.skin = 'office2003';
 	config.toolbarCanCollapse = false;
diff --git a/scripts/admin/kcfinder/core/autoload.php b/scripts/admin/kcfinder/core/autoload.php
index 7f491642df9..678c6f392ff 100644
--- a/scripts/admin/kcfinder/core/autoload.php
+++ b/scripts/admin/kcfinder/core/autoload.php
@@ -10,85 +10,4 @@
 *   @license http://www.opensource.org/licenses/gpl-2.0.php GPLv2
 *   @license http://www.opensource.org/licenses/lgpl-2.1.php LGPLv2
 *      @link http://kcfinder.sunhater.com
-*/
-
-require_once(dirname(__FILE__).'/../../../../config-defaults.php');
-require_once(dirname(__FILE__).'/../../../../common.php');
-require_once(dirname(__FILE__).'/../../../admin_functions.php');
-
-$usquery = "SELECT stg_value FROM ".db_table_name("settings_global")." where stg_name='SessionName'";
-$usresult = db_execute_assoc($usquery,'',true);
-if ($usresult)
-{
-    $usrow = $usresult->FetchRow();
-    @session_name($usrow['stg_value']);
-}
-else
-{
-    session_name("LimeSurveyAdmin");
-}
-
-session_set_cookie_params(0,$relativeurl.'/');
-
-if (session_id() == "") @session_start();
-
-$_SESSION['KCFINDER'] = array();
-
-$sAllowedExtensions = implode(' ',array_map('trim',explode(',',$allowedresourcesuploads)));
-$_SESSION['KCFINDER']['types']=array('files'=>$sAllowedExtensions,
-'flash'=>$sAllowedExtensions,
-'images'=>$sAllowedExtensions);
-
-if ($demoMode === false &&
-isset($_SESSION['loginID']) &&
-isset($_SESSION['FileManagerContext']))
-{
-    // disable upload at survey creation time
-    // because we don't know the sid yet
-    if (preg_match('/^(create|edit):(question|group|answer)/',$_SESSION['FileManagerContext']) != 0 ||
-    preg_match('/^edit:survey/',$_SESSION['FileManagerContext']) !=0 ||
-    preg_match('/^edit:assessments/',$_SESSION['FileManagerContext']) !=0 ||
-    preg_match('/^edit:emailsettings/',$_SESSION['FileManagerContext']) != 0)
-    {
-        $contextarray=explode(':',$_SESSION['FileManagerContext'],3);
-        $surveyid=$contextarray[2];
-
-
-
-        if(bHasSurveyPermission($surveyid,'surveycontent','update'))
-        {
-            $_SESSION['KCFINDER']['disabled'] = false ;
-            $_SESSION['KCFINDER']['uploadURL'] = "{$relativeurl}/upload/surveys/{$surveyid}/" ;
-            $_SESSION['KCFINDER']['uploadDir'] = $uploaddir.'/surveys/'.$surveyid;
-        }
-
-    }
-    elseif (preg_match('/^edit:label/',$_SESSION['FileManagerContext']) != 0)
-    {
-        $contextarray=explode(':',$_SESSION['FileManagerContext'],3);
-        $labelid=$contextarray[2];
-        // check if the user has label management right and labelid defined
-        if ($_SESSION['USER_RIGHT_MANAGE_LABEL']==1 && isset($labelid) && $labelid != '')
-        {
-            $_SESSION['KCFINDER']['disabled'] = false ;
-            $_SESSION['KCFINDER']['uploadURL'] = "{$relativeurl}/upload/labels/{$labelid}/" ;
-            $_SESSION['KCFINDER']['uploadDir'] = "{$uploaddir}/labels/{$labelid}" ;
-        }
-    }
-
-}
-
-function kcfinder_autoload($class) {
-    if ($class == "uploader")
-        require "core/uploader.php";
-    elseif ($class == "browser")
-        require "core/browser.php";
-    elseif (file_exists("core/types/$class.php"))
-        require "core/types/$class.php";
-    elseif (file_exists("lib/class_$class.php"))
-        require "lib/class_$class.php";
-    elseif (file_exists("lib/helper_$class.php"))
-        require "lib/helper_$class.php";
-}
-
-    spl_autoload_register('kcfinder_autoload');
+*/
\ No newline at end of file
diff --git a/scripts/admin/kcfinder/core/browser.php b/scripts/admin/kcfinder/core/browser.php
index 56a3fe12d0d..fac2538c5cc 100644
--- a/scripts/admin/kcfinder/core/browser.php
+++ b/scripts/admin/kcfinder/core/browser.php
@@ -743,14 +743,14 @@ protected function output($data=null, $template=null) {
         if ($template === null)
             $template = $this->action;
 
-        if (file_exists("tpl/tpl_$template.php")) {
+        if (file_exists(ROOT . "/scripts/admin/kcfinder/tpl/tpl_$template.php")) {
             ob_start();
             $eval = "unset(\$data);unset(\$template);unset(\$eval);";
             $_ = $data;
             foreach (array_keys($data) as $key)
                 if (preg_match('/^[a-z\d_]+$/i', $key))
                     $eval .= "\$$key=\$_['$key'];";
-            $eval .= "unset(\$_);require \"tpl/tpl_$template.php\";";
+            $eval .= "unset(\$_);require ROOT . \"/scripts/admin/kcfinder/tpl/tpl_$template.php\";";
             eval($eval);
             return ob_get_clean();
         }
diff --git a/scripts/admin/kcfinder/core/uploader.php b/scripts/admin/kcfinder/core/uploader.php
index 408e809c5ed..2a797b70d3b 100644
--- a/scripts/admin/kcfinder/core/uploader.php
+++ b/scripts/admin/kcfinder/core/uploader.php
@@ -54,7 +54,7 @@ public function __construct() {
             $this->file = &$_FILES[key($_FILES)];
 
         // LOAD DEFAULT CONFIGURATION
-        require "config.php";
+        require ROOT . "/scripts/admin/kcfinder/config.php";
 
         // SETTING UP SESSION
         if (isset($_CONFIG['_sessionLifetime']))
@@ -66,7 +66,7 @@ public function __construct() {
         @session_start();
 
         // RELOAD DEFAULT CONFIGURATION
-        require "config.php";
+        require ROOT . "/scripts/admin/kcfinder/config.php";
         $this->config = $_CONFIG;
 
         // LOAD SESSION CONFIGURATION IF EXISTS
@@ -414,7 +414,7 @@ protected function makeThumb($file, $overwrite=true) {
     }
 
     protected function localize($langCode) {
-        require "lang/{$langCode}.php";
+        require ROOT . "/scripts/admin/kcfinder/lang/{$langCode}.php";
         setlocale(LC_ALL, $lang['_locale']);
         $this->charset = $lang['_charset'];
         $this->dateTimeFull = $lang['_dateTimeFull'];
diff --git a/scripts/admin/kcfinder/css.php b/scripts/admin/kcfinder/css.php
deleted file mode 100644
index c25d9d2b393..00000000000
--- a/scripts/admin/kcfinder/css.php
+++ /dev/null
@@ -1,245 +0,0 @@
-<?php
-
-/** This file is part of KCFinder project
-  *
-  *      @desc Base CSS definitions
-  *   @package KCFinder
-  *   @version 2.21
-  *    @author Pavel Tzonkov <pavelc@users.sourceforge.net>
-  * @copyright 2010 KCFinder Project
-  *   @license http://www.opensource.org/licenses/gpl-2.0.php GPLv2
-  *   @license http://www.opensource.org/licenses/lgpl-2.1.php LGPLv2
-  *      @link http://kcfinder.sunhater.com
-  */
-
-require "core/autoload.php";
-$mtime = @filemtime(__FILE__);
-if ($mtime) httpCache::checkMTime($mtime);
-$browser = new browser();
-$config = $browser->config;
-ob_start();
-
-?>
-html, body {
-    overflow: hidden;
-}
-
-body, form, th, td {
-    margin: 0;
-    padding: 0;
-}
-
-a {
-    cursor:pointer;
-}
-
-* {
-    font-family: Tahoma, Verdana, Arial, sans-serif;
-    font-size: 11px;
-}
-
-table {
-    border-collapse: collapse;
-}
-
-#all {
-    vvisibility: hidden;
-}
-
-#left {
-    float: left;
-    display: block;
-    width: 25%;
-}
-
-#right {
-    float: left;
-    display: block;
-    width: 75%;
-}
-
-#settings {
-    display: none;
-    padding: 0;
-    float: left;
-    width: 100%;
-}
-
-#settings > div {
-    float: left;
-}
-
-#folders {
-    padding: 5px;
-    overflow: auto;
-}
-
-#toolbar {
-    padding: 5px;
-}
-
-#files {
-    padding: 5px;
-    overflow: auto;
-}
-
-#status {
-    padding: 5px;
-    float: left;
-    overflow: hidden;
-}
-
-#fileinfo {
-    float: left;
-}
-
-#clipboard div {
-    width: 16px;
-    height: 16px;
-}
-
-.folders {
-    margin-left: 16px;
-}
-
-div.file {
-    overflow-x: hidden;
-    width: <?php echo $config['thumbWidth'] ?>px;
-    float: left;
-    text-align: center;
-    cursor: default;
-    white-space: nowrap;
-}
-
-div.file .thumb {
-    width: <?php echo $config['thumbWidth'] ?>px;
-    height: <?php echo $config['thumbHeight'] ?>px;
-    background: no-repeat center center;
-}
-
-#files table {
-    width: 100%;
-}
-
-tr.file {
-    cursor: default;
-}
-
-tr.file > td {
-    white-space: nowrap;
-}
-
-tr.file > td.name {
-    background-repeat: no-repeat;
-    background-position: left center;
-    padding-left: 20px;
-    width: 100%;
-}
-
-tr.file > td.time,
-tr.file > td.size {
-    text-align: right;
-}
-
-#toolbar {
-    cursor: default;
-    white-space: nowrap;
-}
-
-#toolbar a {
-    padding-left: 20px;
-    text-decoration: none;
-    background: no-repeat left center;
-}
-
-#toolbar a:hover, a[href="#upload"].uploadHover {
-    color: #000;
-}
-
-#upload {
-    position: absolute;
-    overflow: hidden;
-    opacity: 0;
-    filter: alpha(opacity:0);
-}
-
-#upload input {
-    cursor: pointer;
-}
-
-#uploadResponse {
-    display: none;
-}
-
-span.brace {
-    padding-left: 11px;
-    cursor: default;
-}
-
-span.brace.opened, span.brace.closed {
-    cursor: pointer;
-}
-
-#shadow {
-    position: absolute;
-    top: 0;
-    left: 0;
-    display: none;
-    background: #fff;
-    z-index: 100;
-    opacity: 0.5;
-    filter: alpha(opacity:50);
-}
-
-#dialog, #clipboard {
-    position: absolute;
-    display: none;
-    z-index: 101;
-    cursor: default;
-}
-
-#clipboard {
-    z-index: 99;
-}
-
-#loading {
-    display: none;
-    float: right;
-}
-
-.menu {
-    background: #888;
-    white-space: nowrap;
-}
-
-.menu a {
-    display: block;
-}
-
-.menu .list {
-    max-height: 0;
-    overflow-y: auto;
-    overflow-x: hidden;
-    white-space: nowrap;
-}
-
-.file .access, .file .hasThumb {
-    display: none;
-}
-
-#dialog img {
-    cursor: pointer;
-}
-
-#resizer {
-    position: absolute;
-    z-index: 98;
-    top: 0;
-    background: #000;
-    opacity: 0;
-    filter: alpha(opacity:0);
-}
-<?php
-
-header("Content-Type: text/css");
-echo text::compressCSS(ob_get_clean());
\ No newline at end of file
diff --git a/scripts/admin/kcfinder/js/browser/misc.js b/scripts/admin/kcfinder/js/browser/misc.js
index e783c976015..2542e1a54e8 100644
--- a/scripts/admin/kcfinder/js/browser/misc.js
+++ b/scripts/admin/kcfinder/js/browser/misc.js
@@ -183,7 +183,7 @@ browser.humanSize = function(size) {
 };
 
 browser.baseGetData = function(act) {
-    var data = 'browse.php?type=' + encodeURIComponent(this.type) + '&lng=' + this.lang;
+    var data = 'browse?type=' + encodeURIComponent(this.type) + '&lng=' + this.lang;
     if (act)
         data += "&act=" + act
     return data;
diff --git a/scripts/admin/kcfinder/js_localize.php b/scripts/admin/kcfinder/js_localize.php
deleted file mode 100644
index 7171aae8f72..00000000000
--- a/scripts/admin/kcfinder/js_localize.php
+++ /dev/null
@@ -1,33 +0,0 @@
-<?php
-
-/** This file is part of KCFinder project
-  *
-  *      @desc Load language labels in JavaScript
-  *   @package KCFinder
-  *   @version 2.21
-  *    @author Pavel Tzonkov <pavelc@users.sourceforge.net>
-  * @copyright 2010 KCFinder Project
-  *   @license http://www.opensource.org/licenses/gpl-2.0.php GPLv2
-  *   @license http://www.opensource.org/licenses/lgpl-2.1.php LGPLv2
-  *      @link http://kcfinder.sunhater.com
-  */
-
-require "core/autoload.php";
-
-if (function_exists('set_magic_quotes_runtime'))
-    @set_magic_quotes_runtime(false);
-
-$input = new input();
-if (!isset($input->get['lng']) || ($input->get['lng'] == 'en')) die;
-$file = "lang/" . $input->get['lng'] . ".php";
-$files = glob("lang/*.php");
-if (!in_array($file, $files)) die;
-$mtime = @filemtime($file);
-if ($mtime) httpCache::checkMTime($mtime);
-require $file;
-header("Content-Type: text/javascript; charset={$lang['_charset']}");
-foreach ($lang as $english => $native)
-    if (substr($english, 0, 1) != "_")
-        echo "browser.labels['" . text::jsValue($english) . "']=\"" . text::jsValue($native) . "\";";
-
-?>
diff --git a/scripts/admin/kcfinder/tpl/tpl__css.php b/scripts/admin/kcfinder/tpl/tpl__css.php
index 27fb9f1fb78..6d0bf4c534e 100644
--- a/scripts/admin/kcfinder/tpl/tpl__css.php
+++ b/scripts/admin/kcfinder/tpl/tpl__css.php
@@ -1,2 +1,2 @@
-<link href="css.php" rel="stylesheet" type="text/css" />
-<link href="themes/<?php echo $this->config['theme'] ?>/style.css" rel="stylesheet" type="text/css" />
+<link href="<?php echo Yii::app()->createUrl('admin/kcfinder/css'); ?>" rel="stylesheet" type="text/css" />
+<link href="<?php echo Yii::app()->createUrl("scripts/admin/kcfinder/themes/{$this->config['theme']}/style.css"); ?>" rel="stylesheet" type="text/css" />
diff --git a/scripts/admin/kcfinder/tpl/tpl__javascript.php b/scripts/admin/kcfinder/tpl/tpl__javascript.php
index e94930eede3..2ddf0cef8d4 100644
--- a/scripts/admin/kcfinder/tpl/tpl__javascript.php
+++ b/scripts/admin/kcfinder/tpl/tpl__javascript.php
@@ -1,14 +1,14 @@
-<script src="js/jquery.js" type="text/javascript"></script>
-<script src="js/jquery.rightClick.js" type="text/javascript"></script>
-<script src="js/jquery.drag.js" type="text/javascript"></script>
-<script src="js/helper.js" type="text/javascript"></script>
-<script src="js/browser/joiner.php" type="text/javascript"></script>
-<script src="js_localize.php?lng=<?php echo $this->lang ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl('scripts/admin/kcfinder/js/jquery.js'); ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl('scripts/admin/kcfinder/js/jquery.rightClick.js'); ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl('scripts/admin/kcfinder/js/jquery.drag.js'); ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl('scripts/admin/kcfinder/js/helper.js'); ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl('scripts/admin/kcfinder/js/browser/joiner.php'); ?>" type="text/javascript"></script>
+<script src="<?php echo Yii::app()->createUrl("admin/kcfinder/js_localize?lng={$this->lang}"); ?>" type="text/javascript"></script>
 <?php IF (isset($this->opener['TinyMCE']) && $this->opener['TinyMCE']): ?>
 <script src="<?php echo $this->config['_tinyMCEPath'] ?>/tiny_mce_popup.js" type="text/javascript"></script>
 <?php ENDIF ?>
-<?php IF (file_exists("themes/{$this->config['theme']}/init.js")): ?>
-<script src="themes/<?php echo $this->config['theme'] ?>/init.js" type="text/javascript"></script>
+<?php IF (file_exists(ROOT . "/scripts/admin/kcfinder/themes/{$this->config['theme']}/init.js")): ?>
+<script src="<?php echo Yii::app()->createUrl("scripts/admin/kcfinder/themes/{$this->config['theme']}/init.js"); ?>" type="text/javascript"></script>
 <?php ENDIF ?>
 <script type="text/javascript">
 browser.support.chromeFrame = <?php echo (strpos(strtolower($_SERVER['HTTP_USER_AGENT']), " chromeframe") !== false) ? "true" : "false" ?>;
diff --git a/scripts/admin/kcfinder/tpl/tpl_browser.php b/scripts/admin/kcfinder/tpl/tpl_browser.php
index 674a50d4b08..99679486feb 100644
--- a/scripts/admin/kcfinder/tpl/tpl_browser.php
+++ b/scripts/admin/kcfinder/tpl/tpl_browser.php
@@ -2,8 +2,8 @@
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
 <title>KCFinder: /<?php echo $this->session['dir'] ?></title>
-<?php INCLUDE "tpl/tpl__css.php" ?>
-<?php INCLUDE "tpl/tpl__javascript.php" ?>
+<?php INCLUDE ROOT . "/scripts/admin/kcfinder/tpl/tpl__css.php" ?>
+<?php INCLUDE ROOT . "/scripts/admin/kcfinder/tpl/tpl__javascript.php" ?>
 </head>
 <body>
 <script type="text/javascript">
