diff --git a/lib/receive-incoming-sails-io-msg.js b/lib/receive-incoming-sails-io-msg.js
index 2ddca3b..5be5704 100644
--- a/lib/receive-incoming-sails-io-msg.js
+++ b/lib/receive-incoming-sails-io-msg.js
@@ -1,20 +1,21 @@
 /**
  * Module dependencies
  */
 
 var util = require('util');
+var url = require('url');
 var _ = require('@sailshq/lodash');
 var semver = require('semver');
 var parseSdkMetadata = require('./parse-sdk-metadata');
 var ERRORPACK = require('./errors');
 var proxyaddr = require('proxy-addr');
 
 /**
  * @required  {Object} app
  *
  * @return {Function}     [initialize]
  */
 module.exports = function ToReceiveIncomingSailsIOMsg(app) {
 
   /**
    * Parse an incoming Socket.io message (usually from the sails.io.js client),
@@ -93,30 +94,32 @@ module.exports = function ToReceiveIncomingSailsIOMsg(app) {
       protocol: 'ws', // TODO: consider if this is really helpful or just a waste of LoC
 
       isSocket: true,
 
       ip      : ip,
 
       ips     : ips,
 
       port    : null,
 
       // Access to underlying SIO socket
       socket  : options.socket,
 
       url     : options.incomingSailsIOMsg.url,
 
+      path    : url.parse(options.incomingSailsIOMsg.url).pathname,
+
       method  : options.eventName,
 
       // Attached data becomes simulated HTTP body (`req.body`)
       // (allow `params` or `data` to be specified for backwards/sideways-compatibility)
       body    : _.isArray(options.incomingSailsIOMsg.data) ? options.incomingSailsIOMsg.data : _.extend({}, options.incomingSailsIOMsg.params || {}, options.incomingSailsIOMsg.data || {}),
 
       // Allow optional headers
       headers: _.defaults({
 
         host: app.config.host,
 
         // Default the "cookie" request header to what was provided in the handshake.
         cookie: (function (){
           var _cookie;
           try {
