diff --git a/mistune.py b/mistune.py
index e68c5fb..67b0880 100644
--- a/mistune.py
+++ b/mistune.py
@@ -145,31 +145,31 @@ class BlockGrammar(object):
         r'))+)\n*' % (
             _pure_pattern(fences).replace(r'\1', r'\2'),
             _pure_pattern(list_block).replace(r'\1', r'\3'),
             _pure_pattern(hrule),
             _pure_pattern(heading),
             _pure_pattern(lheading),
             _pure_pattern(block_quote),
             _pure_pattern(def_links),
             _pure_pattern(def_footnotes),
             '<' + _block_tag,
         )
     )
     block_html = re.compile(
         r'^ *(?:%s|%s|%s) *(?:\n{2,}|\s*$)' % (
             r'<!--[\s\S]*?-->',
-            r'<(%s)((?:%s)*?)>([\s\S]+?)<\/\1>' % (_block_tag, _valid_attr),
+            r'<(%s)((?:%s)*?)\s*>([\s\S]+?)<\/\1>' % (_block_tag, _valid_attr),
             r'<%s(?:%s)*?\s*\/?>' % (_block_tag, _valid_attr),
         )
     )
     table = re.compile(
         r'^ *\|(.+)\n *\|( *[-:]+[-| :]*)\n((?: *\|.*(?:\n|$))*)\n*'
     )
     nptable = re.compile(
         r'^ *(\S.*\|.*)\n *([-:]+ *\|[-| :]*)\n((?:.*\|.*(?:\n|$))*)\n*'
     )
     text = re.compile(r'^[^\n]+')
 
 
 class BlockLexer(object):
     """Block level lexer for block grammars."""
     grammar_class = BlockGrammar
@@ -435,31 +435,31 @@ class BlockLexer(object):
         text = m.group(1).rstrip('\n')
         self.tokens.append({'type': 'paragraph', 'text': text})
 
     def parse_text(self, m):
         text = m.group(0)
         self.tokens.append({'type': 'text', 'text': text})
 
 
 class InlineGrammar(object):
     """Grammars for inline level tokens."""
 
     escape = re.compile(r'^\\([\\`*{}\[\]()#+\-.!_>~|])')  # \* \+ \! ....
     inline_html = re.compile(
         r'^(?:%s|%s|%s)' % (
             r'<!--[\s\S]*?-->',
-            r'<(\w+%s)((?:%s)*?)>([\s\S]*?)<\/\1>' % (_valid_end, _valid_attr),
+            r'<(\w+%s)((?:%s)*?)\s*>([\s\S]*?)<\/\1>' % (_valid_end, _valid_attr),
             r'<\w+%s(?:%s)*?\s*\/?>' % (_valid_end, _valid_attr),
         )
     )
     autolink = re.compile(r'^<([^ >]+(@|:)[^ >]+)>')
     link = re.compile(
         r'^!?\[('
         r'(?:\[[^^\]]*\]|[^\[\]]|\](?=[^\[]*\]))*'
         r')\]\('
         r'''\s*(<)?([\s\S]*?)(?(2)>)(?:\s+['"]([\s\S]*?)['"])?\s*'''
         r'\)'
     )
     reflink = re.compile(
         r'^!?\[('
         r'(?:\[[^^\]]*\]|[^\[\]]|\](?=[^\[]*\]))*'
         r')\]\s*\[([^^\]]*)\]'
diff --git a/tests/test_extra.py b/tests/test_extra.py
index 7318444..07c1ca3 100644
--- a/tests/test_extra.py
+++ b/tests/test_extra.py
@@ -65,30 +65,36 @@ def test_parse_inline_html():
     ret = mistune.markdown(
         '<div>**foo**</div>', parse_inline_html=True, escape=False
     )
     assert '<strong>' not in ret
     ret = mistune.markdown(
         '<span>**foo**</span>', parse_inline_html=True, escape=False
     )
     assert '<span><strong>' in ret
 
     ret = mistune.markdown(
         '<a>http://lepture.com</a>', parse_inline_html=True, escape=False
     )
     assert 'href' not in ret
 
 
+def test_block_html():
+    ret = mistune.markdown(
+        '<div ></div>', escape=False
+    )
+    assert '<div ></div>' in ret
+
 def test_parse_block_html():
     ret = mistune.markdown(
         '<div>**foo**</div>', parse_block_html=True, escape=False
     )
     assert '<div><strong>' in ret
     ret = mistune.markdown(
         '<span>**foo**</span>', parse_block_html=True, escape=False
     )
     assert '<strong>' not in ret
 
 
 def test_trigger_more_cases():
     markdown = mistune.Markdown(
         inline=mistune.InlineLexer,
         block=mistune.BlockLexer,
