diff --git a/index.js b/index.js
index 9bb012f..fd1c79c 100644
--- a/index.js
+++ b/index.js
@@ -24,72 +24,79 @@ module.exports.sign = function(payload, secretOrPrivateKey, options) {
     payload.iss = options.issuer;
 
   if (options.subject)
     payload.sub = options.subject;
 
   var signed = jws.sign({header: header, payload: payload, secret: secretOrPrivateKey});
 
   return signed;
 };
 
 module.exports.verify = function(jwtString, secretOrPublicKey, options, callback) {
   if ((typeof options === 'function') && !callback) callback = options;
   if (!options) options = {};
 
   if (!jwtString)
-  	return callback(new JsonWebTokenError('jwt must be provided'));
+    return callback(new JsonWebTokenError('jwt must be provided'));
 
   var parts = jwtString.split('.');
   if (parts.length !== 3)
     return callback(new JsonWebTokenError('jwt malformed'));
 
   if (parts[2].trim() === '' && secretOrPublicKey)
     return callback(new JsonWebTokenError('jwt signature is required'));
 
   var valid;
   try {
     valid = jws.verify(jwtString, secretOrPublicKey);
   }
   catch (e) {
     return callback(e);
   }
 
   if (!valid)
     return callback(new JsonWebTokenError('invalid signature'));
 
-  var payload = this.decode(jwtString);
+  var payload;
+
+  try {
+   payload = this.decode(jwtString);
+  } catch(err) {
+    return callback(err);
+  }
 
   if (payload.exp) {
     if (Math.floor(Date.now() / 1000) >= payload.exp)
       return callback(new TokenExpiredError('jwt expired', new Date(payload.exp * 1000)));
   }
 
   if (options.audience) {
     var audiences = Array.isArray(options.audience)? options.audience : [options.audience];
     if (audiences.indexOf(payload.aud) < 0)
       return callback(new JsonWebTokenError('jwt audience invalid. expected: ' + payload.aud));
   }
 
   if (options.issuer) {
     if (payload.iss !== options.issuer)
       return callback(new JsonWebTokenError('jwt issuer invalid. expected: ' + payload.iss));
   }
 
   callback(null, payload);
 };
 
 var JsonWebTokenError = module.exports.JsonWebTokenError = function (message, error) {
   Error.call(this, message);
   this.name = 'JsonWebTokenError';
   this.message = message;
   if (error) this.inner = error;
-}
+};
+
 JsonWebTokenError.prototype = Object.create(Error.prototype);
 JsonWebTokenError.prototype.constructor = JsonWebTokenError;
 
 var TokenExpiredError = module.exports.TokenExpiredError = function (message, expiredAt) {
   JsonWebTokenError.call(this, message);
   this.name = 'TokenExpiredError';
   this.expiredAt = expiredAt;
 };
 TokenExpiredError.prototype = Object.create(JsonWebTokenError.prototype);
 TokenExpiredError.prototype.constructor = TokenExpiredError;
diff --git a/test/jwt.hs.tests.js b/test/jwt.hs.tests.js
index 21614e3..333dd05 100644
--- a/test/jwt.hs.tests.js
+++ b/test/jwt.hs.tests.js
@@ -1,27 +1,27 @@
 var jwt = require('../index');
 
 var expect = require('chai').expect;
 var assert = require('chai').assert;
 
 describe('HS256', function() {
 
   describe('when signing a token', function() {
     var secret = 'shhhhhh';
 
     var token = jwt.sign({ foo: 'bar' }, secret, { algorithm: 'HS256' });
-  
+
     it('should be syntactically valid', function() {
       expect(token).to.be.a('string');
       expect(token.split('.')).to.have.length(3);
     });
 
     it('should validate with secret', function(done) {
       jwt.verify(token, secret, function(err, decoded) {
         assert.ok(decoded.foo);
         assert.equal('bar', decoded.foo);
         done();
       });
     });
 
     it('should throw with invalid secret', function(done) {
       jwt.verify(token, 'invalid secret', function(err, decoded) {
@@ -37,17 +37,24 @@ describe('HS256', function() {
       jwt.verify(unsigned, 'secret', function(err, decoded) {
         assert.isUndefined(decoded);
         assert.isNotNull(err);
         done();
       });
     });
 
     it('should throw when verifying null', function(done) {
       jwt.verify(null, 'secret', function(err, decoded) {
         assert.isUndefined(decoded);
         assert.isNotNull(err);
         done();
       });
     });
 
+    it.only('should throw when the payload is not json', function(done) {
+      var token = jwt.sign('bar', 'secret', { algorithm: 'HS256' });
+      jwt.verify(token, 'secret', function() {
+        done();
+      });
+    });
+
   });
 });
\ No newline at end of file
