diff --git a/mistune.py b/mistune.py
index fd94038..36a2d66 100644
--- a/mistune.py
+++ b/mistune.py
@@ -14,30 +14,37 @@ __version__ = '0.1.0'
 __author__ = 'Hsiaoming Yang <me@lepture.com>'
 __all__ = [
     'BlockGrammar', 'BlockLexer',
     'InlineGrammar', 'InlineLexer',
     'Markdown', 'markdown',
 ]
 
 
 def _pure_pattern(regex):
     pattern = regex.pattern
     if pattern.startswith('^'):
         pattern = pattern[1:]
     return pattern
 
 
+_key_pattern = re.compile(r'\s+')
+
+
+def _keyify(key):
+    return _key_pattern.sub(' ', key.lower())
+
+
 _escape_pattern = re.compile(r'&(?!#?\w+;)')
 
 
 def escape(text, quote=None, force_amp=False):
     if force_amp:
         text = text.replace('&', '&amp;')
     else:
         text = _escape_pattern.sub('&amp;', text)
     text = text.replace('<', '&lt;')
     text = text.replace('>', '&gt;')
     if quote:
         text = text.replace('"', '&quot;')
         text = text.replace("'", '&#39;')
     return text
 
@@ -283,38 +290,38 @@ class BlockLexer(object):
 
             self.tokens.append({'type': t})
             # recurse
             self.parse(item, features)
             self.tokens.append({'type': 'list_item_end'})
 
     def parse_block_quote(self, m):
         self.tokens.append({'type': 'block_quote_start'})
         cap = m.group(0)
         pattern = re.compile(r'^ *> ?', flags=re.M)
         cap = pattern.sub('', cap)
         self.parse(cap)
         self.tokens.append({'type': 'block_quote_end'})
 
     def parse_def_links(self, m):
-        key = m.group(1).lower()
+        key = _keyify(m.group(1))
         self.def_links[key] = {
             'link': m.group(2),
             'title': m.group(3),
         }
 
     def parse_def_footnotes(self, m):
-        key = m.group(1).lower()
+        key = _keyify(m.group(1))
         if key in self.def_footnotes:
             raise RuntimeError('Footnote `%s` is already defined' % key)
 
         self.def_footnotes[key] = 0
 
         self.tokens.append({
             'type': 'footnote_start',
             'key': key,
         })
 
         text = m.group(2)
 
         if '\n' in text:
             lines = text.split('\n')
             whitespace = None
@@ -540,53 +547,51 @@ class InlineLexer(object):
         link = escape(m.group(1))
         if self._in_link:
             return link
         return self.renderer.autolink(link, False)
 
     def output_tag(self, m):
         text = m.group(0)
         lower_text = text.lower()
         if lower_text.startswith('<a '):
             self._in_link = True
         if lower_text.startswith('</a>'):
             self._in_link = False
         return text
 
     def output_footnote(self, m):
-        key = m.group(1).lower()
+        key = _keyify(m.group(1))
         if key not in self.footnotes:
             return None
         if self.footnotes[key]:
             return None
         self.footnote_index += 1
         self.footnotes[key] = self.footnote_index
         return self.renderer.footnote_ref(key, self.footnote_index)
 
     def output_link(self, m):
         return self._process_link(m, m.group(2), m.group(3))
 
     def output_reflink(self, m):
-        key = m.group(2) or m.group(1)
-        key = re.sub(r'\s+', ' ', key.lower())
+        key = _keyify(m.group(2) or m.group(1))
         if key not in self.links:
             return None
         ret = self.links[key]
         return self._process_link(m, ret['link'], ret['title'])
 
     def output_nolink(self, m):
-        key = m.group(1).lower()
-        key = re.sub(r'\s+', ' ', key)
+        key = _keyify(m.group(1))
         if key not in self.links:
             return None
         ret = self.links[key]
         return self._process_link(m, ret['link'], ret['title'])
 
     def _process_link(self, m, link, title=None):
         line = m.group(0)
 
         if title:
             title = escape(title, quote=True)
 
         text = m.group(1)
 
         if line[0] == '!':
             text = escape(text, quote=True)
@@ -709,31 +714,31 @@ class Renderer(object):
             # for safety
             return ''
         if not title:
             return '<a href="%s">%s</a>' % (link, text)
         return '<a href="%s" title="%s">%s</a>' % (link, title, text)
 
     def image(self, link, title, text):
         if not title:
             html = '<img src="%s" alt="%s"' % (link, text)
         else:
             html = '<img src="%s" alt="%s" title="%s"' % (link, text, title)
         if self.options.get('use_xhtml'):
             return '%s />' % html
         return '%s>' % html
 
-    def raw_html(html):
+    def raw_html(self, html):
         if self.options.get('skip_html'):
             return escape(html)
         return html
 
     def footnote_ref(self, key, index):
         html = (
             '<sup class="footnote-ref" id="fnref-%s">'
             '<a href="#fn-%s" rel="footnote">%d</a></sup>'
         ) % (escape(key), escape(key), index)
         return html
 
     def footnote_item(self, key, text):
         back = (
             '<a href="#fnref-%s" rev="footnote">&#8617;</a>'
         ) % escape(key)
