diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..48a2e24
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,2 @@
+components
+build
diff --git a/Readme.md b/Readme.md
new file mode 100644
index 0000000..01d0b9f
--- /dev/null
+++ b/Readme.md
@@ -0,0 +1,26 @@
+
+# dot
+
+  Get and set object properties with dot notation
+
+## Installation
+
+    $ component install eivindfjeldstad/dot
+
+## API
+
+### dot.set(object, path, value)
+```js
+  dot.set(obj, 'cool.aid', 'rocks');
+  assert(obj.cool.aid === 'rocks');
+```
+
+### dot.get(object, path)
+```js
+var value = dot.get(obj, 'cool.aid');
+assert(value === 'rocks');
+```
+
+## License
+
+  MIT
diff --git a/component.json b/component.json
new file mode 100644
index 0000000..17e444c
--- /dev/null
+++ b/component.json
@@ -0,0 +1,14 @@
+{
+  "name": "dot",
+  "repo": "eivindfjeldstad/dot",
+  "description": "Get and set object properties with dot notation",
+  "version": "0.0.1",
+  "keywords": [],
+  "dependencies": {},
+  "development": {},
+  "license": "MIT",
+  "main": "index.js",
+  "scripts": [
+    "index.js"
+  ]
+}
\ No newline at end of file
diff --git a/index.js b/index.js
new file mode 100644
index 0000000..9df840c
--- /dev/null
+++ b/index.js
@@ -0,0 +1,43 @@
+/**
+ * Set given `path`
+ *
+ * @param {Object} obj
+ * @param {String} path
+ * @param {Mixed} val
+ * @api public
+ */
+
+exports.set = function (obj, path, val) {
+  var segs = path.split('.');
+  var attr = segs.pop();
+  
+  for (var i = 0; i < segs.length; i++) {
+    var seg = segs[i];
+    obj[seg] = obj[seg] || {};
+    obj = obj[seg];
+  }
+  
+  obj[attr] = val;
+};
+
+/**
+ * Get given `path`
+ *
+ * @param {Object} obj
+ * @param {String} path
+ * @return {Mixed}
+ * @api public
+ */
+
+exports.get = function (obj, path) {
+  var segs = path.split('.');
+  var attr = segs.pop();
+  
+  for (var i = 0; i < segs.length; i++) {
+    var seg = segs[i];
+    if (!obj[seg]) return;
+    obj = obj[seg];
+  }
+  
+  return obj[attr];
+};
\ No newline at end of file
diff --git a/test/index.js b/test/index.js
new file mode 100644
index 0000000..5d72a95
--- /dev/null
+++ b/test/index.js
@@ -0,0 +1,24 @@
+var assert = require('assert');
+var dot = require('..');
+
+var tests = module.exports = {
+  'test set': function () {
+    var obj = {};
+    dot.set(obj, 'cool.aid', 'rocks');
+    assert(obj.cool.aid === 'rocks');
+  },
+  
+  'test get': function () {
+    var obj = {};
+    obj.cool = {};
+    obj.cool.aid = 'rocks';
+    var value = dot.get(obj, 'cool.aid');
+    assert(value === 'rocks');
+  }
+}
+
+for (var t in tests) {
+  tests[t]();
+}
+  
+console.log('All tests passed!');
\ No newline at end of file
