diff --git a/application/controllers/admin/participantsaction.php b/application/controllers/admin/participantsaction.php
index 134262095d4..14164d62620 100644
--- a/application/controllers/admin/participantsaction.php
+++ b/application/controllers/admin/participantsaction.php
@@ -553,19 +553,21 @@ function exporttocsvcount()
             {
                 $condition = explode("||", $searchcondition);
                 $query = Participants::model()->getParticipantsSearchMultiple($condition, 0, 0);
+                $count = count($query);
             }
             else // if no search criteria all the participants will be counted
             {
-                $query = Participants::model()->getParticipantsWithoutLimit();
+                $count = Participants::model()->getParticipantsCountWithoutLimit();
             }
         }
         else // If no search criteria it will simply return the number of participants
         {
             $iUserID = Yii::app()->session['loginID'];
             $query = Participants::model()->getParticipantsOwner($iUserID);
+            $count = count($query);
         }
 
-        echo sprintf($clang->gT("Export %s participant(s) to CSV"), count($query));
+        echo sprintf($clang->gT("Export %s participant(s) to CSV"), $count);
     }
 
     /**
@@ -575,21 +577,22 @@ function exporttocsvcountAll()
     {
         if (Yii::app()->session['USER_RIGHT_SUPERADMIN']) //If super admin all the participants in the central table will be counted
         {
-            $query = Participants::model()->getParticipantsWithoutLimit();
+            $count = Participants::model()->getParticipantsCountWithoutLimit();
         }
         else // otherwise only the participants on which the logged in user has the rights
         {
             $iUserID = Yii::app()->session['loginID'];
             $query = Participants::model()->getParticipantsOwner($iUserID);
+            $count = count($query);
         }
 
-        if (count($query) > 0) // If count is greater than 0 it will show the message
+        if ($count > 0) // If count is greater than 0 it will show the message
         {
-            printf($clang->gT("Export %s participant(s) to CSV"), count($query));
+            printf($clang->gT("Export %s participant(s) to CSV"), $count);
         }
         else // else it will return a numeric count which will tell that there is no participant to be exported
         {
-            echo count($query);
+            echo $count;
         }
     }
 
@@ -693,14 +696,15 @@ function getaddtosurveymsg()
         {
             if (Yii::app()->session['USER_RIGHT_SUPERADMIN']) //If super admin all the participants will be visible
             {
-                $query = Participants::model()->getParticipantsWithoutLimit();
+                $count = Participants::model()->getParticipantsCountWithoutLimit();
             }
             else
             {
                 $query = Participants::model()->getParticipantsOwner(Yii::app()->session['loginID']);
+                $count = count($query);
             }
 
-            printf($this->getController()->lang->gT("%s participant(s) are to be copied "), count($query));
+            printf($this->getController()->lang->gT("%s participant(s) are to be copied "), $count);
         }
     }
 
@@ -855,6 +859,9 @@ function exporttocsv()
         }
     }
 
+    /**
+     * Equal to getParticipants_json() but now with a search
+     */
     function getParticipantsResults_json()
     {
     	///admin/participants/getParticipantsResults_json/search/email||contains||com
@@ -999,7 +1006,7 @@ function getParticipants_json()
     {
         $page = Yii::app()->request->getPost('page');
         $limit = Yii::app()->request->getPost('rows');
-    	  $limit = isset($limit) ? $limit : 50; //Stop division by zero errors
+    	$limit = isset($limit) ? $limit : 50; //Stop division by zero errors
 
         $attid = ParticipantAttributeNames::model()->getVisibleAttributes();
         $participantfields = array('participant_id', 'can_edit', 'firstname', 'lastname', 'email', 'blacklisted', 'survey', 'language', 'owner_uid');
@@ -1007,94 +1014,48 @@ function getParticipants_json()
         {
             array_push($participantfields, $value['attribute_id']);
         }
+        $sidx = Yii::app()->request->getPost('sidx');
+        $sidx = !empty($sidx) ? $sidx : "lastname";
+        $sord = Yii::app()->request->getPost('sord');
+        $sord = !empty($sord) ? $sord : "asc";
+        $order = $sidx. " ". $sord;
+        
         $aData = new stdClass;
+        $aData->page = $page;
+        
         //If super admin all the participants will be visible
         if (Yii::app()->session['USER_RIGHT_SUPERADMIN'])
         {
-            $records = Participants::model()->getParticipants($page, $limit,$attid);
-            $aData =  new stdClass;
-            $aData->page = $page;
+            $iUserID = null;
             $aData->records = Participants::model()->count();
-            $aData->total = ceil($aData->records / $limit);
-            $i = 0;
-        	  $sortablearray=array();
-            foreach ($records as $key => $row)
-            {            
-                $surveycount = Participants::model()->getSurveyCount($row['participant_id']);
-                $sortablearray[$i] = array($row['participant_id'], "true", $row['firstname'], $row['lastname'], $row['email'], $row['blacklisted'], $surveycount, $row['language'], $row['ownername']); // since it's the admin he has access to all editing on the participants inspite of what can_edit option is
-                unset($row['participant_id'], $row['firstname'], $row['lastname'], $row['email'], $row['blacklisted'], $row['language'],$row['ownername'],$row['owner_uid']);
-                foreach($row as $key=>$attvalue)
-                {
-                  array_push($sortablearray[$i], $attvalue);
-                }         
-                $i++;
-            }
-
-            $indexsort = array_search(Yii::app()->request->getPost('sidx'), $participantfields);
-        	  if(!empty($sortablearray)) {
-        		$sortedarray = subval_sort($sortablearray, $indexsort, Yii::app()->request->getPost('sord'));
-        		$i = 0;
-        		$count = count($sortedarray[0]);
-        		foreach ($sortedarray as $key => $value)
-        		{
-        			$aData->rows[$i]['id'] = $value[0];
-        			$aData->rows[$i]['cell'] = array();
-        			for ($j = 0; $j < $count; $j++)
-        			{
-        				array_push($aData->rows[$i]['cell'], $value[$j]);
-        			}
-        			$i++;
-        		}
-        	}
-        }
-        // Only the owned and shared participants will be visible
-        else
-        {
+        } else {
             $iUserID = Yii::app()->session['loginID'];
-            $records = Participants::model()->getParticipantsOwner($iUserID);
-            $aData->page = $page;
-            $aData->records = count($records);
-            $aData->total = ceil($aData->records / $limit);
-            $attid = ParticipantAttributeNames::model()->getVisibleAttributes();
-            $i = 0;
-        	$sortablearray=array();
-            foreach ($records as $row)
-            {
-                $surveycount = Participants::model()->getSurveyCount($row['participant_id']);
-                $ownername = User::model()->getName($row['owner_uid']); //for conversion of uid to human readable names
-                $sortablearray[$i] = array($row['participant_id'], $row['can_edit'], $row['firstname'], $row['lastname'], $row['email'], $row['blacklisted'], $surveycount, $row['language'], $ownername[0]['full_name']);
-                foreach ($attid as $iAttributeId)
-                {
-                    $answer = ParticipantAttributeNames::model()->getAttributeValue($row['participant_id'], $iAttributeId['attribute_id']);
-                    if (isset($answer['value']))
-                    {
-                        array_push($sortablearray[$i], $answer['value']);
-                    }
-                    else
-                    {
-                        array_push($sortablearray[$i], "");
-                    }
+            $aData->records = Participants::model()->getParticipantsOwnerCount($iUserID);
+        }
+        $records = Participants::model()->getParticipants($page, $limit,$attid, $order, $iUserID);
+        $aData->total = ceil($aData->records / $limit);
+        $aRowToAdd=array();
+        foreach ($records as $key => $row)
+        {            
+            $surveycount = Participants::model()->getSurveyCount($row['participant_id']);
+            if (array_key_exists('can_edit', $row)) {
+                $sCanEdit = $row['can_edit'];
+                if (is_null($sCanEdit)) {
+                    $sCanEdit = 'true';
                 }
-                $i++;
+            } else {
+                // Super admin
+                $sCanEdit = "true";
             }
-
-            $indexsort = array_search(Yii::app()->request->getPost('sidx'), $participantfields);
-        	if(!empty($sortablearray)) {
-        		$sortedarray = subval_sort($sortablearray, $indexsort, Yii::app()->request->getPost('sord'));
-	            $i = 0;
-	            $count = count($sortedarray[0]);
-	            foreach ($sortedarray as $key => $value)
-	            {
-	                $aData->rows[$i]['id'] = $value[0];
-	                $aData->rows[$i]['cell'] = array();
-	                for ($j = 0; $j < $count; $j++)
-	                {
-	                    array_push($aData->rows[$i]['cell'], $value[$j]);
-	                }
-	                $i++;
-	            }
-        	}
+            $aRowToAdd = array($row['participant_id'], $sCanEdit, $row['firstname'], $row['lastname'], $row['email'], $row['blacklisted'], $surveycount, $row['language'], $row['ownername']);
+            unset($row['participant_id'], $row['firstname'], $row['lastname'], $row['email'], $row['blacklisted'], $row['language'],$row['ownername'],$row['owner_uid'], $row['can_edit']);
+            foreach($row as $key=>$attvalue)
+            {
+              $aRowToAdd[] = $attvalue;
+            }
+            $aData->rows[]['cell'] = $aRowToAdd;
         }
+
         echo ls_json_encode($aData);
     }
 
diff --git a/application/models/Participants.php b/application/models/Participants.php
index 97e62361f9c..9128a92feb7 100644
--- a/application/models/Participants.php
+++ b/application/models/Participants.php
@@ -186,34 +186,46 @@ function getParticipantsOwner($userid)
 
     function getParticipantsOwnerCount($userid)
     {
-
-        return count(Yii::app()->db->createCommand()
-                               ->select('{{participants}}.*,{{participant_shares}}.can_edit')
-                               ->from('{{participants}}')
-                               ->leftJoin('{{participant_shares}}', '{{participants}}.participant_id={{participant_shares}}.participant_id')
-                               ->where('owner_uid = :userid1 OR share_uid = :userid2')
-                               ->bindParam(":userid1", $userid, PDO::PARAM_INT)
-                               ->bindParam(":userid2", $userid, PDO::PARAM_INT)
-                               ->queryAll());
+        $command = Yii::app()->db->createCommand()
+                        ->select('count(*)')
+                        ->from('{{participants}} p')
+                        ->leftJoin('{{participant_shares}} ps', 'ps.participant_id = p.participant_id')
+                        ->where('p.owner_uid = :userid1 OR ps.share_uid = :userid2')
+                        ->bindParam(":userid1", $userid, PDO::PARAM_INT)
+                        ->bindParam(":userid2", $userid, PDO::PARAM_INT);
+        return $command->queryScalar();
     }
 
+    /**
+     * Get the number of participants, no restrictions
+     * 
+     * @return int
+     */
+    function getParticipantsCountWithoutLimit()
+    {
+        return Participants::model()->count();    
+    }
+    
     function getParticipantsWithoutLimit()
     {
         return Yii::app()->db->createCommand()->select('*')->from('{{participants}}')->queryAll();
     }
 
-    /*
+    /**
      * This function combines the shared participant and the central participant
      * table and searches for any reference of owner id in the combined record
      * of the two tables
-     */
-
+     * 
+     * @param  int $userid The id of the owner
+     * @return int The number of participants owned by $userid who are shared
+     */  
     function getParticipantsSharedCount($userid)
     {
-        return count(Yii::app()->db->createCommand()->select('{{participants}}.*, {{participant_shares}}.*')->from('{{participants}}')->join('{{participant_shares}}', '{{participant_shares}}.participant_id = {{participants}}.participant_id')->where('owner_uid = :userid')->bindParam(":userid", $userid, PDO::PARAM_INT)->queryAll());
+        $count = Yii::app()->db->createCommand()->select('count(*)')->from('{{participants}}')->join('{{participant_shares}}', '{{participant_shares}}.participant_id = {{participants}}.participant_id')->where('owner_uid = :userid')->bindParam(":userid", $userid, PDO::PARAM_INT)->queryScalar();
+        return $count;
     }
 
-    function getParticipants($page, $limit,$attid)
+    function getParticipants($page, $limit,$attid, $order = null, $userid = null)
     {
         $start = $limit * $page - $limit;
         $selectValue = array();
@@ -227,11 +239,28 @@ function getParticipants($page, $limit,$attid)
             array_push($selectValue,"attribute".$attid.".value as a".$attid);
             array_push($joinValue,"LEFT JOIN {{participant_attribute}} attribute".$attid." ON attribute".$attid.".participant_id=p.participant_id AND attribute".$attid.".attribute_id=".$attid);
         }
+        if (!is_null($userid)) {
+            // We are not superadmin so we need to limit to our own or shared with us
+            $selectValue[] = '{{participant_shares}}.can_edit';
+            $joinValue[]   = 'LEFT JOIN {{participant_shares}} ON p.participant_id={{participant_shares}}.participant_id';
+            $where = 'p.owner_uid = :userid1 OR {{participant_shares}}.share_uid = :userid2';
+        }
+        
         $data = Yii::app()->db->createCommand()
               ->select($selectValue)
               ->from('{{participants}} p')
               ->limit($limit, $start);
         $data->setJoin($joinValue);
+        
+        if (!empty($order)) {
+            $data->setOrder($order);
+        }
+        
+        if (!is_null($userid)) {
+            $data->setWhere($where);
+            $data->bindParam(":userid1", $userid, PDO::PARAM_INT)
+                 ->bindParam(":userid2", $userid, PDO::PARAM_INT);
+        }
         $allData = $data->queryAll();
         return $allData;
 
@@ -348,95 +377,6 @@ function deleteParticipantTokenAnswer($rows)
 		}
 	}
 
-    /*
-     * Builds a select query for searching through participants limited to searches with only one line (no more than 3 entries in condition array)
-     *
-     * Deprecated, use "getParticipantsSearchMultiple()" instead.
-     *
-     * */
-    function getParticipantsSearch($condition, $page, $limit)
-    {
-        $start = $limit * $page - $limit;
-        $command = new CDbCriteria;
-        $command->condition = '';
-        $lang = Yii::app()->session['adminlang'];
-        $start = $limit * $page - $limit;
-        if(is_numeric($condition[2])) $condition[2]=intval($condition[2]);
-        switch($condition[1])
-        {
-            case 'equal':
-                $operator="=";
-                break;
-            case 'contains':
-                $operator="LIKE";
-                $condition[2]="%".$condition[2]."%";
-                break;
-            case 'notequal':
-                $operator="!=";
-                break;
-            case 'notcontains':
-                $operator="NOT LIKE";
-                $condition[2]="%".$condition[2]."%";
-                break;
-            case 'greaterthan':
-                $operator=">";
-                break;
-            case 'lessthan':
-                $operator="<";
-        }
-        if($condition[0]=="survey")
-        {
-            $lang = Yii::app()->session['adminlang'];
-            $command->addCondition('participant_id IN (SELECT distinct {{survey_links}}.participant_id FROM {{survey_links}}, {{surveys_languagesettings}} WHERE {{survey_links}}.survey_id = {{surveys_languagesettings}}.surveyls_survey_id AND {{surveys_languagesettings}}.surveyls_language=:lang AND ({{surveys_languagesettings}}.surveyls_title '.$operator.' :param1 OR {{survey_links}}.survey_id '.$operator.' :param2))');
-            $command->params=array(':lang'=>$lang, ':param1'=>$condition[2] , ':param2'=>$condition[2]);
-        }
-        elseif($condition[0]=="owner_name")
-        {
-            $userid = Yii::app()->db->createCommand()
-                                ->select('uid')
-                                ->where('full_name '.$operator.' :condition_2')
-                                ->from('{{users}}')
-                                ->bindParam("condition_2", $condition[2], PDO::PARAM_STR)
-                                ->queryAll();
-            $uid = $userid[0];
-            $command->addCondition('owner_uid = :uid');
-            $command->params=array(':uid'=>$uid['uid']);
-        }
-        elseif (is_numeric($condition[0])) //Searching for an attribute
-        {
-            $command->addCondition('participant_id IN (SELECT distinct {{participant_attribute}}.participant_id FROM {{participant_attribute}} WHERE {{participant_attribute}}.attribute_id = :condition_0 AND {{participant_attribute}}.value '.$operator.' :condition_2)');
-            $command->params=array(':condition_0'=>$condition[0], ':condition_2'=>$condition[2]);
-        }
-        else
-        {
-            $command->addCondition($condition[0] . ' '.$operator.' :condition_2');
-            $command->params=array(':condition_2'=>$condition[2]);
-        }
-
-        if ($page == 0 && $limit == 0)
-        {
-            $arr = Participants::model()->findAll($command);
-            $data = array();
-            foreach ($arr as $t)
-            {
-                $data[$t->participant_id] = $t->attributes;
-            }
-        }
-        else
-        {
-            $command->limit = $limit;
-            $command->offset = $start;
-            $arr = Participants::model()->findAll($command);
-            $data = array();
-            foreach ($arr as $t)
-            {
-                $data[$t->participant_id] = $t->attributes;
-            }
-        }
-
-        return $data;
-    }
-
     /*
      * Function builds a select query for searches through participants using the $condition field passed
      * which is in the format "firstfield||sqloperator||value||booleanoperator||secondfield||sqloperator||value||booleanoperator||etc||etc||etc"
@@ -486,7 +426,7 @@ function getParticipantsSearchMultiple($condition, $page, $limit)
         }
 
         $con = count($condition);
-        while ($i < $con)
+        while ($i < $con && $con > 2)
         {
             if ($i < 3) //Special set just for the first query/condition
             {
@@ -659,9 +599,9 @@ function getParticipantsSearchMultiple($condition, $page, $limit)
     function is_owner($participant_id)
     {
         $userid = Yii::app()->session['loginID'];
-        $is_owner = Yii::app()->db->createCommand()->select('participant_id')->where('participant_id = :participant_id AND owner_uid = :userid')->from('{{participants}}')->bindParam(":participant_id", $participant_id, PDO::PARAM_STR)->bindParam(":userid", $userid, PDO::PARAM_INT)->queryAll();
-        $is_shared = Yii::app()->db->createCommand()->select('participant_id')->where('participant_id = :participant_id AND share_uid = :userid')->from('{{participant_shares}}')->bindParam(":participant_id", $participant_id, PDO::PARAM_STR)->bindParam(":userid", $userid, PDO::PARAM_INT)->queryAll();
-        if (count($is_shared) || count($is_owner))
+        $is_owner = Yii::app()->db->createCommand()->select('count(*)')->where('participant_id = :participant_id AND owner_uid = :userid')->from('{{participants}}')->bindParam(":participant_id", $participant_id, PDO::PARAM_STR)->bindParam(":userid", $userid, PDO::PARAM_INT)->queryScalar();
+        $is_shared = Yii::app()->db->createCommand()->select('count(*)')->where('participant_id = :participant_id AND share_uid = :userid')->from('{{participant_shares}}')->bindParam(":participant_id", $participant_id, PDO::PARAM_STR)->bindParam(":userid", $userid, PDO::PARAM_INT)->queryScalar();
+        if ($is_shared > 0 || $is_owner > 0)
         {
             return true;
         }
diff --git a/scripts/admin/displayParticipant.js b/scripts/admin/displayParticipant.js
index f422081a330..d028429aa72 100644
--- a/scripts/admin/displayParticipant.js
+++ b/scripts/admin/displayParticipant.js
@@ -154,7 +154,7 @@ $(document).ready(function() {
             if(can_edit == 'false') {
                 var dialog_buttons={};
                 dialog_buttons[okBtn]=function() {
-                    $( this ).dialog( clostTxt );
+                    $( this ).dialog( "close" );
                 };
                 /* End of building array for button functions */
                 $('#notauthorised').dialog({
@@ -711,7 +711,7 @@ $(document).ready(function() {
         if(can_edit == 'false') {
             var dialog_buttons={};
             dialog_buttons[okBtn]=function(){
-                $( this ).dialog( closeTxt );
+                $( this ).dialog( "close" );
             };
             /* End of building array for button functions */
             $('#notauthorised').dialog({
