diff --git a/reviewboard/htdocs/media/rb/css/reviews.css b/reviewboard/htdocs/media/rb/css/reviews.css
index 839cebafe..98909399a 100644
--- a/reviewboard/htdocs/media/rb/css/reviews.css
+++ b/reviewboard/htdocs/media/rb/css/reviews.css
@@ -933,46 +933,50 @@
   border: 1px solid #808080;
   overflow: hidden;
   position: absolute;
   width: 100%;
   z-index: 9050;
 }
 
 /* Footer */
 .ui-autocomplete-footer {
   background: #ECECEC;
   border-top: 1px #C0C0C0 solid;
   padding: 3px 6px;
 }
 
 /* Results list */
-.ui-autocomplete-results ul
-{
+.ui-autocomplete-results ul {
   margin: 0;
   padding: 0;
   width: 100%;
 }
 
 /* Result item */
-.ui-autocomplete-results li
-{
+.ui-autocomplete-results li {
   cursor: pointer;
   margin: 0;
   padding: 3px 6px;
   white-space: nowrap;
 }
 
+.ui-autocomplete-results li span {
+  margin-left: 6px;
+  position: absolute;
+  right: 6px;
+}
+
 /* Pre-highlighted result item */
 /*
 .ui-autocomplete-results .ui-autocomplete-over
 {
   background: #9cc2eb;
 }
 */
 
 /* Highlighted result item */
 .ui-autocomplete-results .ui-autocomplete-over
 {
   background: #71a5db;
   color: #FFF;
 }
 
diff --git a/reviewboard/htdocs/media/rb/js/reviews.js b/reviewboard/htdocs/media/rb/js/reviews.js
index db2efa556..9c0fc9877 100644
--- a/reviewboard/htdocs/media/rb/js/reviews.js
+++ b/reviewboard/htdocs/media/rb/js/reviews.js
@@ -267,62 +267,76 @@ function setDraftField(field, value) {
                 }
             }
         },
         error: function() {
             gPublishing = false;
         }
     });
 }
 
 
 /*
  * An autocomplete() wrapper that converts our autocomplete data into the
  * format expected by jQuery.ui.autocomplete. It also adds some additional
  * explanatory text to the bottom of the autocomplete list.
  *
- * @param {string} fieldName  The field name ("groups" or "people").
- * @param {string} nameKey    The key containing the name in the result data.
+ * options has the following fields:
+ *
+ *    fieldName   - The field name ("groups" or "people").
+ *    nameKey     - The key containing the name in the result data.
+ *    descKey     - The key containing the description in the result
+ *                  data. This is optional.
+ *    extraParams - Extra parameters to send in the query. This is optional.
+ *
+ * @param {object} options    The options, as listed above.
  *
  * @return {jQuery} This jQuery set.
  */
-$.fn.reviewsAutoComplete = function(fieldName, nameKey) {
+$.fn.reviewsAutoComplete = function(options) {
     return this.each(function() {
         $(this)
             .autocomplete({
                 formatItem: function(data) {
-                    return data[nameKey];
+                    var s = data[options.nameKey];
+
+                    if (options.descKey) {
+                        s += " <span>(" + data[options.descKey] + ")</span>";
+                    }
+
+                    return s;
                 },
                 matchCase: false,
                 multiple: true,
                 parse: function(data) {
                     var jsonData = eval("(" + data + ")");
-                    var items = jsonData[fieldName];
+                    var items = jsonData[options.fieldName];
                     var parsed = [];
 
                     for (var i = 0; i < items.length; i++) {
                         var value = items[i];
 
                         parsed.push({
                             data: value,
-                            value: value[nameKey],
-                            result: value[nameKey]
+                            value: value[options.nameKey],
+                            result: value[options.nameKey]
                         });
                     }
 
                     return parsed;
                 },
-                url: SITE_ROOT + "api/json/" + fieldName + "/"
+                url: SITE_ROOT + "api/json/" + options.fieldName + "/",
+                extraParams: options.extraParams
             })
             .bind("autocompleteshow", function() {
                 /*
                  * Add the footer to the bottom of the results pane the
                  * first time it's created.
                  */
                 var resultsPane = $(".ui-autocomplete-results");
 
                 if ($(".ui-autocomplete-footer", resultsPane).length == 0) {
                     $("<div/>")
                         .addClass("ui-autocomplete-footer")
                         .text("Press Tab to auto-complete.")
                         .appendTo(resultsPane);
                 }
             });
@@ -1637,37 +1651,51 @@ $(document).ready(function() {
 
             var description = $("#description");
             var testing_done = $("#testing_done");
 
             if (description.length > 0) {
                 description.html(linkifyText(description.text()));
                 testing_done.html(linkifyText(testing_done.text()));
             }
 
             var targetGroupsEl = $("#target_groups");
             var targetPeopleEl = $("#target_people");
 
             if (targetGroupsEl.length > 0) {
                 targetGroupsEl
                     .inlineEditor("field")
-                    .reviewsAutoComplete("groups", "name");
+                    .reviewsAutoComplete({
+                        fieldName: "groups",
+                        nameKey: "name",
+                        descKey: "display_name",
+                        extraParams: {
+                            displayname: 1
+                        }
+                    });
             }
 
             if (targetPeopleEl.length > 0) {
                 targetPeopleEl
                     .inlineEditor("field")
-                    .reviewsAutoComplete("users", "username");
+                    .reviewsAutoComplete({
+                        fieldName: "users",
+                        nameKey: "username",
+                        descKey: "fullname",
+                        extraParams: {
+                            fullname: 1
+                        }
+                    });
             }
 
             if (window.google && google.gears) {
                 initScreenshotDnD();
             }
         }
 
         $("pre.reviewtext").each(function() {
             $(this).html(linkifyText($(this).text()));
         });
     }
 
     loadDiffFragments("diff_fragments", "comment_container");
 });
 
diff --git a/reviewboard/webapi/json.py b/reviewboard/webapi/json.py
index 263ab8328..3d985b8ec 100644
--- a/reviewboard/webapi/json.py
+++ b/reviewboard/webapi/json.py
@@ -280,58 +280,75 @@ def repository_info(request, repository_id):
         return WebAPIResponse(request, {
             'info': repository.get_scmtool().get_repository_info()
         })
     except NotImplementedError:
         return WebAPIResponseError(request, REPO_NOT_IMPLEMENTED)
     except:
         return WebAPIResponseError(request, REPO_INFO_ERROR)
 
 @webapi_check_login_required
 def user_list(request):
     """
     Returns a list of all users.
 
     If the q parameter is passed, users with a username beginning with
     the query value will be returned.
+
+    If the fullname parameter is passed along with the q parameter, then full
+    names will be searched as well.
     """
     # XXX Support "query" for backwards-compatibility until after 1.0.
     query = request.GET.get('q', request.GET.get('query', None))
 
-    if not query:
-        u = User.objects.filter(is_active=True)
-    else:
-        u = User.objects.filter(is_active=True, username__istartswith=query)
+    u = User.objects.filter(is_active=True)
+
+    if query:
+        q = Q(username__istartswith=query)
+
+        if request.GET.get('fullname', None):
+            q = q | (Q(first_name__istartswith=query) |
+                     Q(last_name__istartswith=query))
+
+        u = u.filter(q)
 
     return WebAPIResponse(request, {
         'users': u,
     })
 
 @webapi_check_login_required
 def group_list(request):
     """
     Returns a list of all review groups.
 
     If the q parameter is passed, groups with a name beginning with
     the query value will be returned.
+
+    If the displayname parameter is passed along with the q parameter, then
+    full names will be searched as well.
     """
     # XXX Support "query" for backwards-compatibility until after 1.0.
     query = request.GET.get('q', request.GET.get('query', None))
 
-    if not query:
-        u = Group.objects.all()
-    else:
-        u = Group.objects.filter(name__istartswith=query)
+    u = Group.objects.all()
+
+    if query:
+        q = Q(name__istartswith=query)
+
+        if request.GET.get('displayname', None):
+            q = q | Q(display_name__istartswith=query)
+
+        u = u.filter(q)
 
     return WebAPIResponse(request, {
         'groups': u,
     })
 
 @webapi_check_login_required
 def users_in_group(request, group_name):
     """
     Returns a list of users in a group.
     """
     try:
         g = Group.objects.get(name=group_name)
     except Group.DoesNotExist:
         return WebAPIResponseError(request, DOES_NOT_EXIST)
 
