diff --git a/kdcproxy/__init__.py b/kdcproxy/__init__.py
index f88c9cd..5c61fb9 100644
--- a/kdcproxy/__init__.py
+++ b/kdcproxy/__init__.py
@@ -124,31 +124,35 @@ class Application:
         if addr[2] not in (socket.IPPROTO_TCP, socket.IPPROTO_UDP):
             return False
 
         return True
 
     def __call__(self, env, start_response):
         try:
             # Validate the method
             method = env["REQUEST_METHOD"].upper()
             if method != "POST":
                 raise HTTPException(405, "Method not allowed (%s)." % method,
                                     ("Allow", "POST"))
 
             # Parse the request
             try:
-                pr = codec.decode(env["wsgi.input"].read())
+                length = int(env["CONTENT_LENGTH"])
+            except AttributeError:
+                length = -1
+            try:
+                pr = codec.decode(env["wsgi.input"].read(length))
             except codec.ParsingError as e:
                 raise HTTPException(400, e.message)
 
             # Find the remote proxy
             servers = self.__resolver.lookup(pr.realm,
                                   isinstance(pr, codec.KPASSWDProxyRequest))
             if not servers:
                 raise HTTPException(503, "Can't find remote (%s)." % pr)
 
             # Contact the remote server
             reply = None
             wsocks = []
             rsocks = []
             for server in map(urlparse.urlparse, servers):
                 # Enforce valid, supported URIs
