diff --git a/image.jpg b/image.jpg
new file mode 100644
index 0000000..7f1ba44
Binary files /dev/null and b/image.jpg differ
diff --git a/index.js b/index.js
new file mode 100644
index 0000000..63338a3
--- /dev/null
+++ b/index.js
@@ -0,0 +1,50 @@
+var exec = require('child_process').exec, child;
+
+module.exports = function(src, opts, cb) {
+  if (!cb) {
+    var cb = opts;
+    var opts = {};
+  }
+
+  var cmd = module.exports.cmd(src, opts);
+  opts.timeout = opts.timeout || 5000;
+
+  exec(cmd, opts, function(e, stdout, stderr) {
+    if (e) { return cb(e); }
+    if (stderr) { return cb(new Error(stderr)); }
+
+    return cb(null, module.exports.parse(stdout));
+  });
+}
+
+module.exports.cmd = function(src, opts) {
+  var opts = opts || {};
+  var format = [
+    'name=%[name]',
+    'size=%[size]',
+    'format=%m',
+    'colorspace=%[colorspace]',
+    'height=%[height]',
+    'width=%[width]',
+    'orientation=%[orientation]',
+    (opts.exif ? '%[exif:*]' : '')
+  ].join("\n");
+
+  return 'identify -format "' + format + '" ' + src;
+}
+
+module.exports.parse = function(metadata) {
+  var lines = metadata.split('\n'), ret = {}, i;
+
+  for (i = 0; i < lines.length; i++) {
+    if (lines[i]) {
+      lines[i] = lines[i].split('=');
+      ret[lines[i][0]] = lines[i][1];
+    }
+  }
+
+  if (ret.width) { ret.width = parseInt(ret.width, 10); }
+  if (ret.height) { ret.height = parseInt(ret.height, 10); }
+
+  return ret;
+}
diff --git a/package.json b/package.json
new file mode 100644
index 0000000..dfbc294
--- /dev/null
+++ b/package.json
@@ -0,0 +1,29 @@
+{
+  "name": "im-metadata",
+  "version": "0.0.0",
+  "description": "Retrieve image metadata using ImageMagick's identify command",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha test.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git@github.com:Turistforeningen/node-im-identify.git"
+  },
+  "keywords": [
+    "ImageMagick",
+    "identify",
+    "image",
+    "metadata",
+    "exif"
+  ],
+  "author": "Hans Kristian Flaatten <hans.kristian.flaatten@turistforeningen.no>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/Turistforeningen/node-im-identify/issues"
+  },
+  "homepage": "https://github.com/Turistforeningen/node-im-identify",
+  "devDependencies": {
+    "mocha": "^2.2.4"
+  }
+}
diff --git a/test.js b/test.js
new file mode 100644
index 0000000..02bfe1c
--- /dev/null
+++ b/test.js
@@ -0,0 +1,60 @@
+var assert = require('assert');
+var metadata = require('.');
+
+describe('metadata.cmd()', function() {
+  it('returns command without exif data', function() {
+    var cmd = 'identify -format "name=%[name]\nsize=%[size]\nformat=%m\n'
+            + 'colorspace=%[colorspace]\nheight=%[height]\nwidth=%[width]\n'
+            + 'orientation=%[orientation]\n" /foo/bar/baz';
+
+    assert.equal(metadata.cmd('/foo/bar/baz'), cmd);
+  });
+
+  it('returns command with exif data', function() {
+    var cmd = 'identify -format "name=%[name]\nsize=%[size]\nformat=%m\n'
+            + 'colorspace=%[colorspace]\nheight=%[height]\nwidth=%[width]\n'
+            + 'orientation=%[orientation]\n%[exif:*]" /foo/bar/baz';
+
+    assert.equal(metadata.cmd('/foo/bar/baz', {exif: true}), cmd);
+  });
+});
+
+describe('metadata.parse()', function() {
+  it('returns object for metadata string', function() {
+    assert.deepEqual(metadata.parse('foo=bar\nbar=foo'), {
+      foo: 'bar',
+      bar: 'foo'
+    });
+  });
+
+  it('skips empty lines', function() {
+    assert.deepEqual(metadata.parse('foo=bar\n\nbar=foo\n\n'), {
+      foo: 'bar',
+      bar: 'foo'
+    });
+  });
+
+  it('returns object for single value', function() {
+    assert.deepEqual(metadata.parse('foo=bar'), {
+      foo: 'bar'
+    });
+  });
+});
+
+describe('metadata()', function() {
+  it('returns metadata for image', function(done) {
+    metadata('./image.jpg', { exif: true }, function(err, data) {
+      assert.ifError(err);
+
+      assert.equal(data.name, '');
+      assert.equal(data.size, '4.296MB');
+      assert.equal(data.format, 'JPEG');
+      assert.equal(data.colorspace, 'sRGB');
+      assert.equal(data.height, 3456);
+      assert.equal(data.width, 5184);
+      assert.equal(data.orientation, 'TopLeft');
+
+      done()
+    });
+  });
+});
