diff --git a/lib/expression/node/AccessorNode.js b/lib/expression/node/AccessorNode.js
index d8696b5e6..9beee2edd 100644
--- a/lib/expression/node/AccessorNode.js
+++ b/lib/expression/node/AccessorNode.js
@@ -1,24 +1,25 @@
 'use strict';
 
 var stringify = require('../../utils/string').stringify;
 var getSafeProperty = require('../../utils/customs').getSafeProperty;
 
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
+  var IndexNode = load(require('./IndexNode'));
   var access = load(require('./utils/access'));
 
   /**
    * @constructor AccessorNode
    * @extends {Node}
    * Access an object property or get a matrix subset
    *
    * @param {Node} object                 The object from which to retrieve
    *                                      a property or subset.
    * @param {IndexNode} index             IndexNode containing ranges
    */
   function AccessorNode(object, index) {
     if (!(this instanceof AccessorNode)) {
       throw new SyntaxError('Constructor must be called with the new operator');
     }
diff --git a/lib/expression/node/ArrayNode.js b/lib/expression/node/ArrayNode.js
index 23ee84a53..568ad58ae 100644
--- a/lib/expression/node/ArrayNode.js
+++ b/lib/expression/node/ArrayNode.js
@@ -1,17 +1,20 @@
 'use strict';
 
+var map = require('../../utils/array').map;
+var join = require('../../utils/array').join;
+
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
 
   /**
    * @constructor ArrayNode
    * @extends {Node}
    * Holds an 1-dimensional array with items
    * @param {Node[]} [items]   1 dimensional array with items
    */
   function ArrayNode(items) {
     if (!(this instanceof ArrayNode)) {
       throw new SyntaxError('Constructor must be called with the new operator');
     }
@@ -40,36 +43,36 @@ function factory (type, config, load, typed) {
   /**
    * Compile the node to javascript code
    * @param {ArrayNode} node  Node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @private
    */
   function compileArrayNode(node, defs, args) {
     var asMatrix = (defs.math.config().matrix !== 'Array');
 
-    var items = node.items.map(function (item) {
+    var items = map(node.items, function (item) {
       return compile(item, defs, args);
     });
 
     return (asMatrix ? 'math.matrix([' : '[') +
-        items.join(',') +
+        join(items, ',') +
         (asMatrix ? '])' : ']');
   }
 
   // register the compile function
   register(ArrayNode.prototype.type, compileArrayNode);
 
       /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   ArrayNode.prototype.forEach = function (callback) {
     for (var i = 0; i < this.items.length; i++) {
       var node = this.items[i];
       callback(node, 'items[' + i + ']', this);
     }
diff --git a/lib/expression/node/BlockNode.js b/lib/expression/node/BlockNode.js
index 75016de43..c7b8756c8 100644
--- a/lib/expression/node/BlockNode.js
+++ b/lib/expression/node/BlockNode.js
@@ -1,17 +1,20 @@
 'use strict';
 
+var map = require('../../utils/array').map;
+var join = require('../../utils/array').join;
+
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
   var ResultSet = load(require('../../type/resultset/ResultSet'));
 
   /**
    * @constructor BlockNode
    * @extends {Node}
    * Holds a set with blocks
    * @param {Array.<{node: Node} | {node: Node, visible: boolean}>} blocks
    *            An array with blocks, where a block is constructed as an Object
    *            with properties block, which is a Node, and visible, which is
    *            a boolean. The property visible is optional and is true by default
    */
@@ -45,43 +48,43 @@ function factory (type, config, load, typed) {
   /**
    * Compile the node to javascript code
    * @param {BlockNode} node  The node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} js
    * @private
    */
   function compileBlockNode (node, defs, args) {
     defs.ResultSet = ResultSet;
-    var blocks = node.blocks.map(function (param) {
+    var blocks = map(node.blocks, function (param) {
       var js = compile(param.node, defs, args);
       if (param.visible) {
         return 'results.push(' + js + ');';
       }
       else {
         return js + ';';
       }
     });
 
     return '(function () {' +
         'var results = [];' +
-        blocks.join('') +
+        join(blocks, '') +
         'return new ResultSet(results);' +
         '})()';
   }
 
   // register the compile function
   register(BlockNode.prototype.type, compileBlockNode);
 
   /**
    * Execute a callback for each of the child blocks of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   BlockNode.prototype.forEach = function (callback) {
     for (var i = 0; i < this.blocks.length; i++) {
       callback(this.blocks[i].node, 'blocks[' + i + '].node', this);
     }
diff --git a/lib/expression/node/ConstantNode.js b/lib/expression/node/ConstantNode.js
index b90d28408..c10a3bf81 100644
--- a/lib/expression/node/ConstantNode.js
+++ b/lib/expression/node/ConstantNode.js
@@ -117,33 +117,39 @@ function factory (type, config, load, typed) {
         // prevent invalid values
         return String(node.value) === 'true' ? 'true' : 'false';
 
       case 'undefined':
         return 'undefined';
 
       case 'null':
         return 'null';
 
       default:
         // TODO: move this error to the constructor?
         throw new TypeError('Unsupported type of constant "' + node.valueType + '"');
     }
   }
 
+  /**
+   * Test whether value is a string containing a numeric value
+   * @param {String} value
+   * @return {boolean} Returns true when ok
+   */
   function validateNumericValue (value) {
     // The following regexp is relatively permissive
-    if (!/^[\-+]?((\d+\.?\d*)|(\d*\.?\d+))([eE][+\-]?\d+)?$/.test(value)) {
+    if (typeof value !== 'string' ||
+        !/^[\-+]?((\d+\.?\d*)|(\d*\.?\d+))([eE][+\-]?\d+)?$/.test(value)) {
       throw new Error('Invalid numeric value "' + value + '"');
     }
   }
 
   // register the compile function
   register(ConstantNode.prototype.type, compileConstantNode);
 
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   ConstantNode.prototype.forEach = function (callback) {
     // nothing to do, we don't have childs
   };
 
diff --git a/lib/expression/node/FunctionAssignmentNode.js b/lib/expression/node/FunctionAssignmentNode.js
index e003382d4..15cd03d41 100644
--- a/lib/expression/node/FunctionAssignmentNode.js
+++ b/lib/expression/node/FunctionAssignmentNode.js
@@ -1,19 +1,21 @@
 'use strict';
 
 var keywords = require('../keywords');
 var stringify = require('../../utils/string').stringify;
+var forEach = require('../../utils/array').forEach;
+var join = require('../../utils/array').join;
 var latex = require('../../utils/latex');
 var operators = require('../operators');
 var setSafeProperty = require('../../utils/customs').setSafeProperty;
 
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
 
   /**
    * @constructor FunctionAssignmentNode
    * @extends {Node}
    * Function assignment
    *
    * @param {string} name           Function name
@@ -55,48 +57,51 @@ function factory (type, config, load, typed) {
    * @param {FunctionAssignmentNode} node The node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} js
    * @private
    */
   function compileFunctionAssignmentNode(node, defs, args) {
     defs.typed = typed;
     defs.setSafeProperty = setSafeProperty;
 
+    // validate params
+    // FIXME: rename parameters to safe, internal names
+
     // we extend the original args and add the args to the child object
     var childArgs = Object.create(args);
-    node.params.forEach(function (variable) {
+    forEach(node.params, function (variable) {
       childArgs[variable] = true;
     });
 
     // compile the function expression with the child args
     var jsExpr = compile(node.expr, defs, childArgs);
     var jsName = stringify(node.name);
 
     return 'setSafeProperty(scope, ' + jsName + ', ' +
         '  (function () {' +
         '    var fn = typed(' + jsName + ', {' +
-        '      ' + stringify(node.types.join(',')) + ': function (' + node.params.join(',') + ') {' +
+        '      ' + stringify(join(node.types, ',')) + ': function (' + join(node.params, ',') + ') {' +
         '        return ' + jsExpr + '' +
         '      }' +
         '    });' +
-        '    fn.syntax = ' + stringify(node.name + '(' + node.params.join(', ') + ')') + ';' +
+        '    fn.syntax = ' + stringify(node.name + '(' + join(node.params, ', ') + ')') + ';' +
         '    return fn;' +
         '  })())';
   }
 
   // register the compile function
   register(FunctionAssignmentNode.prototype.type, compileFunctionAssignmentNode);
 
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   FunctionAssignmentNode.prototype.forEach = function (callback) {
     callback(this.expr, 'expr', this);
   };
 
diff --git a/lib/expression/node/FunctionNode.js b/lib/expression/node/FunctionNode.js
index f6db26f30..af044fa83 100644
--- a/lib/expression/node/FunctionNode.js
+++ b/lib/expression/node/FunctionNode.js
@@ -1,21 +1,25 @@
 'use strict';
 
 var latex = require('../../utils/latex');
 var stringify = require('../../utils/string').stringify;
 var extend = require('../../utils/object').extend;
+var hasOwnProperty = require('../../utils/object').hasOwnProperty;
+var map = require('../../utils/array').map;
+var join = require('../../utils/array').join;
 var validateSafeMethod = require('../../utils/customs').validateSafeMethod;
+var getUniqueArgumentsName = require('./utils/getUniqueArgumentsName');
 
 function factory (type, config, load, typed, math) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
   var SymbolNode = load(require('./SymbolNode'));
 
   /**
    * @constructor FunctionNode
    * @extends {./Node}
    * invoke a list with arguments on a node
    * @param {./Node | string} fn Node resolving with a function on which to invoke
    *                             the arguments, typically a SymboNode or AccessorNode
    * @param {./Node[]} args
    */
@@ -65,81 +69,80 @@ function factory (type, config, load, typed, math) {
    * Compile the node to javascript code
    * @param {FunctionNode} node The node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} js
    * @private
    */
   function compileFunctionNode(node, defs, args) {
     // compile fn and arguments
     var jsFn = compile(node.fn, defs, args);
-    var jsArgs = node.args.map(function (arg) {
+    var jsArgs = map(node.args, function (arg) {
       return compile(arg, defs, args);
     });
     var jsScope = compileScope(defs, args);
     var argsName;
 
     if (node.fn.isSymbolNode) {
       // we can statically determine whether the function has an rawArgs property
       var name = node.fn.name;
       var fn = defs.math[name];
       var isRaw = (typeof fn === 'function') && (fn.rawArgs == true);
 
       if (isRaw) {
         // pass unevaluated parameters (nodes) to the function
-        argsName = node._getUniqueArgumentsName(defs);
+        argsName = getUniqueArgumentsName(defs);
         defs[argsName] = node.args;
 
-        return jsFn + '(' + argsName + ', math, ' + jsScope + ')';
+        return jsFn + '(' + argsName + ', math, ' + jsScope + ')'; // "raw" evaluation
       }
       else {
-        // "regular" evaluation
-        return jsFn + '(' + jsArgs.join(', ') + ')';
+        return jsFn + '(' + join(jsArgs, ', ') + ')';              // "regular" evaluation
       }
     }
     else if (node.fn.isAccessorNode && node.fn.index.isObjectProperty()) {
       // execute the function with the right context: the object of the AccessorNode
-      argsName = node._getUniqueArgumentsName(defs);
+      argsName = getUniqueArgumentsName(defs);
       defs[argsName] = node.args;
       defs.validateSafeMethod = validateSafeMethod
 
       var jsObject = compile(node.fn.object, defs, args);
       var jsProp = stringify(node.fn.index.getObjectProperty());
 
       return '(function () {' +
           'var object = ' + jsObject + ';' +
           'validateSafeMethod(object, ' + jsProp + ');' +
           'return (object[' + jsProp + '] && object[' + jsProp + '].rawArgs) ' +
-          ' ? object[' + jsProp + '](' + argsName + ', math, ' + jsScope + ')' +
-          ' : object[' + jsProp + '](' + jsArgs.join(', ') + ')' +
+          ' ? object[' + jsProp + '](' + argsName + ', math, ' + jsScope + ')' + // "raw" evaluation
+          ' : object[' + jsProp + '](' + join(jsArgs, ', ') + ')' +              // "regular" evaluation
           '})()';
     }
     else { // node.fn.isAccessorNode && !node.fn.index.isObjectProperty()
       // we have to dynamically determine whether the function has a rawArgs property
-      argsName = node._getUniqueArgumentsName(defs);
+      argsName = getUniqueArgumentsName(defs);
       defs[argsName] = node.args;
 
       return '(function () {' +
           'var fn = ' + jsFn + ';' +
           'return (fn && fn.rawArgs) ' +
-          ' ? fn(' + argsName + ', math, ' + jsScope + ')' +
-          ' : fn(' + jsArgs.join(', ') + ')' +
+          ' ? fn(' + argsName + ', math, ' + jsScope + ')' +  // "raw" evaluation
+          ' : fn(' + join(jsArgs, ', ') + ')' +               // "regular" evaluation
           '})()';
     }
   }
 
   // register the compile function
   register(FunctionNode.prototype.type, compileFunctionNode);
 
   /**
    * Merge function arguments into scope before passing to the actual function.
    * This is needed when the function has `rawArgs=true`. In that case we have
    * to pass the `scope` as third argument, including any variables of
    * enclosing functions.
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
@@ -147,59 +150,38 @@ function factory (type, config, load, typed, math) {
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} js
    * @private
    */
   function compileScope (defs, args) {
     var names = Object.keys(args);
     if (names.length === 0) {
       return 'scope';
     }
     else {
       // merge arguments into scope
       defs.extend = extend;
 
-      var jsArgs = names
-          .map(function (arg) {
-            return stringify(arg) + ': ' + arg;
-          })
-          .join(', ');
+      var jsArgs = map(names, function (arg) {
+          return stringify(arg) + ': ' + arg;
+        });
 
-      return 'extend(extend({}, scope), {' + jsArgs + '})';
+      return 'extend(extend({}, scope), {' + join(jsArgs, ', ') + '})';
     }
   }
 
-  /**
-   * Get a unique name for a arguments to store in defs
-   * @param {Object} defs
-   * @return {string} A string like 'args1', 'args2', ...
-   * @private
-   */
-  FunctionNode.prototype._getUniqueArgumentsName = function (defs) {
-    var argsName;
-    var i = 0;
-
-    do {
-      argsName = 'args' + i;
-      i++;
-    }
-    while (argsName in defs);
-
-    return argsName;
-  };
-
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   FunctionNode.prototype.forEach = function (callback) {
     for (var i = 0; i < this.args.length; i++) {
       callback(this.args[i], 'args[' + i + ']', this);
     }
   };
 
   /**
    * Create a new FunctionNode having it's childs be the results of calling
    * the provided callback function for each of the childs of the original node.
    * @param {function(child: Node, path: string, parent: Node): Node} callback
    * @returns {FunctionNode} Returns a transformed copy of the node
@@ -228,31 +210,31 @@ function factory (type, config, load, typed, math) {
   /**
    * Get string representation. (wrapper function)
    * This overrides parts of Node's toString function.
    * If callback is an object containing callbacks, it
    * calls the correct callback for the current node,
    * otherwise it falls back to calling Node's toString
    * function.
    *
    * @param {Object} options
    * @return {string} str
    * @override
    */
   FunctionNode.prototype.toString = function (options) {
     var customString;
     var name = this.fn.toString(options);
-    if (options && (typeof options.handler === 'object') && options.handler.hasOwnProperty(name)) {
+    if (options && (typeof options.handler === 'object') && hasOwnProperty(options.handler, name)) {
       //callback is a map of callback functions
       customString = options.handler[name](this, options);
     }
 
     if (typeof customString !== 'undefined') {
       return customString;
     }
 
     //fall back to Node's toString
     return nodeToString.call(this, options);
   };
 
   /**
    * Get string representation
    * @param {Object} options
@@ -345,31 +327,31 @@ function factory (type, config, load, typed, math) {
   var nodeToTex = FunctionNode.prototype.toTex;
 
   /**
    * Get LaTeX representation. (wrapper function)
    * This overrides parts of Node's toTex function.
    * If callback is an object containing callbacks, it
    * calls the correct callback for the current node,
    * otherwise it falls back to calling Node's toTex
    * function.
    *
    * @param {Object} options
    * @return {string}
    */
   FunctionNode.prototype.toTex = function (options) {
     var customTex;
-    if (options && (typeof options.handler === 'object') && options.handler.hasOwnProperty(this.name)) {
+    if (options && (typeof options.handler === 'object') && hasOwnProperty(options.handler, this.name)) {
       //callback is a map of callback functions
       customTex = options.handler[this.name](this, options);
     }
 
     if (typeof customTex !== 'undefined') {
       return customTex;
     }
 
     //fall back to Node's toTex
     return nodeToTex.call(this, options);
   };
 
   /**
    * Get LaTeX representation
    * @param {Object} options
diff --git a/lib/expression/node/IndexNode.js b/lib/expression/node/IndexNode.js
index fc6f25c33..005db994f 100644
--- a/lib/expression/node/IndexNode.js
+++ b/lib/expression/node/IndexNode.js
@@ -1,17 +1,20 @@
 'use strict';
 
+var map = require('../../utils/array').map;
+var join = require('../../utils/array').join;
+
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
   var RangeNode = load(require('./RangeNode'));
   var SymbolNode = load(require('./SymbolNode'));
 
   var Range = load(require('../../type/matrix/Range'));
 
   var isArray = Array.isArray;
 
   /**
    * @constructor IndexNode
    * @extends Node
    *
@@ -82,31 +85,31 @@ function factory (type, config, load, typed) {
     defs.range = function (start, end, step) {
       return new Range(
           (start && start.isBigNumber === true) ? start.toNumber() : start,
           (end   && end.isBigNumber === true)   ? end.toNumber()   : end,
           (step  && step.isBigNumber === true)  ? step.toNumber()  : step
       );
     };
 
     // TODO: implement support for bignumber (currently bignumbers are silently
     //       reduced to numbers when changing the value to zero-based)
 
     // TODO: Optimization: when the range values are ConstantNodes,
     //       we can beforehand resolve the zero-based value
 
     // optimization for a simple object property
-    var dimensions = node.dimensions.map(function (range, i) {
+    var dimensions = map(node.dimensions, function (range, i) {
       if (range && range.isRangeNode) {
         if (range.needsEnd()) {
           childArgs.end = true;
 
           // resolve end and create range
           return '(function () {' +
               'var end = size[' + i + ']; ' +
               'return range(' +
               compile(range.start, defs, childArgs) + ', ' +
               compile(range.end, defs, childArgs) + ', ' +
               (range.step ? compile(range.step, defs, childArgs) : '1') +
               '); ' +
               '})()';
         }
         else {
@@ -120,31 +123,31 @@ function factory (type, config, load, typed) {
       }
       if (range.isSymbolNode && range.name === 'end') {
         childArgs.end = true;
 
         // resolve the parameter 'end'
         return '(function () {' +
             'var end = size[' + i + ']; ' +
             'return ' + compile(range, defs, childArgs) + '; ' +
             '})()'
       }
       else { // ConstantNode
         return compile(range, defs, childArgs);
       }
     });
 
-    return 'math.index(' + dimensions.join(', ') + ')';
+    return 'math.index(' + join(dimensions, ', ') + ')';
   }
 
   // register the compile function
   register(IndexNode.prototype.type, compileIndexNode);
 
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   IndexNode.prototype.forEach = function (callback) {
     for (var i = 0; i < this.dimensions.length; i++) {
       callback(this.dimensions[i], 'dimensions[' + i + ']', this);
     }
   };
 
diff --git a/lib/expression/node/Node.js b/lib/expression/node/Node.js
index 113a75367..d1b936812 100644
--- a/lib/expression/node/Node.js
+++ b/lib/expression/node/Node.js
@@ -1,19 +1,20 @@
 'use strict';
 
 var keywords = require('../keywords');
 var deepEqual= require('../../utils/object').deepEqual;
+var hasOwnProperty = require('../../utils/object').hasOwnProperty;
 
 function factory (type, config, load, typed, math) {
   var compile = load(require('./compile')).compile;
 
   /**
    * Node
    */
   function Node() {
     if (!(this instanceof Node)) {
       throw new SyntaxError('Constructor must be called with the new operator');
     }
   }
 
   /**
    * Evaluate the node
@@ -337,30 +338,30 @@ function factory (type, config, load, typed, math) {
   /**
    * Get the content of the current Node.
    * @return {Node} node
    **/
   Node.prototype.getContent = function () {
     return this;
   };
 
   /**
    * Validate the symbol names of a scope.
    * Throws an error when the scope contains an illegal symbol.
    * @param {Object} scope
    */
   function _validateScope(scope) {
     for (var symbol in scope) {
-      if (Object.hasOwnProperty.call(scope, symbol)) {
+      if (hasOwnProperty(scope, symbol)) {
         if (symbol in keywords) {
           throw new Error('Scope contains an illegal symbol, "' + symbol + '" is a reserved keyword');
         }
       }
     }
   }
 
   return Node;
 }
 
 exports.name = 'Node';
 exports.path = 'expression.node';
 exports.math = true; // request access to the math namespace as 5th argument of the factory function
 exports.factory = factory;
diff --git a/lib/expression/node/ObjectNode.js b/lib/expression/node/ObjectNode.js
index 64f4db922..ec7901e10 100644
--- a/lib/expression/node/ObjectNode.js
+++ b/lib/expression/node/ObjectNode.js
@@ -1,18 +1,20 @@
 'use strict';
 
 var stringify = require('../../utils/string').stringify;
+var isSafeProperty = require('../../utils/customs').isSafeProperty;
+var hasOwnProperty = require('../../utils/object').hasOwnProperty;
 
 function factory (type, config, load, typed) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node = load(require('./Node'));
 
   /**
    * @constructor ObjectNode
    * @extends {Node}
    * Holds an object with keys/values
    * @param {Object.<string, Node>} [properties]   array with key/value pairs
    */
   function ObjectNode(properties) {
     if (!(this instanceof ObjectNode)) {
       throw new SyntaxError('Constructor must be called with the new operator');
@@ -40,33 +42,35 @@ function factory (type, config, load, typed) {
    * Compile the node to javascript code
    * @param {ObjectNode} node The node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} code
    * @private
    */
   function compileObjectNode(node, defs, args) {
     var entries = [];
     for (var key in node.properties) {
-      // TODO: introduce a helper function hasOwnProperty in object.js
-      if (Object.hasOwnProperty.call(node.properties, key)) {
-        // FIXME: escape the key
+      if (!isSafeProperty(key)) {
+        throw new Error('No access to property "' + key + '"');
+      }
+
+      if (hasOwnProperty(node.properties, key)) {
         entries.push(stringify(key) + ': ' + compile(node.properties[key], defs, args));
       }
     }
     return '{' + entries.join(', ') + '}';
   }
 
   // register the compile function
   register(ObjectNode.prototype.type, compileObjectNode);
 
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   ObjectNode.prototype.forEach = function (callback) {
     for (var key in this.properties) {
diff --git a/lib/expression/node/OperatorNode.js b/lib/expression/node/OperatorNode.js
index d20e8ad60..49748948f 100644
--- a/lib/expression/node/OperatorNode.js
+++ b/lib/expression/node/OperatorNode.js
@@ -1,18 +1,22 @@
 'use strict';
 
 var latex = require('../../utils/latex');
+var map = require('../../utils/array').map;
+var join = require('../../utils/array').join;
+var stringify = require('../../utils/string').stringify;
+var isSafeMethod = require('../../utils/customs').isSafeMethod;
 var operators = require('../operators');
 
 function factory (type, config, load, typed, math) {
   var register = load(require('./compile')).register;
   var compile = load(require('./compile')).compile;
   var Node         = load(require('./Node'));
   var ConstantNode = load(require('./ConstantNode'));
   var SymbolNode   = load(require('./SymbolNode'));
   var FunctionNode = load(require('./FunctionNode'));
 
   /**
    * @constructor OperatorNode
    * @extends {Node}
    * An operator with two arguments, like 2+3
    *
@@ -52,40 +56,45 @@ function factory (type, config, load, typed, math) {
 
   /**
    * Compile the node to javascript code
    * @param {OperatorNode} node The node to be compiled
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} js
    * @private
    */
   function compileOperatorNode(node, defs, args) {
-    if (!defs.math[node.fn]) {
-      throw new Error('Function ' + node.fn + ' missing in provided namespace "math"');
+    // validate fn
+    if (typeof node.fn !== 'string' || !isSafeMethod(defs.math, node.fn)) {
+      if (!defs.math[node.fn]) {
+        throw new Error('Function ' + node.fn + ' missing in provided namespace "math"');
+      }
+      else {
+        throw new Error('No access to function "' + node.fn + '"');
+      }
     }
 
-    var jsArgs = node.args.map(function (arg) {
+    var jsArgs = map(node.args, function (arg) {
       return compile(arg, defs, args);
     });
 
-    // FIXME: escape/validate fn
-    return 'math.' + node.fn + '(' + jsArgs.join(', ') + ')';
+    return 'math[' + stringify(node.fn) + '](' + join(jsArgs, ', ') + ')';
   }
 
   // register the compile function
   register(OperatorNode.prototype.type, compileOperatorNode);
 
   /**
    * Execute a callback for each of the child nodes of this node
    * @param {function(child: Node, path: string, parent: Node)} callback
    */
   OperatorNode.prototype.forEach = function (callback) {
     for (var i = 0; i < this.args.length; i++) {
       callback(this.args[i], 'args[' + i + ']', this);
     }
   };
 
diff --git a/lib/expression/node/compile.js b/lib/expression/node/compile.js
index fee5b7955..94dbbd277 100644
--- a/lib/expression/node/compile.js
+++ b/lib/expression/node/compile.js
@@ -1,20 +1,22 @@
 // the compile functions which compile a Node into JavaScript are not
 // exposed as class methods for security reasons to prevent being able to
 // override them or create fake Nodes. Instead, only compile functions of
 // registered nodes can be executed
 
+var hasOwnProperty = require('../../utils/object').hasOwnProperty;
+
 function factory () {
   // map with node type as key and compile functions as value
   var compileFunctions = {}
 
   /**
    * Register a compile function for a node
    * @param {string} type
    * @param {function} compileFunction
    *                      The compile function, invoked as
    *                      compileFunction(node, defs, args)
    */
   function register(type, compileFunction) {
     if (compileFunctions[type] === undefined) {
       compileFunctions[type] = compileFunction;
     }
@@ -24,36 +26,36 @@ function factory () {
   }
 
   /**
    * Compile a Node into JavaScript
    * @param {Node} node
    * @param {Object} defs     Object which can be used to define functions
    *                          or constants globally available for the compiled
    *                          expression
    * @param {Object} args     Object with local function arguments, the key is
    *                          the name of the argument, and the value is `true`.
    *                          The object may not be mutated, but must be
    *                          extended instead.
    * @return {string} Returns JavaScript code
    */
   function compile (node, defs, args) {
-    if (Object.hasOwnProperty.call(compileFunctions, node.type)) {
+    if (hasOwnProperty(compileFunctions, node.type)) {
       var compileFunction = compileFunctions[node.type];
       return compileFunction(node, defs, args);
     }
     else if (typeof node._compile === 'function' &&
-        !Object.hasOwnProperty.call(node, '_compile')) {
+        !hasOwnProperty(node, '_compile')) {
       // Compatibility for CustomNodes
       // TODO: this is a security risk, change it such that you have to register CustomNodes separately in math.js, like math.expression.node.register(MyCustomNode)
       return node._compile(defs, args);
     }
     else {
       throw new Error('Cannot compile node: unknown type "' + node.type + '"');
     }
   }
 
   return {
     register: register,
     compile: compile
   }
 }
 
diff --git a/lib/expression/node/utils/getUniqueArgumentsName.js b/lib/expression/node/utils/getUniqueArgumentsName.js
new file mode 100644
index 000000000..a08caa880
--- /dev/null
+++ b/lib/expression/node/utils/getUniqueArgumentsName.js
@@ -0,0 +1,20 @@
+/**
+ * Get a unique name for a arguments to store in defs
+ * @param {Object} defs
+ * @return {string} A string like 'args1', 'args2', ...
+ * @private
+ */
+function getUniqueArgumentsName (defs) {
+  var argsName;
+  var i = 0;
+
+  do {
+    argsName = 'args' + i;
+    i++;
+  }
+  while (argsName in defs);
+
+  return argsName;
+}
+
+module.exports = getUniqueArgumentsName;
diff --git a/lib/expression/transform/map.transform.js b/lib/expression/transform/map.transform.js
index f8c6fe885..1e187d11c 100644
--- a/lib/expression/transform/map.transform.js
+++ b/lib/expression/transform/map.transform.js
@@ -1,27 +1,26 @@
 'use strict';
 
 var maxArgumentCount = require('../../utils/function').maxArgumentCount;
 
 /**
  * Attach a transform function to math.map
  * Adds a property transform containing the transform function.
  *
  * This transform creates a one-based index instead of a zero-based index
  */
 function factory (type, config, load, typed) {
-  var map = load(require('../../function/matrix/map'));
   var matrix = load(require('../../type/matrix/function/matrix'));
 
   return typed('max', {
     'Array, function': function (x, callback) {
       return _map(x, callback, x);
     },
 
     'Matrix, function': function (x, callback) {
       return matrix(_map(x.valueOf(), callback, x));
     }
   });
 }
 
 /**
  * Map for a multi dimensional array. One-based indexes
diff --git a/lib/utils/array.js b/lib/utils/array.js
index 6879b66b2..2ae8e65a9 100644
--- a/lib/utils/array.js
+++ b/lib/utils/array.js
@@ -406,21 +406,48 @@ exports.flatten = function(array) {
   }
   var flat = [];
 
   array.forEach(function callback(value) {
     if (Array.isArray(value)) {
       value.forEach(callback);  //traverse through sub-arrays recursively
     }
     else {
       flat.push(value);
     }
   });
 
   return flat;
 };
 
+/**
+ * A safe map
+ * @param {Array} array
+ * @param {function} callback
+ */
+exports.map = function (array, callback) {
+  return Array.prototype.map.call(array, callback);
+}
+
+/**
+ * A safe forEach
+ * @param {Array} array
+ * @param {function} callback
+ */
+exports.forEach = function (array, callback) {
+  Array.prototype.forEach.call(array, callback);
+}
+
+/**
+ * A safe join
+ * @param {Array} array
+ * @param {string} separator
+ */
+exports.join = function (array, separator) {
+  return Array.prototype.join.call(array, separator);
+}
+
 /**
  * Test whether an object is an array
  * @param {*} value
  * @return {boolean} isArray
  */
 exports.isArray = Array.isArray;
diff --git a/lib/utils/customs.js b/lib/utils/customs.js
index 94c855faf..1ed30b529 100644
--- a/lib/utils/customs.js
+++ b/lib/utils/customs.js
@@ -50,30 +50,40 @@ function setSafeProperty (object, prop, value) {
       // override when the property is defined on the object itself.
       // don't allow overriding inherited properties like .constructor or .toString
       if (Object.hasOwnProperty.call(object, prop)) {
         return object[prop] = value;
       }
     }
     else {
       // this is a new property, that's just ok
       return object[prop] = value;
     }
   }
 
   throw new Error('No access to property "' + prop + '"');
 }
 
+/**
+ * Test whether a property is safe to use for an object.
+ * For example .toString and .constructor are not safe
+ * @param {string} prop
+ * @return {boolean} Returns true when safe
+ */
+function isSafeProperty (prop) {
+  return !(prop in {});
+}
+
 /**
  * Validate whether a method is safe.
  * Throws an error when that's not the case.
  * @param {Object} object
  * @param {string} method
  */
 // TODO: merge this function into assign.js?
 function validateSafeMethod (object, method) {
   if (!isSafeMethod(object, method)) {
     throw new Error('No access to method "' + method + '"');
   }
 }
 
 function isSafeMethod (object, method) {
   // TODO: remove this, replace with whitelist
@@ -355,17 +365,18 @@ var safeMethods = {
   typeof: true,
   unaryMinus: true,
   unaryPlus: true,
   unequal: true,
   unit: true,
   usolve: true,
   valueOf: true,
   var: true,
   xgcd: true,
   xor: true,
   zeros: true
 }
 
 exports.getSafeProperty = getSafeProperty;
 exports.setSafeProperty = setSafeProperty;
+exports.isSafeProperty = isSafeProperty;
 exports.validateSafeMethod = validateSafeMethod;
 exports.isSafeMethod = isSafeMethod;
diff --git a/lib/utils/object.js b/lib/utils/object.js
index 842e53887..18ca0a176 100644
--- a/lib/utils/object.js
+++ b/lib/utils/object.js
@@ -224,30 +224,39 @@ exports.traverse = function(object, path) {
 
   if (path) {
     var names = path.split('.');
     for (var i = 0; i < names.length; i++) {
       var name = names[i];
       if (!(name in obj)) {
         obj[name] = {};
       }
       obj = obj[name];
     }
   }
 
   return obj;
 };
 
+/**
+ * A safe hasOwnProperty
+ * @param {Object} object
+ * @param {string} property
+ */
+exports.hasOwnProperty = function (object, property) {
+  return object && Object.hasOwnProperty.call(object, property);
+}
+
 /**
  * Test whether an object is a factory. a factory has fields:
  *
  * - factory: function (type: Object, config: Object, load: function, typed: function [, math: Object])   (required)
  * - name: string (optional)
  * - path: string    A dot separated path (optional)
  * - math: boolean   If true (false by default), the math namespace is passed
  *                   as fifth argument of the factory function
  *
  * @param {*} object
  * @returns {boolean}
  */
 exports.isFactory = function (object) {
   return object && typeof object.factory === 'function';
 };
diff --git a/test/expression/security.test.js b/test/expression/security.test.js
index 9e20ab5cc..8550fbb9b 100644
--- a/test/expression/security.test.js
+++ b/test/expression/security.test.js
@@ -98,30 +98,44 @@ describe('security', function () {
 
   it ('should not allow calling Function via a symbol', function () {
     assert.throws(function () {
       math.eval('O = {}.constructor\n' +
           'd = O.getOwnPropertyDescriptor(O.__proto__, "constructor")\n' +
           'eval("value", d)("console.log(\'hacked...\')")()');
     }, /Error: No access to property "constructor/);
   })
 
   it ('should not allow calling Function via a specially encoded constructor property name', function () {
     assert.throws(function () {
       math.eval('[].map["\\x63onstructor"]("console.log(\'hacked...\')")()');
     }, /Error: Cannot access method "map" as a property/);
   })
 
+  it ('should not allow creating an Object with forbidden properties', function () {
+    assert.throws(function () {
+      math.eval('{hasOwnProperty: 2}');
+    }, /Error: No access to property "hasOwnProperty/);
+
+    assert.throws(function () {
+      math.eval('{constructor: 2}');
+    }, /Error: No access to property "constructor/);
+
+    assert.throws(function () {
+      math.eval('{toString: 2}');
+    }, /Error: No access to property "toString/);
+  })
+
   it ('should not allow calling Object via a an object constructor', function () {
     assert.throws(function () {
             math.eval('{}.constructor.assign(expression.node.AssignmentNode.prototype, ' +
                 '{_compile: "".toString.bind("console.log(\'hacked...\')")})\n' +
                 'eval("a = 2")');
     }, /Error: No access to property "constructor/);
   })
 
   it ('should not allow calling Object via a __defineGetter__', function () {
     assert.throws(function () {
       math.eval('expression.node.AssignmentNode.prototype.__defineGetter__("_compile", ' +
           '{}.valueOf.bind("".toString.bind("console.log(\'hacked...\')"))); eval("a = 2")')
     }, /Error: Undefined symbol expression/);
   })
 
@@ -179,68 +193,68 @@ describe('security', function () {
     }, /Error: No access to property "value/);
   })
 
   it ('should not allow using restricted properties via subset (3)', function () {
     assert.throws(function () {
       math.eval('subset(parse("1"), index("value"), "eval").compile().eval()("console.log(\'hacked...\')")')
     }, /Error: No access to property "value/);
   })
 
   it ('should not allow inserting fake nodes with bad code via node.map or node.transform', function () {
     assert.throws(function () {
       math.eval("badValue = {\"isNode\": true, \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
     }, /Error: Cannot compile node: unknown type "undefined"/);
 
     assert.throws(function () {
-      math.eval("badValue = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"string\", \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
+      math.eval("badValue = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"string\", \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked...\')\")")
     }); // The error message is vague but well...
   })
 
-  it.skip ('should not allow injecting code via an adjusted ConstantNode value', function () {
+  it ('should not allow replacing validateSafeMethod with a local variant', function () {
     assert.throws(function () {
-      // TODO: write unit test
-      }, /Error: No access to property "constructor/);
+      math.eval("eval(\"f(validateSafeMethod)=cos.constructor(\\\"return eval\\\")()\")(eval(\"f(x,y)=0\"))(\"console.log('hacked...')\")")
+    }, /Error: Cannot compile node/);
   })
 
-  it.skip ('should not allow injecting code via an adjusted SymbolNode name', function () {
+  it ('should not allow abusing toString', function () {
     assert.throws(function () {
-      // TODO: write unit test
-      }, /Error: No access to property "constructor/);
+      math.eval("badToString = eval(\"f() = 1\"); badReplace = eval(\"f(a, b) = \\\"eval\\\"\"); badNumber = {toString:badToString, replace:badReplace}; badNode = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"number\", \"value\": badNumber}; x = eval(\"f(child, path, parent) = badNode\", {badNode:badNode}); parse(\"(1)\").map(x).compile().eval()(\"console.log('hacked...')\")")
+    }, /Error: No access to property "toString"/);
   })
 
-  it.skip ('should not allow injecting code via an adjusted FunctionNode name', function () {
+  it ('should not allow creating a bad FunctionAssignmentNode', function () {
     assert.throws(function () {
-      // TODO: write unit test
-      }, /Error: No access to property "constructor/);
+      math.eval("badNode={isNode:true,type:\"FunctionAssignmentNode\",expr:parse(\"1\"),types:{join:eval(\"f(a)=\\\"\\\"\")},params:{\"forEach\":eval(\"f(x)=1\"),\"join\":eval(\"f(x)=\\\"){return eval;}});return fn;})())}});return fn;})());}};//\\\"\")}};parse(\"f()=x\").map(eval(\"f(a,b,c)=badNode\",{\"badNode\":badNode})).compile().eval()()()(\"console.log('hacked...')\")")
+    }, /TypeError: fn is not a function/);
   })
 
-  it.skip ('should not allow injecting code via an adjusted function name', function () {
+  it ('should not allow creating a bad OperatorNode (1)', function () {
     assert.throws(function () {
-        // TODO: write unit test
-      }, /Error: No access to property "constructor/);
+      math.eval("badNode={isNode:true,type:\"FunctionAssignmentNode\",expr:parse(\"1\"),types:{join:eval(\"f(a)=\\\"\\\"\")},params:{\"forEach\":eval(\"f(x)=1\"),\"join\":eval(\"f(x)=\\\"){return eval;}});return fn;})())}});return fn;})());}};//\\\"\")}};parse(\"f()=x\").map(eval(\"f(a,b,c)=badNode\",{\"badNode\":badNode})).compile().eval()()()(\"console.log('hacked...')\")")
+    }, /TypeError: fn is not a function/);
   })
 
-  it.skip ('should not allow injecting code via a property name', function () {
+  it ('should not allow creating a bad OperatorNode (2)', function () {
     assert.throws(function () {
-        // TODO
-      }, /Error: No access to property "constructor/);
+      math.eval("parse(\"(0)\").map(eval(\"f(a,b,c)=d\",{d:{isNode:true,type:\"OperatorNode\",fn:\"__lookupGetter__\",args:{map:eval(\"f(a)=b\",{b:{join:eval(\"f(a)=\\\"1)||eval;}};//\\\"\")}})}}})).compile().eval()(\"console.log('hacked...')\")")
+    }, /TypeError: fn is not a function/);
   })
 
-  it.skip ('should not allow inserting a fake ObjectNode', function () {
+  it ('should not allow creating a bad ConstantNode', function () {
     assert.throws(function () {
-      // TODO: write unit test
-      }, /Error: No access to property "constructor/);
+      math.eval('f(x,y)="eval";g()=3;fakeConstantNode={"isNode": true, "type": "ConstantNode", "valueType": "number", "value": {"replace": f, "toString": g}};injectFakeConstantNode(child,path,parent)=path=="value"?fakeConstantNode:child;parse("a=3").map(injectFakeConstantNode).compile().eval()("console.log(\'hacked...\')")')
+    }, /Error: No access to property "toString"/);
   })
 
   it ('should allow calling functions on math', function () {
     assert.equal(math.eval('sqrt(4)'), 2);
   })
 
   it ('should allow accessing properties on an object', function () {
     assert.deepEqual(math.eval('obj.a', {obj: {a:42}}), 42);
   })
 
   it ('should not allow accessing inherited properties on an object', function () {
     assert.throws(function () {
       math.eval('obj.constructor', {obj: {a:42}});
     }, /Error: No access to property "constructor"/)
   })
