diff --git a/index.js b/index.js
index e022cb3..d7df1f6 100644
--- a/index.js
+++ b/index.js
@@ -1,24 +1,23 @@
 var htmlparser = require('htmlparser2');
 var extend = require('xtend');
 var quoteRegexp = require('regexp-quote');
-require('array-includes').shim(); // Array.prototype.includes polyfill
 
 function each(obj, cb) {
-  obj && Object.keys(obj).forEach(function (key) {
-    cb(obj[key], key)
-  })
+  if (obj) Object.keys(obj).forEach(function (key) {
+    cb(obj[key], key);
+  });
 }
 
 module.exports = sanitizeHtml;
 
 // Ignore the _recursing flag; it's there for recursive
 // invocation as a guard against this exploit:
 // https://github.com/fb55/htmlparser2/issues/105
 
 function sanitizeHtml(html, options, _recursing) {
   var result = '';
 
   function Frame(tag, attribs) {
     var that = this;
     this.tag = tag;
     this.attribs = attribs || {};
@@ -91,81 +90,81 @@ function sanitizeHtml(html, options, _recursing) {
     onopentag: function(name, attribs) {
       var frame = new Frame(name, attribs);
       stack.push(frame);
 
       var skip = false;
       if (transformTagsMap[name]) {
         var transformedTag = transformTagsMap[name](name, attribs);
 
         frame.attribs = attribs = transformedTag.attribs;
         if (name !== transformedTag.tagName) {
           frame.name = name = transformedTag.tagName;
           transformMap[depth] = transformedTag.tagName;
         }
       }
 
-      if (options.allowedTags && !options.allowedTags.includes(name)) {
+      if (options.allowedTags && options.allowedTags.indexOf(name) === -1) {
         skip = true;
-        if (nonTextTagsArray.includes(name)) {
+        if (nonTextTagsArray.indexOf(name) !== -1) {
           skipText = true;
         }
         skipMap[depth] = true;
       }
       depth++;
       if (skip) {
         // We want the contents but not this tag
         return;
       }
       result += '<' + name;
       if (!allowedAttributesMap || allowedAttributesMap[name]) {
         each(attribs, function(value, a) {
-          if (!allowedAttributesMap || allowedAttributesMap[name].includes(a) ||
+          if (!allowedAttributesMap || allowedAttributesMap[name].indexOf(a) !== -1 ||
               (allowedAttributesGlobMap[name] && allowedAttributesGlobMap[name].test(a))) {
             if ((a === 'href') || (a === 'src')) {
               if (naughtyHref(name, value)) {
                 delete frame.attribs[a];
                 return;
               }
             }
             if (a === 'class') {
               value = filterClasses(value, allowedClassesMap[name]);
               if (!value.length) {
                 delete frame.attribs[a];
                 return;
               }
             }
             result += ' ' + a;
             if (value.length) {
               result += '="' + escapeHtml(value) + '"';
             }
           } else {
             delete frame.attribs[a];
           }
         });
       }
-      if (options.selfClosing.includes(name)) {
+      if (options.selfClosing.indexOf(name) !== -1) {
         result += " />";
       } else {
         result += ">";
       }
     },
     ontext: function(text) {
       if (skipText) {
         return;
       }
       var tag = stack[stack.length-1] && stack[stack.length-1].tag;
-      if (nonTextTagsArray.includes(tag)) {
+      if (nonTextTagsArray.indexOf(tag) !== -1) {
         result += text;
       } else {
         var escaped = escapeHtml(text);
         if (options.textFilter) {
           result += options.textFilter(escaped);
         } else {
           result += escaped;
         }
       }
       if (stack.length) {
            var frame = stack[stack.length - 1];
            frame.text += text;
       }
     },
     onclosetag: function(name) {
@@ -182,31 +181,31 @@ function sanitizeHtml(html, options, _recursing) {
         return;
       }
 
       if (transformMap[depth]) {
         name = transformMap[depth];
         delete transformMap[depth];
       }
 
       if (options.exclusiveFilter && options.exclusiveFilter(frame)) {
          result = result.substr(0, frame.tagPosition);
          return;
       }
 
       frame.updateParentNodeText();
 
-      if (options.selfClosing.includes(name)) {
+      if (options.selfClosing.indexOf(name) !== -1) {
          // Already output />
          return;
       }
 
       result += "</" + name + ">";
     }
   }, { decodeEntities: true });
 
   parser.write(html);
   parser.end();
 
   return result;
 
   function escapeHtml(s) {
     if (typeof(s) !== 'string') {
@@ -221,53 +220,55 @@ function sanitizeHtml(html, options, _recursing) {
     // https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet#Embedded_tab
     href = href.replace(/[\x00-\x20]+/g, '');
     // Clobber any comments in URLs, which the browser might
     // interpret inside an XML data island, allowing
     // a javascript: URL to be snuck through
     href = href.replace(/<\!\-\-.*?\-\-\>/g, '');
     // Case insensitive so we don't get faked out by JAVASCRIPT #1
     var matches = href.match(/^([a-zA-Z]+)\:/);
     if (!matches) {
       // No scheme = no way to inject js (right?)
       return false;
     }
     var scheme = matches[1].toLowerCase();
 
     if (options.allowedSchemesByTag[name]) {
-      return !options.allowedSchemesByTag[name].includes(scheme);
+      return options.allowedSchemesByTag[name].indexOf(scheme) === -1;
     }
 
-    return !options.allowedSchemes || !options.allowedSchemes.includes(scheme);
+    return !options.allowedSchemes || options.allowedSchemes.indexOf(scheme) === -1;
   }
 
   function filterClasses(classes, allowed) {
     if (!allowed) {
       // The class attribute is allowed without filtering on this tag
       return classes;
     }
     classes = classes.split(/\s+/);
     return classes.filter(function(clss) {
-      return allowed.includes(clss);
+      return allowed.indexOf(clss) !== -1;
     }).join(' ');
   }
 }
 
 // Defaults are accessible to you so that you can use them as a starting point
 // programmatically if you wish
 
 sanitizeHtml.defaults = {
-  allowedTags: [ 'h3', 'h4', 'h5', 'h6', 'blockquote', 'p', 'a', 'ul', 'ol', 'nl', 'li', 'b', 'i', 'strong', 'em', 'strike', 'code', 'hr', 'br', 'div', 'table', 'thead', 'caption', 'tbody', 'tr', 'th', 'td', 'pre' ],
+  allowedTags: [ 'h3', 'h4', 'h5', 'h6', 'blockquote', 'p', 'a', 'ul', 'ol',
+    'nl', 'li', 'b', 'i', 'strong', 'em', 'strike', 'code', 'hr', 'br', 'div',
+    'table', 'thead', 'caption', 'tbody', 'tr', 'th', 'td', 'pre' ],
   allowedAttributes: {
     a: [ 'href', 'name', 'target' ],
     // We don't currently allow img itself by default, but this
     // would make sense if we did
     img: [ 'src' ]
   },
   // Lots of these won't come up by default because we don't allow them
   selfClosing: [ 'img', 'br', 'hr', 'area', 'base', 'basefont', 'input', 'link', 'meta' ],
   // URL schemes we permit
   allowedSchemes: [ 'http', 'https', 'ftp', 'mailto' ],
   allowedSchemesByTag: {}
 };
 
 sanitizeHtml.simpleTransform = function(newTagName, newAttribs, merge) {
   merge = (merge === undefined) ? true : merge;
diff --git a/package.json b/package.json
index af07077..ba27789 100644
--- a/package.json
+++ b/package.json
@@ -9,21 +9,20 @@
   "repository": {
     "type": "git",
     "url": "https://github.com/punkave/sanitize-html.git"
   },
   "keywords": [
     "html",
     "parser",
     "sanitizer",
     "html",
     "sanitizer",
     "apostrophe"
   ],
   "author": "P'unk Avenue LLC",
   "license": "MIT",
   "dependencies": {
-    "array-includes": "^2.0.0",
     "htmlparser2": "3.8.x",
     "regexp-quote": "0.0.0",
     "xtend": "^4.0.0"
   }
 }
