diff --git a/include/default_controls.js b/include/default_controls.js
index 2864ee7..ceb374b 100644
--- a/include/default_controls.js
+++ b/include/default_controls.js
@@ -44,32 +44,32 @@ load: function(target) {
     html += '          onmouseout="DefaultControls.canvasFocus();">';
     html += '          <ul>';
     html += '            <li><input id="VNC_encrypt"';
     html += '                type="checkbox"> Encrypt</li>';
     html += '            <li><input id="VNC_true_color"';
     html += '                type="checkbox" checked> True Color</li>';
     html += '            <li><input id="VNC_cursor"';
     html += '                type="checkbox"> Local Cursor</li>';
     html += '            <li><input id="VNC_connectTimeout"';
     html += '                type="input"> Connect Timeout (s)</li>';
     html += '            <hr>';
 
     // Stylesheet selection dropdown
     html += '            <li><select id="VNC_stylesheet" name="vncStyle">';
     html += '              <option value="default">default</option>';
-    sheet = Util.selectStylesheet();
-    sheets = Util.getStylesheets();
+    sheet = WebUtil.selectStylesheet();
+    sheets = WebUtil.getStylesheets();
     for (i = 0; i < sheets.length; i += 1) {
         html += '<option value="' + sheets[i].title + '">' + sheets[i].title + '</option>';
     }
     html += '              </select> Style</li>';
 
     // Logging selection dropdown
     html += '            <li><select id="VNC_logging" name="vncLogging">';
     llevels = ['error', 'warn', 'info', 'debug'];
     for (i = 0; i < llevels.length; i += 1) {
         html += '<option value="' + llevels[i] + '">' + llevels[i] + '</option>';
     }
     html += '              </select> Logging</li>';
 
     html += '            <hr>';
     html += '            <li><input type="button" id="VNC_apply" value="Apply"';
@@ -90,121 +90,121 @@ load: function(target) {
     html += '<div id="VNC_clipboard">';
     html += '  VNC Clipboard:';
     html += '  <input id="VNC_clipboard_clear_button"';
     html += '      type="button" value="Clear"';
     html += '      onclick="DefaultControls.clipClear();">';
     html += '  <br>';
     html += '  <textarea id="VNC_clipboard_text" cols=80 rows=5';
     html += '    onfocus="DefaultControls.canvasBlur();"';
     html += '    onblur="DefaultControls.canvasFocus();"';
     html += '    onchange="DefaultControls.clipSend();"></textarea>';
     html += '</div>';
     $(target).innerHTML = html;
 
     // Settings with immediate effects
     DC.initSetting('logging', 'warn');
-    Util.init_logging(DC.getSetting('logging'));
+    WebUtil.init_logging(DC.getSetting('logging'));
     DC.initSetting('stylesheet', 'default');
 
-    Util.selectStylesheet(null); // call twice to get around webkit bug
-    Util.selectStylesheet(DC.getSetting('stylesheet'));
+    WebUtil.selectStylesheet(null); // call twice to get around webkit bug
+    WebUtil.selectStylesheet(DC.getSetting('stylesheet'));
 
     /* Populate the controls if defaults are provided in the URL */
     DC.initSetting('host', '');
     DC.initSetting('port', '');
     DC.initSetting('password', '');
     DC.initSetting('encrypt', false);
     DC.initSetting('true_color', true);
     DC.initSetting('cursor', false);
     DC.initSetting('connectTimeout', 2);
 
     DC.rfb = RFB({'target': 'VNC_canvas',
                   'updateState': DC.updateState,
                   'clipboardReceive': DC.clipReceive});
 
     // Unfocus clipboard when over the VNC area
     $('VNC_screen').onmousemove = function () {
             var canvas = DC.rfb.get_canvas();
             if ((! canvas) || (! canvas.get_focused())) {
                 $('VNC_clipboard_text').blur();
             }
         };
 
 },
 
 // Read form control compatible setting from cookie
 getSetting: function(name) {
     var val, ctrl = $('VNC_' + name);
-    val = Util.readCookie(name);
+    val = WebUtil.readCookie(name);
     if (ctrl.type === 'checkbox') {
         if (val.toLowerCase() in {'0':1, 'no':1, 'false':1}) {
             val = false;
         } else {
             val = true;
         }
     }
     return val;
 },
 
 // Update cookie and form control setting. If value is not set, then
 // updates from control to current cookie setting.
 updateSetting: function(name, value) {
     var i, ctrl = $('VNC_' + name);
     // Save the cookie for this session
     if (typeof value !== 'undefined') {
-        Util.createCookie(name, value);
+        WebUtil.createCookie(name, value);
     }
 
     // Update the settings control
     value = DefaultControls.getSetting(name);
     if (ctrl.type === 'checkbox') {
         ctrl.checked = value;
     } else if (typeof ctrl.options !== 'undefined') {
         for (i = 0; i < ctrl.options.length; i += 1) {
             if (ctrl.options[i].value === value) {
                 ctrl.selectedIndex = i;
                 break;
             }
         }
     } else {
         ctrl.value = value;
     }
 },
 
 // Save control setting to cookie
 saveSetting: function(name) {
     var val, ctrl = $('VNC_' + name);
     if (ctrl.type === 'checkbox') {
         val = ctrl.checked;
     } else if (typeof ctrl.options !== 'undefined') {
         val = ctrl.options[ctrl.selectedIndex].value;
     } else {
         val = ctrl.value;
     }
-    Util.createCookie(name, val);
+    WebUtil.createCookie(name, val);
     //Util.Debug("Setting saved '" + name + "=" + val + "'");
     return val;
 },
 
 // Initial page load read/initialization of settings
 initSetting: function(name, defVal) {
     var val;
 
     // Check Query string followed by cookie
-    val = Util.getQueryVar(name);
+    val = WebUtil.getQueryVar(name);
     if (val === null) {
-        val = Util.readCookie(name, defVal);
+        val = WebUtil.readCookie(name, defVal);
     }
     DefaultControls.updateSetting(name, val);
     //Util.Debug("Setting '" + name + "' initialized to '" + val + "'");
     return val;
 },
 
 
 // Toggle the settings menu:
 //   On open, settings are refreshed from saved cookies.
 //   On close, settings are applied
 clickSettingsMenu: function() {
     var DC = DefaultControls;
     if (DC.settingsOpen) {
         DC.settingsApply();
 
@@ -255,32 +255,32 @@ settingsDisabled: function(disabled, rfb) {
 
 // Save/apply settings when 'Apply' button is pressed
 settingsApply: function() {
     //Util.Debug(">> settingsApply");
     var DC = DefaultControls;
     DC.saveSetting('encrypt');
     DC.saveSetting('true_color');
     if (DC.rfb.get_canvas().get_cursor_uri()) {
         DC.saveSetting('cursor');
     }
     DC.saveSetting('connectTimeout');
     DC.saveSetting('stylesheet');
     DC.saveSetting('logging');
 
     // Settings with immediate (non-connected related) effect
-    Util.selectStylesheet(DC.getSetting('stylesheet'));
-    Util.init_logging(DC.getSetting('logging'));
+    WebUtil.selectStylesheet(DC.getSetting('stylesheet'));
+    WebUtil.init_logging(DC.getSetting('logging'));
 
     //Util.Debug("<< settingsApply");
 },
 
 
 
 setPassword: function() {
     DefaultControls.rfb.sendPassword($('VNC_password').value);
     return false;
 },
 
 sendCtrlAltDel: function() {
     DefaultControls.rfb.sendCtrlAltDel();
 },
 
diff --git a/include/util.js b/include/util.js
index 91883f4..9610a0b 100644
--- a/include/util.js
+++ b/include/util.js
@@ -50,33 +50,30 @@ Array.prototype.push32 = function (num) {
 };
 
 /* 
  * ------------------------------------------------------
  * Namespaced in Util
  * ------------------------------------------------------
  */
 
 /*
  * Logging/debug routines
  */
 
 Util._log_level = 'warn';
 Util.init_logging = function (level) {
     if (typeof level === 'undefined') {
-        Util._log_level = (document.location.href.match(
-                /logging=([A-Za-z0-9\._\-]*)/) ||
-                ['', Util._log_level])[1];
         level = Util._log_level;
     } else {
         Util._log_level = level;
     }
     if (typeof window.console === "undefined") {
         if (typeof window.opera !== "undefined") {
             window.console = {
                 'log'  : window.opera.postError,
                 'warn' : window.opera.postError,
                 'error': window.opera.postError };
         } else {
             window.console = {
                 'log'  : function(m) {},
                 'warn' : function(m) {},
                 'error': function(m) {}};
@@ -84,63 +81,35 @@ Util.init_logging = function (level) {
     }
 
     Util.Debug = Util.Info = Util.Warn = Util.Error = function (msg) {};
     switch (level) {
         case 'debug': Util.Debug = function (msg) { console.log(msg); };
         case 'info':  Util.Info  = function (msg) { console.log(msg); };
         case 'warn':  Util.Warn  = function (msg) { console.warn(msg); };
         case 'error': Util.Error = function (msg) { console.error(msg); };
         case 'none':
             break;
         default:
             throw("invalid logging type '" + level + "'");
     }
 };
 Util.get_logging = function () {
-        return Util._log_level;
-    }
+    return Util._log_level;
+}
 // Initialize logging level
 Util.init_logging();
 
-Util.dirObj = function (obj, depth, parent) {
-    var i, msg = "", val = "";
-    if (! depth) { depth=2; }
-    if (! parent) { parent= ""; }
-
-    // Print the properties of the passed-in object 
-    for (i in obj) {
-        if ((depth > 1) && (typeof obj[i] === "object")) { 
-            // Recurse attributes that are objects
-            msg += Util.dirObj(obj[i], depth-1, parent + "." + i);
-        } else {
-            //val = new String(obj[i]).replace("\n", " ");
-            val = obj[i].toString().replace("\n", " ");
-            if (val.length > 30) {
-                val = val.substr(0,30) + "...";
-            } 
-            msg += parent + "." + i + ": " + val + "\n";
-        }
-    }
-    return msg;
-};
-
-// Read a query string variable
-Util.getQueryVar = function(name, defVal) {
-    var re = new RegExp('[?][^#]*' + name + '=([^&#]*)');
-    if (typeof defVal === 'undefined') { defVal = null; }
-    return (document.location.href.match(re) || ['',defVal])[1];
-};
 
 // Set defaults for Crockford style function namespaces
 Util.conf_default = function(cfg, api, v, type, defval, desc) {
     // Description
     api['get_' + v + '_desc'] = desc;
     // Default getter
     if (typeof api['get_' + v] === 'undefined') {
         api['get_' + v] = function () {
                 return cfg[v];
             };
     }
     // Default setter
     if (typeof api['set_' + v] === 'undefined') {
         api['set_' + v] = function (val) {
                 if (type in {'boolean':1, 'bool':1}) {
@@ -257,82 +226,16 @@ Util.Engine = {
 };
 
 Util.Flash = (function(){
     var v, version;
     try {
         v = navigator.plugins['Shockwave Flash'].description;
     } catch(err1) {
         try {
             v = new ActiveXObject('ShockwaveFlash.ShockwaveFlash').GetVariable('$version');
         } catch(err2) {
             v = '0 r0';
         }
     }
     version = v.match(/\d+/g);
     return {version: parseInt(version[0] || 0 + '.' + version[1], 10) || 0, build: parseInt(version[2], 10) || 0};
-}()); 
-
-/*
- * Cookie handling. Dervied from: http://www.quirksmode.org/js/cookies.html
- */
-
-// No days means only for this browser session
-Util.createCookie = function(name,value,days) {
-    var date, expires;
-    if (days) {
-        date = new Date();
-        date.setTime(date.getTime()+(days*24*60*60*1000));
-        expires = "; expires="+date.toGMTString();
-    }
-    else {
-        expires = "";
-    }
-    document.cookie = name+"="+value+expires+"; path=/";
-};
-
-Util.readCookie = function(name, defaultValue) {
-    var i, c, nameEQ = name + "=", ca = document.cookie.split(';');
-    for(i=0; i < ca.length; i += 1) {
-        c = ca[i];
-        while (c.charAt(0) === ' ') { c = c.substring(1,c.length); }
-        if (c.indexOf(nameEQ) === 0) { return c.substring(nameEQ.length,c.length); }
-    }
-    return (typeof defaultValue !== 'undefined') ? defaultValue : null;
-};
-
-Util.eraseCookie = function(name) {
-    Util.createCookie(name,"",-1);
-};
-
-/*
- * Alternate stylesheet selection
- */
-Util.getStylesheets = function() { var i, links, sheets = [];
-    links = document.getElementsByTagName("link");
-    for (i = 0; i < links.length; i += 1) {
-        if (links[i].title &&
-            links[i].rel.toUpperCase().indexOf("STYLESHEET") > -1) {
-            sheets.push(links[i]);
-        }
-    }
-    return sheets;
-};
-
-// No sheet means try and use value from cookie, null sheet used to
-// clear all alternates.
-Util.selectStylesheet = function(sheet) {
-    var i, link, sheets = Util.getStylesheets();
-    if (typeof sheet === 'undefined') {
-        sheet = 'default';
-    }
-    for (i=0; i < sheets.length; i += 1) {
-        link = sheets[i];
-        if (link.title === sheet) {    
-            Util.Debug("Using stylesheet " + sheet);
-            link.disabled = false;
-        } else {
-            //Util.Debug("Skipping stylesheet " + link.title);
-            link.disabled = true;
-        }
-    }
-    return sheet;
-};
+}()); 
\ No newline at end of file
diff --git a/include/vnc.js b/include/vnc.js
index 043ae63..a11f28d 100644
--- a/include/vnc.js
+++ b/include/vnc.js
@@ -19,30 +19,31 @@ var VNC_native_ws, WEB_SOCKET_SWF_LOCATION;
 function get_VNC_uri_prefix() {
     return (typeof VNC_uri_prefix !== "undefined") ? VNC_uri_prefix : "include/";
 }
 
 (function () {
     var extra = "", start, end;
 
     start = "<script src='" + get_VNC_uri_prefix();
     end = "'><\/script>";
 
     // Uncomment to activate firebug lite
     //extra += "<script src='http://getfirebug.com/releases/lite/1.2/" + 
     //         "firebug-lite-compressed.js'><\/script>";
 
     extra += start + "util.js" + end;
+    extra += start + "webutil.js" + end;
     extra += start + "base64.js" + end;
     extra += start + "des.js" + end;
     extra += start + "canvas.js" + end;
     extra += start + "rfb.js" + end;
 
     /* If no builtin websockets then load web_socket.js */
     if (window.WebSocket) {
         VNC_native_ws = true;
     } else {
         VNC_native_ws = false;
         WEB_SOCKET_SWF_LOCATION = get_VNC_uri_prefix() +
                     "web-socket-js/WebSocketMain.swf";
         extra += start + "web-socket-js/swfobject.js" + end;
         extra += start + "web-socket-js/FABridge.js" + end;
         extra += start + "web-socket-js/web_socket.js" + end;
diff --git a/include/webutil.js b/include/webutil.js
new file mode 100644
index 0000000..162cae4
--- /dev/null
+++ b/include/webutil.js
@@ -0,0 +1,144 @@
+/*
+ * noVNC: HTML5 VNC client
+ * Copyright (C) 2010 Joel Martin
+ * Licensed under LGPL-3 (see LICENSE.txt)
+ *
+ * See README.md for usage and integration instructions.
+ */
+
+"use strict";
+/*jslint bitwise: false, white: false */
+/*global window, console, document, navigator, ActiveXObject*/
+
+// Globals defined here
+var WebUtil = {}, $;
+
+/*
+ * Simple DOM selector by ID
+ */
+if (!window.$) {
+    $ = function (id) {
+        if (document.getElementById) {
+            return document.getElementById(id);
+        } else if (document.all) {
+            return document.all[id];
+        } else if (document.layers) {
+            return document.layers[id];
+        }
+        return undefined;
+    };
+}
+
+
+/* 
+ * ------------------------------------------------------
+ * Namespaced in WebUtil
+ * ------------------------------------------------------
+ */
+
+// init log level reading the logging HTTP param
+WebUtil.init_logging = function() {
+    Util._log_level = (document.location.href.match(
+         /logging=([A-Za-z0-9\._\-]*)/) ||
+         ['', Util._log_level])[1];
+    
+    Util.init_logging()
+}
+WebUtil.init_logging();
+
+
+WebUtil.dirObj = function (obj, depth, parent) {
+    var i, msg = "", val = "";
+    if (! depth) { depth=2; }
+    if (! parent) { parent= ""; }
+
+    // Print the properties of the passed-in object 
+    for (i in obj) {
+        if ((depth > 1) && (typeof obj[i] === "object")) { 
+            // Recurse attributes that are objects
+            msg += WebUtil.dirObj(obj[i], depth-1, parent + "." + i);
+        } else {
+            //val = new String(obj[i]).replace("\n", " ");
+            val = obj[i].toString().replace("\n", " ");
+            if (val.length > 30) {
+                val = val.substr(0,30) + "...";
+            } 
+            msg += parent + "." + i + ": " + val + "\n";
+        }
+    }
+    return msg;
+};
+
+// Read a query string variable
+WebUtil.getQueryVar = function(name, defVal) {
+    var re = new RegExp('[?][^#]*' + name + '=([^&#]*)');
+    if (typeof defVal === 'undefined') { defVal = null; }
+    return (document.location.href.match(re) || ['',defVal])[1];
+};
+
+
+/*
+ * Cookie handling. Dervied from: http://www.quirksmode.org/js/cookies.html
+ */
+
+// No days means only for this browser session
+WebUtil.createCookie = function(name,value,days) {
+    var date, expires;
+    if (days) {
+        date = new Date();
+        date.setTime(date.getTime()+(days*24*60*60*1000));
+        expires = "; expires="+date.toGMTString();
+    }
+    else {
+        expires = "";
+    }
+    document.cookie = name+"="+value+expires+"; path=/";
+};
+
+WebUtil.readCookie = function(name, defaultValue) {
+    var i, c, nameEQ = name + "=", ca = document.cookie.split(';');
+    for(i=0; i < ca.length; i += 1) {
+        c = ca[i];
+        while (c.charAt(0) === ' ') { c = c.substring(1,c.length); }
+        if (c.indexOf(nameEQ) === 0) { return c.substring(nameEQ.length,c.length); }
+    }
+    return (typeof defaultValue !== 'undefined') ? defaultValue : null;
+};
+
+WebUtil.eraseCookie = function(name) {
+    WebUtil.createCookie(name,"",-1);
+};
+
+/*
+ * Alternate stylesheet selection
+ */
+WebUtil.getStylesheets = function() { var i, links, sheets = [];
+    links = document.getElementsByTagName("link");
+    for (i = 0; i < links.length; i += 1) {
+        if (links[i].title &&
+            links[i].rel.toUpperCase().indexOf("STYLESHEET") > -1) {
+            sheets.push(links[i]);
+        }
+    }
+    return sheets;
+};
+
+// No sheet means try and use value from cookie, null sheet used to
+// clear all alternates.
+WebUtil.selectStylesheet = function(sheet) {
+    var i, link, sheets = WebUtil.getStylesheets();
+    if (typeof sheet === 'undefined') {
+        sheet = 'default';
+    }
+    for (i=0; i < sheets.length; i += 1) {
+        link = sheets[i];
+        if (link.title === sheet) {    
+            Util.Debug("Using stylesheet " + sheet);
+            link.disabled = false;
+        } else {
+            //Util.Debug("Skipping stylesheet " + link.title);
+            link.disabled = true;
+        }
+    }
+    return sheet;
+};
diff --git a/tests/base64.html b/tests/base64.html
index 58eb7ac..b13253e 100644
--- a/tests/base64.html
+++ b/tests/base64.html
@@ -1,20 +1,21 @@
 <!DOCTYPE html> 
 <html xmlns="http://www.w3.org/1999/xhtml"> 
   <head> 
     <title>Native Base64 Tests</title> 
     <script src="include/util.js"></script> 
+    <script src="include/webutil.js"></script> 
     <script src="include/base64.js"></script> 
   </head> 
   <body> 
         <h1>Native Base64 Tests</h1> 
  
         <br> 
         Messages:<br> 
         <textarea id="debug" style="font-size: 9px;" cols=80 rows=25></textarea> 
     
         <br>
   </body> 
  
 <script> 
     function debug(str) {
         console.log(str);
diff --git a/tests/canvas.html b/tests/canvas.html
index f271953..382e1dd 100644
--- a/tests/canvas.html
+++ b/tests/canvas.html
@@ -1,23 +1,24 @@
 <html>
     <head>
         <title>Canvas Performance Test</title>
         <!--
         <script type='text/javascript' 
             src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
         -->
         <script src="include/util.js"></script>
+        <script src="include/webutil.js"></script>
         <script src="include/base64.js"></script>
         <script src="include/canvas.js"></script>
         <script src="face.png.js"></script>
     </head>
     <body>
         Iterations: <input id='iterations' style='width:50' value="100">&nbsp;
 
         Width: <input id='width' style='width:50' value="640">&nbsp;
         Height: <input id='height' style='width:50' value="480">&nbsp;
 
         <input id='startButton' type='button' value='Do Performance Test'
             style='width:150px' onclick="begin();">&nbsp;
 
         <br><br>
 
diff --git a/tests/cursor.html b/tests/cursor.html
index 867b55d..0e0dd9a 100644
--- a/tests/cursor.html
+++ b/tests/cursor.html
@@ -1,25 +1,26 @@
 <!DOCTYPE html> 
 <html xmlns="http://www.w3.org/1999/xhtml"> 
   <head> 
     <title>Cursor Change test</title> 
     <meta charset="UTF-8"> 
     <!--
     <script type='text/javascript' 
                     src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
     -->
     <script src="include/util.js"></script> 
+    <script src="include/webutil.js"></script> 
     <script src="include/base64.js"></script> 
     <script src="include/canvas.js"></script> 
   </head> 
   <body> 
         <h1>Roll over the buttons to test cursors</h1>
         <br>
         <input id=button1 type="button" value="Cursor from file (smiley face)">
         <input id=button2 type="button" value="Data URI cursor (crosshair)">
     
         <br> 
         <br> 
         <br> 
         Debug:<br> 
         <textarea id="debug" style="font-size: 9px;" cols=80 rows=25></textarea> 
         <br>
diff --git a/tests/input.html b/tests/input.html
index 43791ed..b3fc8ed 100644
--- a/tests/input.html
+++ b/tests/input.html
@@ -7,30 +7,31 @@
         <canvas id="canvas" width="640" height="20"
                 style="border-style: dotted; border-width: 1px;">
             Canvas not supported.
         </canvas>
 
         <br>
         Results:<br>
         <textarea id="messages" style="font-size: 9;" cols=80 rows=25></textarea>
     </body>
 
     <!--
     <script type='text/javascript' 
         src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
     -->
     <script src="include/util.js"></script>
+    <script src="include/webutil.js"></script> 
     <script src="include/base64.js"></script>
     <script src="include/canvas.js"></script>
     <script>
         var msg_cnt = 0;
         var width = 400, height = 200;
         var iterations;
 
         function message(str) {
             console.log(str);
             cell = $('messages');
             cell.innerHTML += msg_cnt + ": " + str + "\n";
             cell.scrollTop = cell.scrollHeight;
         }
 
         function mouseButton(x, y, down, bmask) {
diff --git a/tests/vnc_playback.html b/tests/vnc_playback.html
index 4c52e5c..d777a84 100644
--- a/tests/vnc_playback.html
+++ b/tests/vnc_playback.html
@@ -38,31 +38,31 @@
     -->
 
     <script src="include/vnc.js"></script>
     <script src="include/playback.js"></script>
 
     <script>
         var fname, start_time;
 
         function message(str) {
             console.log(str);
             var cell = $('messages');
             cell.innerHTML += str + "\n";
             cell.scrollTop = cell.scrollHeight;
         }
 
-        fname = Util.getQueryVar('data', null);
+        fname = WebUtil.getQueryVar('data', null);
 
         if (fname) {
             message("Loading " + fname);
             document.write('<script src="' + fname + '"><\/script>');
         } else {
             message("Must specify data=FOO in query string.");
         }
 
         updateState = function (rfb, state, oldstate, msg) {
             switch (state) {
                 case 'failed':
                 case 'fatal':
                     message("noVNC sent '" + state + "' state during iteration " + iteration + " frame " + frame_idx);
                     test_state = 'failed';
                     break;
@@ -97,32 +97,32 @@
         function finish() {
             // Finished with all iterations
             var total_time, end_time = (new Date()).getTime();
             total_time = end_time - start_time;
 
             iter_time = parseInt(total_time / iterations, 10);
             message(iterations + " iterations took " + total_time + "ms, " +
                     iter_time + "ms per iteration");
             rfb.get_canvas().stop();   // Shut-off event interception
             $('startButton').disabled = false;
             $('startButton').value = "Start";
 
         }
 
         window.onload = function() {
-            iterations = Util.getQueryVar('iterations', 3);
+            iterations = WebUtil.getQueryVar('iterations', 3);
             $('iterations').value = iterations;
-            mode = Util.getQueryVar('mode', 3);
+            mode = WebUtil.getQueryVar('mode', 3);
             if (mode === 'realtime') {
                 $('mode2').checked = true;
             } else {
                 $('mode1').checked = true;
             }
             if (fname) {
                 message("VNC_frame_data.length: " + VNC_frame_data.length);
                 rfb = RFB({'target': 'VNC_canvas',
                         'updateState': updateState});
                 rfb.testMode(send_array);
             }
         }
     </script>
 </html>
diff --git a/tests/ws.html b/tests/ws.html
index 24c8ea1..e6bf5ac 100644
--- a/tests/ws.html
+++ b/tests/ws.html
@@ -1,21 +1,22 @@
 <html>
 
     <head>
         <title>WebSockets Test</title>
         <script src="include/base64.js"></script>
         <script src="include/util.js"></script>
+        <script src="include/webutil.js"></script> 
         <!-- Uncomment to activate firebug lite -->
         <!--
         <script type='text/javascript' 
             src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
         -->
 
 
     </head>
 
     <body>
 
         Host: <input id='host' style='width:100'>&nbsp;
         Port: <input id='port' style='width:50'>&nbsp;
         Encrypt: <input id='encrypt' type='checkbox'>&nbsp;
         Send Delay (ms): <input id='sendDelay' style='width:50' value="100">&nbsp;
diff --git a/tests/wsencoding.html b/tests/wsencoding.html
index 7c0d6bc..48277c9 100644
--- a/tests/wsencoding.html
+++ b/tests/wsencoding.html
@@ -12,30 +12,31 @@
 
         <br>
         Messages:<br>
         <textarea id="messages" style="font-size: 9;" cols=80 rows=25></textarea>
     </body>
 
 
     <!-- Uncomment to activate firebug lite -->
     <!--
     <script type='text/javascript' 
         src='http://getfirebug.com/releases/lite/1.2/firebug-lite-compressed.js'></script>
     -->
 
     <script src="include/base64.js"></script>
     <script src="include/util.js"></script>
+    <script src="include/webutil.js"></script>
 
     <script>
 
         var host = null, port = null;
         var ws = null;
         var VNC_native_ws = true;
 
         function message(str) {
             console.log(str);
             cell = $('messages');
             cell.innerHTML += str + "\n";
             cell.scrollTop = cell.scrollHeight;
         }
 
         function print_response(str) {
diff --git a/vnc_auto.html b/vnc_auto.html
index 91e0463..9481017 100644
--- a/vnc_auto.html
+++ b/vnc_auto.html
@@ -79,35 +79,35 @@
 
             if (state === "normal") { cad.disabled = false; }
             else                    { cad.disabled = true; }
 
             if (typeof(msg) !== 'undefined') {
                 sb.setAttribute("class", klass);
                 s.innerHTML = msg;
             }
         }
 
         window.onload = function () {
             var host, port, password;
 
             $('sendCtrlAltDelButton').onclick = sendCtrlAltDel;
 
-            host = Util.getQueryVar('host', null);
-            port = Util.getQueryVar('port', null);
-            password = Util.getQueryVar('password', '');
+            host = WebUtil.getQueryVar('host', null);
+            port = WebUtil.getQueryVar('port', null);
+            password = WebUtil.getQueryVar('password', '');
             if ((!host) || (!port)) {
                 updateState('failed',
                     "Must specify host and port in URL");
                 return;
             }
 
-            rfb = new RFB({'encrypt':      Util.getQueryVar('encrypt', false),
-                       'true_color':   Util.getQueryVar('true_color', true),
-                       'local_cursor': Util.getQueryVar('cursor', true),
+            rfb = new RFB({'encrypt':      WebUtil.getQueryVar('encrypt', false),
+                       'true_color':   WebUtil.getQueryVar('true_color', true),
+                       'local_cursor': WebUtil.getQueryVar('cursor', true),
                        'updateState':  updateState});
             rfb.connect(host, port, password);
         };
         </script>
 
     </body>
 </html>
 
