diff --git a/index.js b/index.js
index b0d94a0..639b1bd 100644
--- a/index.js
+++ b/index.js
@@ -53,17 +53,21 @@ exports.get = function (obj, path) {
  * @param {String} path
  * @return {Mixed}
  * @api public
  */
 
 exports.delete = function (obj, path) {
   var segs = path.split('.');
   var attr = segs.pop();
 
   for (var i = 0; i < segs.length; i++) {
     var seg = segs[i];
     if (!obj[seg]) return;
     obj = obj[seg];
   }
 
-  delete obj[attr];
+  if (Array.isArray(obj)) {
+    obj.splice(path, 1);
+  } else {
+    delete obj[attr];
+  }
 };
diff --git a/test/index.js b/test/index.js
index 4977610..1f5e4f5 100644
--- a/test/index.js
+++ b/test/index.js
@@ -9,25 +9,28 @@ var tests = module.exports = {
     assert(obj === ret);
   },
 
   'test get': function () {
     var obj = {};
     obj.cool = {};
     obj.cool.aid = 'rocks';
     var value = dot.get(obj, 'cool.aid');
     assert(value === 'rocks');
   },
 
   'test delete': function () {
     var obj = {};
     obj.cool = {};
     obj.cool.aid = 'rocks';
+    obj.cool.hello = ['world'];
     dot.delete(obj, 'cool.aid');
+    dot.delete(obj, 'cool.hello.0');
     assert(!obj.cool.hasOwnProperty('aid'))
+    assert(obj.cool.hello.length == 0);
   }
 }
 
 for (var t in tests) {
   tests[t]();
 }
 
 console.log('All tests passed!');
