diff --git a/src/node/utils/Minify.js b/src/node/utils/Minify.js
index 21a663c0e..a16e480af 100644
--- a/src/node/utils/Minify.js
+++ b/src/node/utils/Minify.js
@@ -132,30 +132,31 @@ function minify(req, res, next)
   }
 
   /* Handle static files for plugins:
      paths like "plugins/ep_myplugin/static/js/test.js"
      are rewritten into ROOT_PATH_OF_MYPLUGIN/static/js/test.js,
      commonly ETHERPAD_ROOT/node_modules/ep_myplugin/static/js/test.js
   */
   var match = filename.match(/^plugins\/([^\/]+)\/static\/(.*)/);
   if (match) {
     var pluginName = match[1];
     var resourcePath = match[2];
     var plugin = plugins.plugins[pluginName];
     if (plugin) {
       var pluginPath = plugin.package.realPath;
       filename = path.relative(ROOT_DIR, pluginPath + '/static/' + resourcePath);
+      filename = filename.replace(/\\/g, '/'); // Windows (safe generally?)
     }
   }
 
   // What content type should this be?
   // TODO: This should use a MIME module.
   var contentType;
   if (filename.match(/\.js$/)) {
     contentType = "text/javascript";
   } else if (filename.match(/\.css$/)) {
     contentType = "text/css";
   } else if (filename.match(/\.html$/)) {
     contentType = "text/html";
   } else if (filename.match(/\.txt$/)) {
     contentType = "text/plain";
   } else if (filename.match(/\.png$/)) {
