diff --git a/README.md b/README.md
index d4779fc..0484520 100644
--- a/README.md
+++ b/README.md
@@ -21,58 +21,61 @@ npm install im-resize --save
 
 ## API
 
 ```js
 var resize = require('im-resize');
 ```
 
 ### resize(**object** `image`, **object[]** `versions`, **function** `cb`)
 
 Resize a given source `image` into several `versions`.
 
 * **object** `image` - source image to resize
   * **integer** `width` - image pixel width
   * **integer** `height` - image pixel height
   * **string** `path` - complete path to source image
-* **object[]** `versions` - Array of version objects
-  * **string** `suffix` - suffix for the resized image (ex. "-small")
-  * **integer** `maxWidth` - max width for resized image
-  * **integer** `maxHeight` - max height for resized image
-  * **integer** `quality` - quality for resized image
-  * **string** `ratio` - force aspectratio on resized image (ex. "4:3")
-  * **boolean** `flatten` - used in conjunction with background
-  * **string** `background` - set background to transparent image
-  * **string** `format` - image format for resized image (ex. "png")
+* **object** `output` - image resize output config
+  * **object[]** `versions` - Array of version objects
+    * **string** `suffix` - suffix for the resized image (ex. "-small")
+    * **integer** `maxWidth` - max width for resized image
+    * **integer** `maxHeight` - max height for resized image
+    * **integer** `quality` - quality for resized image
+    * **string** `ratio` - force aspectratio on resized image (ex. "4:3")
+    * **boolean** `flatten` - used in conjunction with background
+    * **string** `background` - set background to transparent image
+    * **string** `format` - image format for resized image (ex. "png")
 * **function** `cb` - callback function (**Error** `error`, **object[]** `versions`)
   * **Error** `error` - error output if command failed
   * **object[]** `versions` - resized image versions
     * **string** `path` path to the resized image
 
 #### Example
 
 ```js
 var image = {
   path: '/path/to/image.jpg',
   width: 1024,
   height: 768
 };
 
-var versions = [{
-  suffix: '-thumb',
-  maxHeight: 150,
-  maxWidth: 150,
-  aspect: "3:2"
-},{
-  suffix: '-square',
-  maxHeight: 200,
-  maxWidth: 200,
-  aspect: "1:1"
-]};
-
-resize(image, versions, function(error, versions) {
+var output = {
+  versions: [{
+    suffix: '-thumb',
+    maxHeight: 150,
+    maxWidth: 150,
+    aspect: "3:2"
+  },{
+    suffix: '-square',
+    maxHeight: 200,
+    maxWidth: 200,
+    aspect: "1:1"
+  ]}
+};
+
+resize(image, output, function(error, versions) {
   if (error) { console.error(error); }
   console.log(versions[0].path); // /path/to/image-thumb.jpg
   console.log(versions[1].path); // /path/to/image-square.jpg
 });
 ```
 
 ## [MIT License](https://github.com/Turistforeningen/node-im-resize/blob/master/LICENSE)
diff --git a/index.js b/index.js
index 2929185..7f69599 100644
--- a/index.js
+++ b/index.js
@@ -1,30 +1,30 @@
 var exec = require('child_process').exec, child;
 var aspect = require('aspectratio');
 var dirname = require('path').dirname;
 var basename = require('path').basename;
 var extname = require('path').extname;
 var join = require('path').join;
 var sprintf = require('util').format;
 
-module.exports = function(image, versions, cb) {
-  var cmd = module.exports.cmd(image, versions);
+module.exports = function(image, output, cb) {
+  var cmd = module.exports.cmd(image, output.versions);
   exec(cmd, {timeout: 10000}, function(e, stdout, stderr) {
     if (e) { return cb(e); }
     if (stderr) { return cb(new Error(stderr)); }
 
-    return cb(null, versions);
+    return cb(null, output.versions);
   });
 };
 
 /**
  * Get cropped geometry for given aspectratio
  *
  * @param string image - original image metadata
  * @param string ratio - new aspect ratio
  *
  * @return string geometry; false if no crop applies
  */
 module.exports.crop = function(image, ratio) {
   if (!ratio) { return false; }
 
   var g = aspect.fixed(image.width, image.height, ratio);
diff --git a/test.js b/test.js
index 7458007..529d81e 100644
--- a/test.js
+++ b/test.js
@@ -59,96 +59,98 @@ describe('resize.cmdVersion()', function() {
     var version = {
       suffix: '-b',
       quality: 50,
       maxWidth: 500,
       maxHeight: 500
     };
 
     var cmd = resize.cmdVersion(image, version);
     var out = 'mpr:./a.jpg -quality 50 -resize "500x500" -write a-b.jpg +delete';
 
     assert.equal(cmd, out);
   });
 });
 
 describe('resize()', function() {
-  var versions;
+  var output;
 
   beforeEach(function() {
-    versions = [{
-      suffix: '-full',
-      maxHeight: 1920,
-      maxWidth: 1920
-    },{
-      suffix: '-1200',
-      maxHeight: 1200,
-      maxWidth: 1200,
-      aspect: "3:2"
-    },{
-      suffix: '-800',
-      maxHeight: 800,
-      maxWidth: 800,
-      aspect: "3:2"
-    },{
-      suffix: '-500',
-      maxHeight: 500,
-      maxWidth: 500,
-      aspect: "3:2"
-    },{
-      suffix: '-260',
-      maxHeight: 260,
-      maxWidth: 260,
-      aspect: "3:2"
-    },{
-      suffix: '-150',
-      maxHeight: 150,
-      maxWidth: 150,
-      aspect: "3:2"
-    },{
-      suffix: '-square-200',
-      maxHeight: 200,
-      maxWidth: 200,
-      aspect: "1:1"
-    },{
-      suffix: '-square-50',
-      maxHeight: 50,
-      maxWidth: 50,
-      aspect: "1:1"
-    }];
+    output = {
+      versions: [{
+        suffix: '-full',
+        maxHeight: 1920,
+        maxWidth: 1920
+      },{
+        suffix: '-1200',
+        maxHeight: 1200,
+        maxWidth: 1200,
+        aspect: "3:2"
+      },{
+        suffix: '-800',
+        maxHeight: 800,
+        maxWidth: 800,
+        aspect: "3:2"
+      },{
+        suffix: '-500',
+        maxHeight: 500,
+        maxWidth: 500,
+        aspect: "3:2"
+      },{
+        suffix: '-260',
+        maxHeight: 260,
+        maxWidth: 260,
+        aspect: "3:2"
+      },{
+        suffix: '-150',
+        maxHeight: 150,
+        maxWidth: 150,
+        aspect: "3:2"
+      },{
+        suffix: '-square-200',
+        maxHeight: 200,
+        maxWidth: 200,
+        aspect: "1:1"
+      },{
+        suffix: '-square-50',
+        maxHeight: 50,
+        maxWidth: 50,
+        aspect: "1:1"
+      }]
+    };
   });
 
   it('resisizes horizontal image', function(done) {
     this.timeout(10000);
 
     var image = {
       path: './assets/horizontal.jpg',
       width: 5184,
       height: 2623
     };
 
     var paths = [
       'assets/horizontal-full.jpg',
       'assets/horizontal-1200.jpg',
       'assets/horizontal-800.jpg',
       'assets/horizontal-500.jpg',
       'assets/horizontal-260.jpg',
       'assets/horizontal-150.jpg',
       'assets/horizontal-square-200.jpg',
       'assets/horizontal-square-50.jpg'
     ];
 
-    resize(image, versions, function(err, versions) {
+    resize(image, output, function(err, versions) {
       assert.ifError(err);
       assert(versions instanceof Array);
 
       for(var i = 0; i < versions.length; i++) {
         assert.equal(versions[i].path, paths[i]);
       }
 
       done();
     });
   });
 
   it('resisizes vertical image', function(done) {
     this.timeout(10000);
 
     var image = {
@@ -156,65 +158,65 @@ describe('resize()', function() {
       width: 1929,
       height: 3456
     };
 
     var paths = [
       'assets/vertical-full.jpg',
       'assets/vertical-1200.jpg',
       'assets/vertical-800.jpg',
       'assets/vertical-500.jpg',
       'assets/vertical-260.jpg',
       'assets/vertical-150.jpg',
       'assets/vertical-square-200.jpg',
       'assets/vertical-square-50.jpg'
     ];
 
-    resize(image, versions, function(err, versions) {
+    resize(image, output, function(err, versions) {
       assert.ifError(err);
       assert(versions instanceof Array);
 
       for(var i = 0; i < versions.length; i++) {
         assert.equal(versions[i].path, paths[i]);
       }
 
       done();
     });
   });
 
   it('resizes transparent image', function(done) {
     this.timeout(10000);
 
     var image = {
       path: './assets/transparent.png',
       width: 800,
       height: 600
     };
 
-    for (var i = 0; i < versions.length; i++) {
-      versions[i].flatten = true;
-      versions[i].background = 'red';
-      versions[i].format = 'jpg';
+    for (var i = 0; i < output.versions.length; i++) {
+      output.versions[i].flatten = true;
+      output.versions[i].background = 'red';
+      output.versions[i].format = 'jpg';
     }
 
     var paths = [
       'assets/transparent-full.jpg',
       'assets/transparent-1200.jpg',
       'assets/transparent-800.jpg',
       'assets/transparent-500.jpg',
       'assets/transparent-260.jpg',
       'assets/transparent-150.jpg',
       'assets/transparent-square-200.jpg',
       'assets/transparent-square-50.jpg'
     ];
 
-    resize(image, versions, function(err, versions) {
+    resize(image, output, function(err, versions) {
       assert.ifError(err);
       assert(versions instanceof Array);
 
       for(var i = 0; i < versions.length; i++) {
         assert.equal(versions[i].path, paths[i]);
       }
 
       done();
     });
   });
 });
