diff --git a/.gitattributes b/.gitattributes
new file mode 100644
index 0000000..759c2c5
--- /dev/null
+++ b/.gitattributes
@@ -0,0 +1,14 @@
+# Enforce Unix newlines
+*.*     text eol=lf
+*.css   text eol=lf
+*.html  text eol=lf
+*.js    text eol=lf
+*.json  text eol=lf
+*.less  text eol=lf
+*.md    text eol=lf
+*.yml   text eol=lf
+
+*.jpg binary
+*.gif binary
+*.png binary
+*.jpeg binary
\ No newline at end of file
diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..096c55d
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,52 @@
+# Numerous always-ignore extensions
+*.csv
+*.dat
+*.diff
+*.err
+*.gz
+*.log
+*.orig
+*.out
+*.pid
+*.rar
+*.rej
+*.seed
+*.swo
+*.swp
+*.vi
+*.yo-rc.json
+*.zip
+*~
+.ruby-version
+lib-cov
+npm-debug.log
+
+# Always-ignore dirs
+/bower_components/
+/node_modules/
+/temp/
+/tmp/
+/vendor/
+_gh_pages
+
+# OS or Editor folders
+*.esproj
+*.komodoproject
+.komodotools
+*.sublime-*
+._*
+.cache
+.DS_Store
+.idea
+.project
+.settings
+.tmproj
+nbproject
+Thumbs.db
+
+# grunt-html-validation
+validation-status.json
+validation-report.json
+
+# misc
+TODO.md
\ No newline at end of file
diff --git a/.jshintrc b/.jshintrc
new file mode 100644
index 0000000..e4104fe
--- /dev/null
+++ b/.jshintrc
@@ -0,0 +1,22 @@
+{
+  "esnext": true,
+  "boss": true,
+  "curly": true,
+  "eqeqeq": true,
+  "eqnull": true,
+  "immed": true,
+  "latedef": true,
+  "newcap": true,
+  "noarg": true,
+  "node": true,
+  "sub": true,
+  "undef": true,
+  "unused": true,
+  "globals": {
+    "define": true,
+    "before": true,
+    "after": true,
+    "describe": true,
+    "it": true
+  }
+}
\ No newline at end of file
diff --git a/.travis.yml b/.travis.yml
new file mode 100644
index 0000000..ff74a05
--- /dev/null
+++ b/.travis.yml
@@ -0,0 +1,3 @@
+language: node_js
+node_js:
+  - '0.10'
\ No newline at end of file
diff --git a/.verbrc.md b/.verbrc.md
new file mode 100644
index 0000000..8bc895a
--- /dev/null
+++ b/.verbrc.md
@@ -0,0 +1,44 @@
+# {%= name %} {%= badge("fury") %}
+
+> {%= description %}
+
+## Related
+
+ - [object.defaults](https://github.com/jonschlinkert/object.defaults)
+ - [object.filter](https://github.com/jonschlinkert/object.filter)
+ - [object.map](https://github.com/jonschlinkert/object.map)
+ - [object.mixin](https://github.com/jonschlinkert/object.mixin)
+ - [object.omit](https://github.com/jonschlinkert/object.omit)
+ - [object.pick](https://github.com/jonschlinkert/object.pick)
+ - [object.reduce](https://github.com/jonschlinkert/object.reduce)
+
+
+## Usage
+
+```js
+var defaults = require('{%= name %}');
+
+defaults({a: {one: 'one'}}, {a: {two: 'two'}})
+//=> {a: {one: 'one', two: 'two'}};
+```
+
+## Install
+{%= include("install-npm", {save: true}) %}
+{%= include("install-bower", {save: true}) %}
+
+## Run tests
+
+```bash
+npm test
+```
+
+## Author
+{%= include("author") %}
+
+## License
+{%= copyright() %}
+{%= license() %}
+
+***
+
+{%= include("footer") %}
\ No newline at end of file
diff --git a/LICENSE-MIT b/LICENSE-MIT
new file mode 100644
index 0000000..9bd36fe
--- /dev/null
+++ b/LICENSE-MIT
@@ -0,0 +1,22 @@
+Copyright (c) 2014 Jon Schlinkert, contributors.
+
+Permission is hereby granted, free of charge, to any person
+obtaining a copy of this software and associated documentation
+files (the "Software"), to deal in the Software without
+restriction, including without limitation the rights to use,
+copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following
+conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+OTHER DEALINGS IN THE SOFTWARE.
diff --git a/README.md b/README.md
new file mode 100644
index 0000000..dba2a76
--- /dev/null
+++ b/README.md
@@ -0,0 +1,56 @@
+# defaults-deep [![NPM version](https://badge.fury.io/js/defaults-deep.svg)](http://badge.fury.io/js/defaults-deep)
+
+> Like `extend` but recursively copies only the missing properties/values to the target object.
+
+## Related
+
+ - [object.defaults](https://github.com/jonschlinkert/object.defaults)
+ - [object.filter](https://github.com/jonschlinkert/object.filter)
+ - [object.map](https://github.com/jonschlinkert/object.map)
+ - [object.mixin](https://github.com/jonschlinkert/object.mixin)
+ - [object.omit](https://github.com/jonschlinkert/object.omit)
+ - [object.pick](https://github.com/jonschlinkert/object.pick)
+ - [object.reduce](https://github.com/jonschlinkert/object.reduce)
+
+
+## Usage
+
+```js
+var defaults = require('defaults-deep');
+
+defaults({a: {one: 'one'}}, {a: {two: 'two'}})
+//=> {a: {one: 'one', two: 'two'}};
+```
+
+## Install
+#### Install with [npm](npmjs.org)
+
+```bash
+npm i defaults-deep --save
+```
+#### Install with [bower](https://github.com/bower/bower)
+
+```bash
+bower install defaults-deep --save
+```
+
+## Run tests
+
+```bash
+npm test
+```
+
+## Author
+
+**Jon Schlinkert**
+ 
++ [github/jonschlinkert](https://github.com/jonschlinkert)
++ [twitter/jonschlinkert](http://twitter.com/jonschlinkert) 
+
+## License
+Copyright (c) 2014 Jon Schlinkert, contributors.  
+Released under the MIT license
+
+***
+
+_This file was generated by [verb-cli](https://github.com/assemble/verb-cli) on October 12, 2014._
\ No newline at end of file
diff --git a/index.js b/index.js
new file mode 100644
index 0000000..3857550
--- /dev/null
+++ b/index.js
@@ -0,0 +1,40 @@
+/*!
+ * defaults-deep <https://github.com/jonschlinkert/defaults-deep>
+ *
+ * Copyright (c) 2014 Jon Schlinkert, contributors.
+ * Licensed under the MIT license.
+ */
+
+'use strict';
+
+var isObject = require('is-plain-object');
+var slice = require('array-slice');
+var forOwn = require('for-own');
+
+module.exports = function defaults(o, objects) {
+  if (o == null) return {};
+  if (objects == null) return o;
+  forEach(slice(arguments, 1), function (obj) {
+    forOwn(obj, function (value, key) {
+      var val = o[key];
+      if (val == null) {
+        o[key] = value;
+      } else if (isObject(val) && isObject(value)) {
+        defaults(val, value);
+      }
+    });
+  });
+  return o;
+};
+
+function forEach(arr, cb) {
+  if (arr == null) return;
+  var len = arr.length;
+  var i = -1;
+
+  while (++i < len) {
+    if (cb(arr[i], i, arr) === false) {
+      break;
+    }
+  }
+}
\ No newline at end of file
diff --git a/package.json b/package.json
new file mode 100644
index 0000000..d7b30d8
--- /dev/null
+++ b/package.json
@@ -0,0 +1,56 @@
+{
+  "name": "defaults-deep",
+  "description": "Like `extend` but recursively copies only the missing properties/values to the target object.",
+  "version": "0.1.0",
+  "homepage": "https://github.com/jonschlinkert/defaults-deep",
+  "author": {
+    "name": "Jon Schlinkert",
+    "url": "https://github.com/jonschlinkert"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/jonschlinkert/defaults-deep.git"
+  },
+  "bugs": {
+    "url": "https://github.com/jonschlinkert/defaults-deep/issues"
+  },
+  "licenses": [
+    {
+      "type": "MIT",
+      "url": "https://github.com/jonschlinkert/defaults-deep/blob/master/LICENSE-MIT"
+    }
+  ],
+  "main": "index.js",
+  "engines": {
+    "node": ">=0.10.0"
+  },
+  "scripts": {
+    "test": "mocha -R spec"
+  },
+  "devDependencies": {
+    "mocha": "*",
+    "should": "^4.0.4",
+    "verb": ">= 0.2.6"
+  },
+  "keywords": [
+    "copy",
+    "default",
+    "defaults",
+    "extend",
+    "merge",
+    "object",
+    "properties",
+    "property",
+    "value",
+    "values",
+    "javascript",
+    "js",
+    "util",
+    "utils"
+  ],
+  "dependencies": {
+    "array-slice": "^0.2.2",
+    "for-own": "^0.1.2",
+    "is-plain-object": "^0.1.0"
+  }
+}
diff --git a/test.js b/test.js
new file mode 100644
index 0000000..24b64e3
--- /dev/null
+++ b/test.js
@@ -0,0 +1,39 @@
+/*!
+ * defaults-deep <https://github.com/jonschlinkert/defaults-deep>
+ *
+ * Copyright (c) 2014 Jon Schlinkert, contributors.
+ * Licensed under the MIT License
+ */
+
+'use strict';
+
+var should = require('should');
+var deepDefaults = require('./');
+
+describe('deep-defaults', function () {
+  it('should copy only missing properties defaults', function () {
+    deepDefaults({a: 'c'}, {a: 'bbb', d: 'c'}).should.eql({a: 'c', d: 'c'});
+  });
+
+  it('should copy properties from multiple objects', function () {
+    deepDefaults({a: 'b'}, {c: 'd'}, {e: 'f'}).should.eql({a: 'b', c: 'd', e: 'f'});
+  });
+
+  it('should fill in values that are null', function () {
+    deepDefaults({a: null}, {a: 'c', d: 'c'}).should.eql({a: 'c', d: 'c'});
+  });
+
+  it('should copy nested values.', function () {
+    deepDefaults({a: {b: 'c'}}, {a: {d: 'e'}}).should.eql({a: {b: 'c', d: 'e'}});
+  });
+
+  it('should clone when an empty object is passed as the first arg.', function () {
+    var obj = {};
+    deepDefaults(obj, {a: {b: 'c'}}, {a: {d: 'e'}});
+    obj.should.eql({a: {b: 'c', d: 'e'}});
+  });
+
+  it('should return an empty object when the first arg is null.', function () {
+    deepDefaults(null).should.eql({});
+  });
+});
